$FreeBSD$

--- scrotwm.c.orig	2012-01-11 19:42:57.000000000 +0100
+++ scrotwm.c	2012-01-11 21:36:07.000000000 +0100
@@ -4481,14 +4481,16 @@
 					" -nb $bar_color"
 					" -nf $bar_font_color"
 					" -sb $bar_border"
-					" -sf $bar_color",	0);
+					" -sf $bar_color"
+					" -b",	0);
 	setconfspawn("name_workspace",	"dmenu"
 					" -p Workspace"
 					" -fn $bar_font"
 					" -nb $bar_color"
 					" -nf $bar_font_color"
 					" -sb $bar_border"
-					" -sf $bar_color",	0);
+					" -sf $bar_color"
+					" -b",	0);
 }
 
 /* key bindings */
