From 06d69c5277f08d9a57f74836c247d3927ca390e6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Pawe=C5=82=20Chmielowski?= <pchmielowski@process-one.net>
Date: Mon, 7 Jan 2019 16:08:58 +0100
Subject: [PATCH] Fix decompressing of custom elements with custom namespace

--- src/xml_compress.erl.orig	2018-12-21 09:44:15 UTC
+++ src/xml_compress.erl
@@ -506,8 +506,8 @@ decode_child(<<2:8, Rest/binary>>, PNs, J1, J2) ->
   {Children, Rest4} = decode_children(Rest3, PNs, J1, J2),
   {{xmlel, Name, Attrs, Children}, Rest4};
 decode_child(<<3:8, Rest/binary>>, PNs, J1, J2) ->
-  {Name, Rest2} = decode_string(Rest),
-  {Ns, Rest3} = decode_string(Rest2),
+  {Ns, Rest2} = decode_string(Rest),
+  {Name, Rest3} = decode_string(Rest2),
   {Attrs, Rest4} = decode_attrs(Rest3),
   {Children, Rest5} = decode_children(Rest4, Ns, J1, J2),
   {{xmlel, Name, add_ns(PNs, Ns, Attrs), Children}, Rest5};
 src/xml_compress.erl       | 4 ++--
 tools/xml_compress_gen.erl | 4 ++--
 2 files changed, 4 insertions(+), 4 deletions(-)

--- tools/xml_compress_gen.erl.orig	2018-12-21 09:44:15 UTC
+++ tools/xml_compress_gen.erl
@@ -117,8 +117,8 @@ gen_decode(Dev, Data, VerId) ->
 		   "  {Children, Rest4} = decode_children(Rest3, PNs, J1, J2),~n"
 		   "  {{xmlel, Name, Attrs, Children}, Rest4};~n", []),
     io:format(Dev, "decode_child(<<3:8, Rest/binary>>, PNs, J1, J2) ->~n"
-		   "  {Name, Rest2} = decode_string(Rest),~n"
-		   "  {Ns, Rest3} = decode_string(Rest2),~n"
+		   "  {Ns, Rest2} = decode_string(Rest),~n"
+		   "  {Name, Rest3} = decode_string(Rest2),~n"
 		   "  {Attrs, Rest4} = decode_attrs(Rest3),~n"
 		   "  {Children, Rest5} = decode_children(Rest4, Ns, J1, J2),~n"
 		   "  {{xmlel, Name, add_ns(PNs, Ns, Attrs), Children}, Rest5};~n", []),
