Index: ejabberd_s2s_in.erl
===================================================================
--- ejabberd_s2s_in.erl	(revision 1026)
+++ ejabberd_s2s_in.erl	(working copy)
@@ -30,8 +30,7 @@
 
 -include("ejabberd.hrl").
 -include("jlib.hrl").
--include_lib("ssl/include/PKIX1Explicit88.hrl").
--include_lib("ssl/include/PKIX1Implicit88.hrl").
+-include_lib("ssl/include/OTP-PKIX.hrl").
 -include("XmppAddr.hrl").
 
 -define(DICT, dict).
