Index: src/applets/main.c
===================================================================
RCS file: /home/cvs/DeforaOS/Apps/Desktop/src/Panel/src/applets/main.c,v
retrieving revision 1.36
diff -p -u -r1.36 main.c
--- src/applets/main.c	21 Mar 2012 21:07:42 -0000	1.36
+++ src/applets/main.c	23 Oct 2012 09:08:40 -0000
@@ -209,14 +209,21 @@ static GtkWidget * _main_applications(Ma

 static void _applications_on_activate(gpointer data)
 {
-	char const * program = data;
+	char * program = data;
+	char * p;

	if(program == NULL)
		return;
+	if((program = strdup(program)) == NULL)
+		return; /* XXX report error */
+	/* XXX crude way to ignore %f, %F, %u and %U */
+	if((p = strchr(program, '%')) != NULL)
+		*p = '\0';
 #ifdef DEBUG
	fprintf(stderr, "DEBUG: %s() \"%s\"", __func__, program);
 #endif
	g_spawn_command_line_async(program, NULL);
+	free(program);
 }

 static void _applications_categories(GtkWidget * menu, GtkWidget ** menus)
