--- common/posib_err.hpp	2004-12-17 09:23:33.000000000 +0100
+++ common/posib_err.hpp	2011-05-09 19:46:45.000000000 +0200
@@ -158,6 +158,16 @@
     void del();
   };
 
+  template <>
+  class PosibErr<void> : public PosibErrBase
+  {
+  public:
+    PosibErr(const PosibErrBase & other) 
+      : PosibErrBase(other) {}
+
+    PosibErr() {}
+  };
+
   template <typename Ret>
   class PosibErr : public PosibErrBase
   {
@@ -186,16 +196,6 @@
     Ret data;
   };
 
-  template <>
-  class PosibErr<void> : public PosibErrBase
-  {
-  public:
-    PosibErr(const PosibErrBase & other) 
-      : PosibErrBase(other) {}
-
-    PosibErr() {}
-  };
-
 //
 //
 //
--- common/vector.hpp	2011-05-09 19:51:59.000000000 +0200
+++ common/vector.hpp	2011-05-09 19:51:54.000000000 +0200
@@ -24,7 +24,7 @@
       this->push_back(t);
     }
     void append(const T * begin, unsigned int size) {
-      insert(this->end(), begin, begin+size);
+      this->insert(this->end(), begin, begin+size);
     }
     void append(const T * begin, const T * end) {
       insert(this->end(), begin, end);
--- common/hash.hpp	2011-05-09 19:52:05.000000000 +0200
+++ common/hash.hpp	2011-05-09 19:51:35.000000000 +0200
@@ -318,7 +318,7 @@
       : Base(s, Parms(h,e)) {}
     data_type & operator[](const key_type & k) 
     {
-      return (*((insert(value_type(k, data_type()))).first)).second;
+      return (*((this->insert(value_type(k, data_type()))).first)).second;
     }
   };
 
