$FreeBSD$

--- src/ogmrip-pref.c.orig	2005-02-24 02:25:46.143332080 +0100
+++ src/ogmrip-pref.c	2005-02-24 02:27:45.306216568 +0100
@@ -124,22 +124,22 @@
 }
 
 static void
-ogmrip_pref_output_dir_notified (GConfClient *gconf, guint id, GConfEntry *entry, GtkWidget *label)
+ogmrip_pref_output_dir_notified (GConfClient *gconf, guint id, GConfEntry *entry, GtkWidget *widget)
 {
   /*
    * Sanity checks
    */
-  gtk_label_set_text (GTK_LABEL (label),
+  gtk_entry_set_text (GTK_ENTRY (widget),
       gconf_value_get_string (entry->value));
 }
 
 static void
-ogmrip_pref_tmp_dir_notified (GConfClient *gconf, guint id, GConfEntry *entry, GtkWidget *label)
+ogmrip_pref_tmp_dir_notified (GConfClient *gconf, guint id, GConfEntry *entry, GtkWidget *widget)
 {
   /*
    * Sanity checks
    */
-  gtk_label_set_text (GTK_LABEL (label),
+  gtk_entry_set_text (GTK_ENTRY (widget),
       gconf_value_get_string (entry->value));
 }
 
