diff -urN -x .svn ../../vendor/vpopmail/vchkpw.c ./vchkpw.c
--- ../../vendor/vpopmail/vchkpw.c	Tue Oct  3 17:40:12 2006
+++ ./vchkpw.c	Tue Oct  3 17:58:48 2006
@@ -441,7 +441,8 @@
     /* Check CRAM-MD5 auth */
     if(ConnType == SMTP_CONN) {
       /* printf("vchkpw: smtp auth\n"); */
-      cramaccepted = authcram(ThePass,TheChallenge,vpw->pw_clear_passwd);
+      /* cramaccepted = authcram(ThePass,TheChallenge,vpw->pw_clear_passwd); */
+      cramaccepted = authcram(TheChallenge,ThePass,vpw->pw_clear_passwd);
       if(cramaccepted == 0) strcpy(AuthType, "CRAM-MD5");
     }
 
