--- vmnet-only/freebsd/vmnet_linux.c.orig	Tue Feb 11 22:12:40 2003
+++ vmnet-only/freebsd/vmnet_linux.c	Tue Feb 11 22:25:20 2003
@@ -32,6 +32,7 @@
 #include <sys/socket.h>
 #include <sys/proc.h>
 #include <sys/file.h>
+#include <sys/filedesc.h>
 
 #include <net/if.h>
 
--- vmmon-only/freebsd/driver.c.orig	Tue Feb 11 22:12:39 2003
+++ vmmon-only/freebsd/driver.c	Tue Feb 11 22:22:35 2003
@@ -45,8 +45,9 @@
 #include <sys/kernel.h>
 #include <sys/module.h>
 #include <sys/conf.h>
 #include <sys/selinfo.h>
 #include <sys/fcntl.h>
+#include <sys/filedesc.h>
 #include <sys/malloc.h>
 #include <sys/proc.h>
 #include <sys/ioccom.h>
--- vmmon-only/freebsd/linux_emu.c.orig	Tue Feb 11 22:12:40 2003
+++ vmmon-only/freebsd/linux_emu.c	Tue Feb 11 22:24:42 2003
@@ -36,6 +36,7 @@
 #include <sys/disklabel.h>
 #include <sys/cdio.h>
 #include <sys/file.h>
+#include <sys/filedesc.h>
 #include <sys/selinfo.h>
 
 #include <sys/fdcio.h>
