--- vmmon-only/freebsd/hostif.c.orig	Tue Dec  4 20:49:08 2001
+++ vmmon-only/freebsd/hostif.c	Tue Dec  4 20:49:44 2001
@@ -170,7 +170,7 @@
 {
 #define DEB(x) 
    caddr_t addr = (caddr_t)VPN_2_VA(ppn);
-   pt_entry_t pteptr = (pt_entry_t)vtopte(addr);
+   pt_entry_t *pteptr = vtopte(addr);
    PTE pte;
 
    DEB(printf("FindMPN: for page %d address %p(phys %p) pteptr %p", ppn, addr, (caddr_t)vtophys(addr), pteptr));
