diff --git pngpriv.h pngpriv.h
index 583c26f9bd..1ebddeccaf 100644
--- pngpriv.h
+++ pngpriv.h
@@ -492,7 +492,9 @@
    static_cast<type>(static_cast<const void*>(value))
 #else
 #  define png_voidcast(type, value) (value)
-#  ifdef _WIN64
+#  ifdef __UINTPTR_TYPE__
+      typedef __UINTPTR_TYPE__ png_ptruint;
+#  elif defined(_WIN64)
 #     ifdef __GNUC__
          typedef unsigned long long png_ptruint;
 #     else
