--- Makefile.orig	2018-06-02 13:26:25.000000000 -0400
+++ Makefile	2018-06-06 22:23:38.079036000 -0400
@@ -79,7 +79,7 @@
 # Construct the architecture-version string, which will be used to name the
 # libraries upon installation.
 VERSION                        := $(shell cat version)
-ARCH_VERS                      := $(ARCH)-$(VERSION)
+ARCH_VERS                      := 
 
 # --- Library names ---
 ALL_FLAMEC_LIB_NAME            := libflame.a
@@ -117,25 +117,25 @@
 # --- Define install target names for static libraries ---
 MK_FLAMEC_LIBS                    := $(MK_ALL_FLAMEC_LIB)
 MK_FLAMEC_LIBS_INST               := $(patsubst $(BASE_LIB_DIR)/%.a, \
-                                                $(INSTALL_PREFIX)/lib/%.a, \
+                                                $(STAGEDIR)$(INSTALL_PREFIX)/lib/%.a, \
                                                 $(MK_FLAMEC_LIBS))
 MK_FLAMEC_LIBS_INST_W_ARCH_VERS   := $(patsubst $(BASE_LIB_DIR)/%.a, \
-                                                $(INSTALL_PREFIX)/lib/%-$(ARCH_VERS).a, \
+                                                $(STAGEDIR)$(INSTALL_PREFIX)/lib/%.a, \
                                                 $(MK_FLAMEC_LIBS))
 #MK_FLAMEC_LIBS_INST_W_ARCH_ONLY   := $(patsubst $(BASE_LIB_DIR)/%.a, \
-#                                                $(INSTALL_PREFIX)/lib/%-$(ARCH).a, \
+#                                                $(STAGEDIR)$(INSTALL_PREFIX)/lib/%.a, \
 #                                                $(MK_FLAMEC_LIBS))
 
 # --- Define install target names for dynamic libraries ---
 MK_FLAMEC_DLLS                    := $(MK_ALL_FLAMEC_DLL)
 MK_FLAMEC_DLLS_INST               := $(patsubst $(BASE_LIB_DIR)/%.so, \
-                                                $(INSTALL_PREFIX)/lib/%.so, \
+                                                $(STAGEDIR)$(INSTALL_PREFIX)/lib/%.so, \
                                                 $(MK_FLAMEC_DLLS))
 MK_FLAMEC_DLLS_INST_W_ARCH_VERS   := $(patsubst $(BASE_LIB_DIR)/%.so, \
-                                                $(INSTALL_PREFIX)/lib/%-$(ARCH_VERS).so, \
+                                                $(STAGEDIR)$(INSTALL_PREFIX)/lib/%.so, \
                                                 $(MK_FLAMEC_DLLS))
 #MK_FLAMEC_DLLS_INST_W_ARCH_ONLY   := $(patsubst $(BASE_LIB_DIR)/%.so, \
-#                                                $(INSTALL_PREFIX)/lib/%-$(ARCH).so, \
+#                                                $(STAGEDIR)$(INSTALL_PREFIX)/lib/%-$(ARCH).so, \
 #                                                $(MK_FLAMEC_DLLS))
 
 # --- Determine which libraries to build ---
@@ -159,9 +159,9 @@
 endif
 
 # --- Set the include directory names ---
-MK_INCL_DIR_INST                  := $(INSTALL_PREFIX)/include
-MK_INCL_DIR_INST_W_ARCH_VERS      := $(INSTALL_PREFIX)/include-$(ARCH_VERS)
-#MK_INCL_DIR_INST_W_ARCH_ONLY      := $(INSTALL_PREFIX)/include-$(ARCH)
+MK_INCL_DIR_INST                  := $(STAGEDIR)$(INSTALL_PREFIX)/include
+MK_INCL_DIR_INST_W_ARCH_VERS      := $(STAGEDIR)$(INSTALL_PREFIX)/include
+#MK_INCL_DIR_INST_W_ARCH_ONLY      := $(STAGEDIR)$(INSTALL_PREFIX)/include-$(ARCH)
 
 
 
@@ -452,22 +452,22 @@
 	@$(INSTALL) -m 0644 $(MK_HEADER_FILES) $(@)
 endif
 
-$(INSTALL_PREFIX)/lib/%-$(ARCH_VERS).a: $(BASE_LIB_DIR)/%.a
+$(STAGEDIR)$(INSTALL_PREFIX)/lib/%.a: $(BASE_LIB_DIR)/%.a
 ifeq ($(FLA_ENABLE_VERBOSE_MAKE_OUTPUT),yes)
 	$(INSTALL) -m 0755 -d $(@D)
 	$(INSTALL) -m 0644 $< $@
 else
-	@echo "Installing $(@F) into $(INSTALL_PREFIX)/lib/"
+	@echo "Installing $(@F) into $(STAGEDIR)$(INSTALL_PREFIX)/lib/"
 	@$(INSTALL) -m 0755 -d $(@D)
 	@$(INSTALL) -m 0644 $< $@
 endif
 
-$(INSTALL_PREFIX)/lib/%-$(ARCH_VERS).so: $(BASE_LIB_DIR)/%.so
+$(STAGEDIR)$(INSTALL_PREFIX)/lib/%.so: $(BASE_LIB_DIR)/%.so
 ifeq ($(FLA_ENABLE_VERBOSE_MAKE_OUTPUT),yes)
 	$(INSTALL) -m 0755 -d $(@D)
 	$(INSTALL) -m 0644 $< $@
 else
-	@echo "Installing $(@F) into $(INSTALL_PREFIX)/lib/"
+	@echo "Installing $(@F) into $(STAGEDIR)$(INSTALL_PREFIX)/lib/"
 	@$(INSTALL) -m 0755 -d $(@D)
 	@$(INSTALL) -m 0644 $< $@
 endif
@@ -475,30 +475,30 @@
 
 
 # --- Install-symlinks rules ---
-install-lib-symlinks: check-config $(MK_LIBS_INST)
+#install-lib-symlinks: check-config $(MK_LIBS_INST)
+#
+#install-header-symlinks: check-config $(MK_INCL_DIR_INST)
+#
+#$(MK_INCL_DIR_INST): $(MK_INCL_DIR_INST_W_ARCH_VERS)
+#ifeq ($(FLA_ENABLE_VERBOSE_MAKE_OUTPUT),yes)
+#	$(SYMLINK) $(<F) $(@F)
+#	$(MV) $(@F) $(INSTALL_PREFIX)
+#else
+#	@echo "Installing symlink $(@F) into $(INSTALL_PREFIX)/"
+#	@$(SYMLINK) $(<F) $(@F)
+#	@$(MV) $(@F) $(INSTALL_PREFIX)
+#endif
+#
+#$(INSTALL_PREFIX)/lib/%.a: $(INSTALL_PREFIX)/lib/%-$(ARCH_VERS).a
+#ifeq ($(FLA_ENABLE_VERBOSE_MAKE_OUTPUT),yes)
+#	$(SYMLINK) $(<F) $(@F)
+#	$(MV) $(@F) $(INSTALL_PREFIX)/lib/
+#else
+#	@echo "Installing symlink $(@F) into $(INSTALL_PREFIX)/lib/"
+#	@$(SYMLINK) $(<F) $(@F)
+#	@$(MV) $(@F) $(INSTALL_PREFIX)/lib/
+#endif
 
-install-header-symlinks: check-config $(MK_INCL_DIR_INST)
-
-$(MK_INCL_DIR_INST): $(MK_INCL_DIR_INST_W_ARCH_VERS)
-ifeq ($(FLA_ENABLE_VERBOSE_MAKE_OUTPUT),yes)
-	$(SYMLINK) $(<F) $(@F)
-	$(MV) $(@F) $(INSTALL_PREFIX)
-else
-	@echo "Installing symlink $(@F) into $(INSTALL_PREFIX)/"
-	@$(SYMLINK) $(<F) $(@F)
-	@$(MV) $(@F) $(INSTALL_PREFIX)
-endif
-
-$(INSTALL_PREFIX)/lib/%.a: $(INSTALL_PREFIX)/lib/%-$(ARCH_VERS).a
-ifeq ($(FLA_ENABLE_VERBOSE_MAKE_OUTPUT),yes)
-	$(SYMLINK) $(<F) $(@F)
-	$(MV) $(@F) $(INSTALL_PREFIX)/lib/
-else
-	@echo "Installing symlink $(@F) into $(INSTALL_PREFIX)/lib/"
-	@$(SYMLINK) $(<F) $(@F)
-	@$(MV) $(@F) $(INSTALL_PREFIX)/lib/
-endif
-
 #$(INSTALL_PREFIX)/lib/%-$(ARCH).a: $(INSTALL_PREFIX)/lib/%-$(ARCH_VERS).a
 #ifeq ($(FLA_ENABLE_VERBOSE_MAKE_OUTPUT),yes)
 #	$(SYMLINK) $(<F) $(@F)
@@ -509,15 +509,15 @@
 #	@$(MV) $(@F) $(INSTALL_PREFIX)/lib/
 #endif
 
-$(INSTALL_PREFIX)/lib/%.so: $(INSTALL_PREFIX)/lib/%-$(ARCH_VERS).so
-ifeq ($(FLA_ENABLE_VERBOSE_MAKE_OUTPUT),yes)
-	$(SYMLINK) $(<F) $(@F)
-	$(MV) $(@F) $(INSTALL_PREFIX)/lib/
-else
-	@echo "Installing symlink $(@F) into $(INSTALL_PREFIX)/lib/"
-	@$(SYMLINK) $(<F) $(@F)
-	@$(MV) $(@F) $(INSTALL_PREFIX)/lib/
-endif
+#$(INSTALL_PREFIX)/lib/%.so: $(INSTALL_PREFIX)/lib/%-$(ARCH_VERS).so
+#ifeq ($(FLA_ENABLE_VERBOSE_MAKE_OUTPUT),yes)
+#	$(SYMLINK) $(<F) $(@F)
+#	$(MV) $(@F) $(INSTALL_PREFIX)/lib/
+#else
+#	@echo "Installing symlink $(@F) into $(INSTALL_PREFIX)/lib/"
+#	@$(SYMLINK) $(<F) $(@F)
+#	@$(MV) $(@F) $(INSTALL_PREFIX)/lib/
+#endif
 
 #$(INSTALL_PREFIX)/lib/%-$(ARCH).so: $(INSTALL_PREFIX)/lib/%-$(ARCH_VERS).so
 #ifeq ($(FLA_ENABLE_VERBOSE_MAKE_OUTPUT),yes)
