diff --git a/llvm/config.bzl b/llvm/config.bzl
--- a/llvm/config.bzl
+++ b/llvm/config.bzl
@@ -85,7 +85,7 @@
 os_defines = select({
     "@bazel_tools//src/conditions:windows": win32_defines,
     "@bazel_tools//src/conditions:darwin": macos_defines,
-    "@bazel_tools//src/conditions:freebsd": posix_defines,
+    "@platforms//os:freebsd": posix_defines,
     "//conditions:default": linux_defines,
 })
 
@@ -108,6 +108,7 @@
     "@bazel_tools//src/conditions:linux_ppc64le": native_arch_defines("PowerPC", "powerpc64le-unknown-linux-gnu"),
     "@bazel_tools//src/conditions:linux_riscv64": native_arch_defines("RISCV", "riscv64-unknown-linux-gnu"),
     "@bazel_tools//src/conditions:linux_s390x": native_arch_defines("SystemZ", "systemz-unknown-linux_gnu"),
+    "@platforms//os:freebsd": native_arch_defines("X86", "x86_64-unknown-freebsd14-elf"),
     "//conditions:default": native_arch_defines("X86", "x86_64-unknown-linux-gnu"),
 }) + [
     "LLVM_VERSION_MAJOR={}".format(LLVM_VERSION_MAJOR),
