Index: configure
===================================================================
--- configure (revision 16772)
+++ configure (revision 16895)
@@ -164,4 +164,5 @@
   echo "  --disable-x11            disable X11 support"
   echo "  --x11-path=X11LOC        location of X11 include files [$x11_include_path-path]"
+#  echo "  --enable-glx-procaddrarb use glXGetProcAddressARB() instead of glXGetProcAddress()"
   echo "  --disable-xrandr         disable X11 resolution switching"
   echo "  --disable-xv             disable XVideo   (X11 video output accel.)"
@@ -905,4 +906,5 @@
     dvdv
     fribidi
+    glx_proc_addr_arb
 '
 
@@ -1190,4 +1192,5 @@
 xvmc_vld="yes"
 xvmcw="yes"
+glx_proc_addr_arb="no"
 
 # libraries
@@ -1353,4 +1356,8 @@
   ;;
   --xvmc-lib=*) xvmc_lib="$optval"
+  ;;
+  --enable-glx-procaddrarb) glx_proc_addr_arb="yes"
+  ;;
+  --disable-glx-procaddrarb) glx_proc_addr_arb="no"
   ;;
   --enable-mac-accel) dvdv="yes"
Index: libs/libmythtv/libmythtv.pro
===================================================================
--- libs/libmythtv/libmythtv.pro (revision 16142)
+++ libs/libmythtv/libmythtv.pro (revision 16895)
@@ -313,4 +313,6 @@
     using_opengl_video:SOURCES += openglvideo.cpp
 
+    using_glx_proc_addr_arb:DEFINES += USING_GLX_PROC_ADDR_ARB
+
     # Misc. frontend
     HEADERS += guidegrid.h              infostructs.h
Index: /branches/release-0-21-fixes/mythtv/libs/libmythtv/util-opengl.cpp
===================================================================
--- libs/libmythtv/util-opengl.cpp (revision 15141)
+++ libs/libmythtv/util-opengl.cpp (revision 16895)
@@ -307,5 +307,7 @@
     __GLXextFuncPtr ret = NULL;
 
-#if GLX_VERSION_1_4
+#if USING_GLX_PROC_ADDR_ARB
+    X11S(ret = glXGetProcAddressARB((const GLubyte*)procName.latin1()));
+#elif GLX_VERSION_1_4
     X11S(ret = glXGetProcAddress((const GLubyte*)procName.latin1()));
 #elif GLX_ARB_get_proc_address
