--- src/ghtml-webkit.c.orig	2012-10-21 14:12:06.000000000 +0200
+++ src/ghtml-webkit.c	2012-10-21 14:13:02.000000000 +0200
@@ -160,7 +160,8 @@
 		"/etc/ssl/certs/ca-certificates.crt",
 		"/etc/openssl/certs/ca-certificates.crt",
 		PREFIX "/etc/ssl/certs/ca-certificates.crt",
-		PREFIX "/etc/openssl/certs/ca-certificates.crt"
+		PREFIX "/etc/openssl/certs/ca-certificates.crt",
+		PREFIX "/share/certs/ca-root-nss.crt"
 	};
 	size_t i;
 #endif
