--- lib/engine/components/hal-dbus/Makefile.in.orig	2009-10-25 01:38:37.000000000 +0400
+++ lib/engine/components/hal-dbus/Makefile.in	2009-10-25 01:38:58.000000000 +0400
@@ -648,10 +648,10 @@
 
 
 hal-marshal.h: hal-marshal.list
-	$(LIBTOOL) --mode=execute glib-genmarshal --prefix=gm_hal_dbus_marshal $^ --header > $@.tmp && mv $@.tmp $@
+	$(LIBTOOL) --mode=execute glib-genmarshal --prefix=gm_hal_dbus_marshal $^ --header < hal-marshal.list > $@.tmp && mv $@.tmp $@
 
 hal-marshal.c: hal-marshal.list
-	$(LIBTOOL) --mode=execute glib-genmarshal --prefix=gm_hal_dbus_marshal $^ --body > $@.tmp && mv $@.tmp $@
+	$(LIBTOOL) --mode=execute glib-genmarshal --prefix=gm_hal_dbus_marshal $^ --body < hal-marshal.list > $@.tmp && mv $@.tmp $@
 
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
