--- lisp/mule/mule-init.el.orig	Fri Oct 20 14:08:56 2000
+++ lisp/mule/mule-init.el	Fri Oct 20 14:09:12 2000
@@ -76,7 +76,8 @@
 
 (defvar auto-language-alist
   '(("^ja" . "Japanese")
-    ("^zh" . "Chinese")
+    ("^zh_CN" . "Chinese-GB")
+    ("^zh_TW" . "Chinese-BIG5")
     ("^ko" . "Korean"))
   "Alist of LANG patterns vs. corresponding language environment.
 Each element looks like (REGEXP . LANGUAGE-ENVIRONMENT).
