--- status.c.orig	2011/04/18 20:49:05	1.73
+++ status.c	2011/04/24 22:32:07	1.74
@@ -1,4 +1,4 @@
-/* $OpenBSD: status.c,v 1.73 2011/04/18 19:49:05 nicm Exp $ */
+/* $OpenBSD: status.c,v 1.74 2011/04/24 21:32:07 nicm Exp $ */
 
 /*
  * Copyright (c) 2007 Nicholas Marriott <nicm@users.sourceforge.net>
@@ -408,6 +408,13 @@ status_replace1(struct client *c, struct session *s, s
 	case 'H':
 		if (gethostname(tmp, sizeof tmp) != 0)
 			fatal("gethostname failed");
+		ptr = tmp;
+		goto do_replace;
+	case 'h':
+		if (gethostname(tmp, sizeof tmp) != 0)
+			fatal("gethostname failed");
+		if ((ptr = strchr(tmp, '.')) != NULL)
+			*ptr = '\0';
 		ptr = tmp;
 		goto do_replace;
 	case 'I':
--- tmux.1.orig	2011/04/19 22:31:33	1.223
+++ tmux.1	2011/04/24 22:32:07	1.224
@@ -2005,6 +2005,7 @@ may contain any of the following special character seq
 .It Li "#(shell-command)" Ta "First line of the command's output"
 .It Li "#[attributes]" Ta "Colour or attribute change"
 .It Li "#H" Ta "Hostname of local host"
+.It Li "#h" Ta "Hostname of local host without the domain name"
 .It Li "#F" Ta "Current window flag"
 .It Li "#I" Ta "Current window index"
 .It Li "#P" Ta "Current pane index"
