diff -urN Net-FreeDB-0.06/FreeDB.pm roam/FreeDB.pm
--- FreeDB.pm	Wed Mar 20 22:19:50 2002
+++ FreeDB.pm	Mon Oct 17 13:03:00 2005
@@ -1,6 +1,6 @@
 package Net::FreeDB;
 
-use 5.006;
+use 5.00503;
 use strict;
 use warnings;
 use IO::Socket;
@@ -13,7 +13,9 @@
 require DynaLoader;
 use AutoLoader;
 
-our @ISA = qw(Exporter DynaLoader Net::Cmd IO::Socket::INET);
+use vars qw/@ISA %EXPORT_TAGS @EXPORT_OK @EXPORT $VERSION $ERROR/;
+
+@ISA = qw(Exporter DynaLoader Net::Cmd IO::Socket::INET);
 
 # Items to export into callers namespace by default. Note: do not export
 # names by default without a very good reason. Use EXPORT_OK instead.
@@ -22,22 +24,21 @@
 # This allows declaration	use Net::FreeDB ':all';
 # If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
 # will save memory.
-our %EXPORT_TAGS = ( 'all' => [ qw() ] );
+%EXPORT_TAGS = ( 'all' => [ qw() ] );
 
-our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
+@EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
 
-our @EXPORT = qw();
+@EXPORT = qw();
 
-our $VERSION = '0.06';
+$VERSION = '0.06';
 
-our $ERROR;
 sub AUTOLOAD {
     # This AUTOLOAD is used to 'autoload' constants from the constant()
     # XS function.  If a constant is not found then control is passed
     # to the AUTOLOAD in AutoLoader.
 
     my $constname;
-    our $AUTOLOAD;
+    use vars qw/$AUTOLOAD/;
     ($constname = $AUTOLOAD) =~ s/.*:://;
     croak "& not defined" if $constname eq 'constant';
     my $val = constant($constname, @_ ? $_[0] : 0);
diff -urN Net-FreeDB-0.06/FreeDB.xs.linux roam/FreeDB.xs.linux
--- FreeDB.xs.linux	Wed Mar 20 22:19:50 2002
+++ FreeDB.xs.linux	Mon Oct 17 12:09:00 2005
@@ -36,7 +38,8 @@
 	RETVAL
 
 char*
-discid(char* dev)
+discid(dev)
+	char *dev;
     INIT:
 		int i;
 		char id[30];
@@ -54,7 +57,8 @@
 		RETVAL
 
 SV*
-discinfo(char* dev)
+discinfo(dev)
+	char *dev;
     INIT:
 	int i;
         char id[30];
diff -urN Net-FreeDB-0.06/lib/discid.h roam/lib/discid.h
--- lib/discid.h	Mon Oct 17 13:05:09 2005
+++ lib/discid.h	Mon Oct 17 12:00:02 2005
@@ -39,6 +39,7 @@
 
 #ifdef __FreeBSD__
 #include <sys/cdio.h>
+#include <sys/fcntl.h>
 #include "freebsd.h"
 #endif // __FreeBSD__
 
