--- FreeDB.pm.orig	Tue Jan  3 08:58:51 2006
+++ FreeDB.pm	Sun Jun 18 19:50:43 2006
@@ -1,6 +1,6 @@
 package Net::FreeDB;
 
-use 5.006;
+use 5.00503;
 use strict;
 use warnings;
 use IO::Socket;
@@ -13,7 +13,8 @@
 require DynaLoader;
 use AutoLoader;
 
-our @ISA = qw(Exporter DynaLoader Net::Cmd IO::Socket::INET);
+use vars qw/@ISA %EXPORT_TAGS @EXPORT_OK @EXPORT $VERSION $ERROR/;
+@ISA = qw(Exporter DynaLoader Net::Cmd IO::Socket::INET);
 
 # Items to export into callers namespace by default. Note: do not export
 # names by default without a very good reason. Use EXPORT_OK instead.
@@ -22,22 +23,21 @@
 # This allows declaration	use Net::FreeDB ':all';
 # If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
 # will save memory.
-our %EXPORT_TAGS = ( 'all' => [ qw() ] );
+%EXPORT_TAGS = ( 'all' => [ qw() ] );
 
-our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
+@EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
 
-our @EXPORT = qw();
+@EXPORT = qw();
 
-our $VERSION = '0.07';
+$VERSION = '0.07';
 
-our $ERROR;
 sub AUTOLOAD {
     # This AUTOLOAD is used to 'autoload' constants from the constant()
     # XS function.  If a constant is not found then control is passed
     # to the AUTOLOAD in AutoLoader.
 
     my $constname;
-    our $AUTOLOAD;
+    use vars qw/$AUTOLOAD/;
     ($constname = $AUTOLOAD) =~ s/.*:://;
     croak "& not defined" if $constname eq 'constant';
     my $val = constant($constname, @_ ? $_[0] : 0);
--- FreeDB.xs.linux.orig	Mon Oct 17 07:11:29 2005
+++ FreeDB.xs.linux	Sun Jun 18 19:53:17 2006
@@ -38,7 +38,8 @@
 	RETVAL
 
 char*
-discid(char* dev)
+discid(dev)
+	char *dev;
     INIT:
 		int i;
 		char id[30];
@@ -56,7 +57,8 @@
 		RETVAL
 
 SV*
-discinfo(char* dev)
+discinfo(dev)
+	char *dev;
     INIT:
 	int i;
         char id[30];
--- lib/discid.h.orig	Wed Sep 28 08:45:55 2005
+++ lib/discid.h	Sun Jun 18 19:54:13 2006
@@ -39,6 +39,7 @@
 
 #ifdef __FreeBSD__
 #include <sys/cdio.h>
+#include <sys/fcntl.h>
 #include "freebsd.h"
 #endif // __FreeBSD__
 
