--- CMakeLists.txt	2023-06-20 14:32:15.000000000 +0100
+++ CMakeLists.txt	2023-08-25 16:15:38.899041000 +0100
@@ -37,7 +37,6 @@
     KIO
     ItemViews
     Sonnet
-    DocTools
     GlobalAccel
     WindowSystem
 )
@@ -50,7 +49,6 @@
 add_definitions(-DQT_NO_CAST_FROM_ASCII -DQT_NO_CAST_FROM_BYTEARRAY)
 
 add_subdirectory( pixmaps ) 
-add_subdirectory( doc )
 add_subdirectory(kconf_update)
 
 set(kmenueditcommon_STAT_SRCS basictab.cpp
@@ -120,4 +118,3 @@
         )
 
 ki18n_install(po)
-kdoctools_install(po)
