--- celestia-1.6.1.orig/src/celestia/celx.cpp	2014-02-09 23:23:10.308810741 +0000
+++ celestia-1.6.1/src/celestia/celx.cpp	2014-02-09 23:24:57.627017784 +0000
@@ -272,9 +272,13 @@ 
                            const char* name,
                            lua_CFunction func)
 {
+#if LUA_VER >= 0x050200
+    luaL_requiref(l, name, func, 1);
+#else
     lua_pushcfunction(l, func);
     lua_pushstring(l, name);
     lua_call(l, 1, 0);
+#endif
 }
 #endif
 
