Index: server/listen.c
===================================================================
RCS file: /home/cvspublic/httpd-2.0/server/listen.c,v
retrieving revision 1.95
diff -u -r1.95 listen.c
--- server/listen.c     1 Jan 2004 13:26:23 -0000       1.95
+++ server/listen.c     3 Feb 2004 17:34:08 -0000
@@ -199,7 +199,13 @@

 #if APR_HAS_SO_ACCEPTFILTER
 #ifndef ACCEPT_FILTER_NAME
+#define ACCEPT_FILTER_NAME "httpready"
+#ifdef __FreeBSD_version
+#if __FreeBSD_version < 411000 /* httpready broken before 4.1.1 */
+#undef ACCEPT_FILTER_NAME
 #define ACCEPT_FILTER_NAME "dataready"
+#endif
+#endif
 #endif
     apr_socket_accept_filter(s, ACCEPT_FILTER_NAME, "");
 #endif
