--- a/source/CMakeLists.txt
+++ b/source/CMakeLists.txt
@@ -780,7 +780,7 @@ if(POWER)
     # IBM Power8
     option(ENABLE_ALTIVEC "Enable ALTIVEC profiling instrumentation" ON)
     if(ENABLE_ALTIVEC)
-        add_definitions(-DHAVE_ALTIVEC=1 -maltivec -mabi=altivec)
+        add_definitions(-DHAVE_ALTIVEC=1 -maltivec)
         add_definitions(-flax-vector-conversions -fpermissive)
     else()
         add_definitions(-DHAVE_ALTIVEC=0)
