--- vmmon-only/include/vm_types.h.patched	Tue Apr  8 15:10:25 2003
+++ vmmon-only/include/vm_types.h	Tue Apr  8 15:12:07 2003
@@ -144,10 +144,6 @@
 #define offsetof(type, member) ( (int) & ((type*)0) -> member )
 #endif
 
-
-#define MIN(_a, _b)   (((_a) < (_b)) ? (_a) : (_b))
-#define MAX(_a, _b)   (((_a) > (_b)) ? (_a) : (_b))
-
 #define ROUNDUP(x,y)		(((x) + (y) - 1) / (y) * (y))
 #define ROUNDDOWN(x,y)		((x) / (y) * (y))
 #define ROUNDUPBITS(x, bits)	(((uint32) (x) + MASK(bits)) & ~MASK(bits))
