--- lmtp2nntp_l2.c.orig	Wed Oct 19 19:11:54 2005
+++ lmtp2nntp_l2.c	Wed Oct 19 19:12:26 2005
@@ -60,7 +60,7 @@
 }
 
 /* configure channel */
-static l2_result_t hook_configure(l2_context_t *ctx, l2_channel_t *ch, const char *fmt, va_list ap)
+static l2_result_t hook_configure(l2_context_t *ctx, l2_channel_t *ch, const char *fmt, va_list *ap)
 {
     l2_ch_var_t *cfg;
 
@@ -69,7 +69,7 @@
         return L2_ERR_ARG;
 
     /* special parameter parsing */
-    if ((cfg->var = (var_t *)va_arg(ap, void *)) == NULL)
+    if ((cfg->var = (var_t *)va_arg(*ap, void *)) == NULL)
         return L2_ERR_ARG;
 
     return L2_OK;
