--- dialogline.cc
+++ dialogline.cc
@@ -16,6 +16,7 @@
 
 #include <config.h>
 
+#include <cstdlib>
 #include <stdarg.h>
 
 #include "dialogline.h"
--- editbox2.cc
+++ editbox2.cc
@@ -14,6 +14,7 @@
 // along with this program; if not, write to the Free Software
 // Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111, USA.
 
+#include <cstdlib>
 #include <config.h>
 
 #include "editbox.h"
--- editbox.h
+++ editbox.h
@@ -768,7 +768,7 @@
 			    int para_num
 			);
 
-    void EditBox::redraw_unwrapped_paragraph(
+    void redraw_unwrapped_paragraph(
 			    Paragraph &p,
 			    int window_start_line,
 			    bool only_cursor,
@@ -780,7 +780,7 @@
 			    bool eop_is_selected
 			);
 
-    void EditBox::redraw_wrapped_paragraph(
+    void redraw_wrapped_paragraph(
 			    Paragraph &p,
 			    int window_start_line,
 			    bool only_cursor,
--- io.cc
+++ io.cc
@@ -16,6 +16,7 @@
 
 #include <config.h>
 
+#include <cstdlib>
 #include <stdarg.h>
 #include <fcntl.h>  // file primitives
 #include <unistd.h>
--- transtbl.cc
+++ transtbl.cc
@@ -16,6 +16,7 @@
 
 #include <config.h>
 
+#include <cstdlib>
 #include <stdio.h>
 #include <errno.h>
