--- configure.orig	2012-01-19 17:19:29.000000000 -0200
+++ configure	2013-10-09 15:07:40.051081546 -0300
@@ -9336,7 +9336,7 @@
     if test x${ffmpeg_top_incl} = x; then
     { $as_echo "$as_me:${as_lineno-$LINENO}: checking location of avcodec.h" >&5
 $as_echo_n "checking location of avcodec.h... " >&6; }
-        if test x"$PKG_CONFIG" != x; then       if $PKG_CONFIG --exists libavcodec; then                                 ffmpeg_pkg=`$PKG_CONFIG --cflags-only-I libavcodec`
+        if test x"$PKG_CONFIG" != x; then       if $PKG_CONFIG --exists libavcodec0; then                                 ffmpeg_pkg=`$PKG_CONFIG --cflags-only-I libavcodec0`
         ffmpeg_top_incl=`echo ${ffmpeg_pkg} | sed  -e 's:-I::'`
                         for i in "" ffmpeg libavcodec ffmpeg/libavcodec; do
           if test -f ${ffmpeg_top_incl}/${i}/avcodec.h; then
@@ -9568,16 +9568,16 @@
 else
 
     if test x"${with_ffmpeg_lib}" != x ; then
-      if test -f ${with_ffmpeg_lib}/libavcodec.a -o -f ${with_ffmpeg_lib}/libavcodec.${shlibext}; then
+      if test -f ${with_ffmpeg_lib}/libavcodec0.a -o -f ${with_ffmpeg_lib}/libavcodec0.${shlibext}; then
 	      ac_cv_path_ffmpeg_lib="-L`(cd ${with_ffmpeg_lib}; pwd)`"
-          libavcodec="-lavcodec"
+          libavcodec="-lavcodec0"
           LIBS="${ac_cv_path_ffmpeg_lib} $LIBS"           top_lib_dir=${with_ffmpeg_lib}
       else
 	      as_fn_error "${with_ffmpeg_lib} directory doesn't contain libavcodec libraries." "$LINENO" 5
       fi
-      if test -f ${with_ffmpeg_lib}/libavformat.a -o -f ${with_ffmpeg_lib}/libavformat.${shlibext}; then
+      if test -f ${with_ffmpeg_lib}/libavformat0.a -o -f ${with_ffmpeg_lib}/libavformat0.${shlibext}; then
 	      ac_cv_path_ffmpeg_lib="-L`(cd ${with_ffmpeg_lib}; pwd)`"
-          libavformat="-lavformat"
+          libavformat="-lavformat0"
           top_lib_dir=${with_ffmpeg_lib}
       else
 	      as_fn_error "${with_ffmpeg_lib} directory doesn't contain libavformat libraries." "$LINENO" 5
@@ -9869,13 +9869,13 @@
                     { $as_echo "$as_me:${as_lineno-$LINENO}: checking for libavutil library" >&5
 $as_echo_n "checking for libavutil library... " >&6; }
     if test x"$PKG_CONFIG" != x -a x${cross_compiling} = xno; then
-      $PKG_CONFIG --exists libavutil && libavutil=`$PKG_CONFIG --libs-only-l libavutil`
+      $PKG_CONFIG --exists libavutil0 && libavutil=`$PKG_CONFIG --libs-only-l libavutil0`
     else
       libavutil=""
     fi
     if test x"${libavutil}" = x; then
-      if test -f ${top_lib_dir}/libavutil.a -o -f ${top_lib_dir}/libavutil.${shlibext}; then
-        ac_cv_path_ffmpeg_lib="${ac_cv_path_ffmpeg_lib} -lavutil"
+      if test -f ${top_lib_dir}/libavutil0.a -o -f ${top_lib_dir}/libavutil0.${shlibext}; then
+        ac_cv_path_ffmpeg_lib="${ac_cv_path_ffmpeg_lib} -lavutil0"
         { $as_echo "$as_me:${as_lineno-$LINENO}: result: ${top_lib_dir}/libavutil" >&5
 $as_echo "${top_lib_dir}/libavutil" >&6; }
       else
@@ -9888,7 +9888,7 @@
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lavutil  $LIBS"
+LIBS="-lavutil0  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -9919,7 +9919,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_avutil_av_log" >&5
 $as_echo "$ac_cv_lib_avutil_av_log" >&6; }
 if test "x$ac_cv_lib_avutil_av_log" = x""yes; then :
-  ac_cv_path_ffmpeg_lib="${ac_cv_path_ffmpeg_lib} -lavutil"
+  ac_cv_path_ffmpeg_lib="${ac_cv_path_ffmpeg_lib} -lavutil0"
 fi
 
         fi
@@ -10126,13 +10126,13 @@
                     { $as_echo "$as_me:${as_lineno-$LINENO}: checking for libswscale library" >&5
 $as_echo_n "checking for libswscale library... " >&6; }
     if test x"$PKG_CONFIG" != x -a x${cross_compiling} = xno; then
-      $PKG_CONFIG --exists libswscale  && libsws=`$PKG_CONFIG --libs-only-l libswscale`
+      $PKG_CONFIG --exists libswscale0  && libsws=`$PKG_CONFIG --libs-only-l libswscale0`
     else
       libsws=""
     fi
     if test x"${libsws}" = x; then
-      if test -f ${top_lib_dir}/libswscale.a -o -f ${top_lib_dir}/libswscale.${shlibext}; then
-        ac_cv_path_ffmpeg_lib="${ac_cv_path_ffmpeg_lib} -lswscale"
+      if test -f ${top_lib_dir}/libswscale0.a -o -f ${top_lib_dir}/libswscale0.${shlibext}; then
+        ac_cv_path_ffmpeg_lib="${ac_cv_path_ffmpeg_lib} -lswscale0"
         { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
 $as_echo "yes" >&6; }
       else
@@ -10145,7 +10145,7 @@
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lswscale  $LIBS"
+LIBS="-lswscale0  $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -10176,7 +10176,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_swscale_sws_scale" >&5
 $as_echo "$ac_cv_lib_swscale_sws_scale" >&6; }
 if test "x$ac_cv_lib_swscale_sws_scale" = x""yes; then :
-  ac_cv_path_ffmpeg_lib="${ac_cv_path_ffmpeg_lib} -lswscale"
+  ac_cv_path_ffmpeg_lib="${ac_cv_path_ffmpeg_lib} -lswscale0"
 fi
 
         fi
