--- ../mkspecs/freebsd-g++/qmake.conf.orig	2012-09-10 18:36:35.000000000 -0700
+++ ../mkspecs/freebsd-g++/qmake.conf	2012-09-16 16:13:00.894513577 -0700
@@ -13,8 +13,8 @@
 QMAKE_CXXFLAGS_THREAD	= $$QMAKE_CFLAGS_THREAD
 
 # Addon software goes into /usr/local on the BSDs, by default we will look there
-QMAKE_INCDIR		= /usr/local/include
-QMAKE_LIBDIR		= /usr/local/lib
+QMAKE_INCDIR		= $$[QT_INSTALL_HEADERS] /usr/local/include
+QMAKE_LIBDIR		= $$[QT_INSTALL_LIBS] /usr/local/lib
 QMAKE_INCDIR_X11	= /usr/X11R6/include
 QMAKE_LIBDIR_X11	= /usr/X11R6/lib
 QMAKE_INCDIR_QT		= $$[QT_INSTALL_HEADERS]
--- ../mkspecs/freebsd-g++46/qmake.conf.orig	2012-09-10 18:36:35.000000000 -0700
+++ ../mkspecs/freebsd-g++46/qmake.conf	2012-09-16 16:13:00.894513577 -0700
@@ -13,8 +13,8 @@
 QMAKE_CXXFLAGS_THREAD	= $$QMAKE_CFLAGS_THREAD
 
 # Addon software goes into /usr/local on the BSDs, by default we will look there
-QMAKE_INCDIR		= /usr/local/include
-QMAKE_LIBDIR		= /usr/local/lib
+QMAKE_INCDIR		= $$[QT_INSTALL_HEADERS] /usr/local/include
+QMAKE_LIBDIR		= $$[QT_INSTALL_LIBS] /usr/local/lib
 QMAKE_INCDIR_X11	= /usr/X11R6/include
 QMAKE_LIBDIR_X11	= /usr/X11R6/lib
 QMAKE_INCDIR_QT		= $$[QT_INSTALL_HEADERS]
