--- app/config/core.php	Mon Feb 19 13:45:27 2007
+++ app/config/core.php	Mon Feb 19 13:39:40 2007
@@ -37,7 +37,7 @@
  * release/app/.htaccess
  * release/app/webroot/.htaccess
  */
-//	define ('BASE_URL', env('SCRIPT_NAME'));
+	define ('BASE_URL', env('SCRIPT_NAME'));
 /**
  * Set debug level here:
  * - 0: production
@@ -143,4 +143,4 @@
  */
 	define('ACL_CLASSNAME', 'DB_ACL');
 	define('ACL_FILENAME', 'dbacl' . DS . 'db_acl');
-?>
\ No newline at end of file
+?>
