--- src/classes/Makefile.in.orig	2012-09-03 12:10:18.000000000 -0400
+++ src/classes/Makefile.in	2012-10-05 15:41:54.000000000 -0400
@@ -531,7 +531,7 @@
 
 @ENABLE_ZLIB_TRUE@vm.zip: $(VM_JAVA_FILES)
 @ENABLE_ZLIB_TRUE@	$(mkdir_p) classes
-@ENABLE_ZLIB_TRUE@	$(JAVAC) -bootclasspath $(BOOTCLASSPATH) -source 1.5 -target 1.5 -d classes $(VM_JAVA_FILES)
+@ENABLE_ZLIB_TRUE@	$(JAVAC) $(JAVACFLAGS) --bootclasspath=$(BOOTCLASSPATH) -fsource=1.5 -ftarget=1.5 -d classes $(VM_JAVA_FILES)
 @ENABLE_ZLIB_TRUE@	@if test "$(JAR)" = "zip" -o "$(JAR)" = "zip.exe"; then \
 @ENABLE_ZLIB_TRUE@	    cd classes && $(JAR) -r -D $(VM_ZIP) .; \
 @ENABLE_ZLIB_TRUE@	else \
@@ -539,7 +539,7 @@
 @ENABLE_ZLIB_TRUE@	fi
 @ENABLE_ZLIB_FALSE@nozip: $(VM_JAVA_FILES)
 @ENABLE_ZLIB_FALSE@	$(mkdir_p) classes
-@ENABLE_ZLIB_FALSE@	$(JAVAC) -bootclasspath $(BOOTCLASSPATH) -source 1.5 -target 1.5 -d classes $(VM_JAVA_FILES)
+@ENABLE_ZLIB_FALSE@	$(JAVAC) $(JAVACFLAGS) --bootclasspath=$(BOOTCLASSPATH) -fsource=1.5 -ftarget=1.5 -d classes $(VM_JAVA_FILES)
 
 @ENABLE_ZLIB_FALSE@install-data-local: nozip
 @ENABLE_ZLIB_FALSE@	$(mkinstalldirs) $(DESTDIR)$(pkgdatadir)
