--- ispell-da-1.4.22/Makefile.orig	Tue Jan  2 10:47:45 2001
+++ ispell-da-1.4.22/Makefile	Wed Jan 10 09:10:51 2001
@@ -7,7 +7,7 @@
 	install -o root -g root -m 0644 dansk.aff $(installdir)
 
 dansk.hash: dansk.aff words-da.ispell
-	buildhash words-da.ispell dansk.aff dansk.hash
+	${BUILDHASH} words-da.ispell dansk.aff dansk.hash
 
 clean:
 	rm -f words-da.ispell.stat dansk.hash words-da.ispell.cnt *~
