--- ispell-da-1.4.49/Makefile.orig	Thu Apr 21 11:29:52 2005
+++ ispell-da-1.4.49/Makefile	Sun May  1 21:35:42 2005
@@ -9,13 +9,13 @@
 lang=da
 version=1.4.49
 
-installdir=`ispell -vv | grep LIBDIR | cut -d'"' -f2`
+installdir=`${ISPELL} -vv | grep LIBDIR | cut -d'"' -f2`
 
 all: maskbits $(sprog).hash
 
 maskbits:
-	@test "`ispell -vv | grep MASKBITS`" != ""              || ( echo Ispell should be compiled with MASKBITS set to at least 64. ; exit -2 )
-	@test `ispell -vv | grep MASKBITS | cut -d= -f2` -ge 64 || ( echo Ispell should be compiled with MASKBITS set to at least 64. ; exit -3 )
+	@test "`${ISPELL} -vv | grep MASKBITS`" != ""              || ( echo Ispell should be compiled with MASKBITS set to at least 64. ; exit -2 )
+	@test `${ISPELL} -vv | grep MASKBITS | cut -d= -f2` -ge 64 || ( echo Ispell should be compiled with MASKBITS set to at least 64. ; exit -3 )
 
 install: maskbits $(sprog).hash $(sprog).aff
 	install -o root -g root -m 0644 $(sprog).hash $(installdir)
@@ -26,7 +26,7 @@
 	ln -fs $(installdir)/$(sprog).aff  $(installdir)/$(sprog_en_lower_case).aff
 
 $(sprog).hash: maskbits $(sprog).aff words-$(language_code).ispell
-	buildhash words-$(language_code).ispell $(sprog).aff $(sprog).hash
+	${BUILDHASH} words-$(language_code).ispell $(sprog).aff $(sprog).hash
 
 clean:
 	rm -f words-$(language_code).ispell.stat $(sprog).hash words-$(language_code).ispell.cnt *~
