--- Makefile.orig	2007-12-11 15:52:44.000000000 +0100
+++ Makefile	2008-04-01 15:44:16.000000000 +0200
@@ -62,16 +62,16 @@
 #ZIP_OPTS = -j -r -9 -X
 ZIP = 7z
 ZIP_OPTS = a -tzip -mx=9
-BUILDHASH = buildhash
+BUILDHASH = ../ispell-3.3.02/buildhash
 
-ISPELL = ispell
+ISPELL = ../ispell-3.3.02/ispell
 HUNSPELL_BIN = hunspell
 # Hunspell starting with 1.1.9 (?) uses automatic charset transcoding,
 # we need to make sure we use latin1:
 HUNSPELL = $(HUNSPELL_BIN) `$(HUNSPELL_BIN) -h 2>&1 | grep -q "hunspell \\-i" && echo "-i latin1"`
 
 SQ = sq
-UNSQ = unsq
+UNSQ = ../ispell-3.3.02/unsq
 #   if for some reason you don't have Ispell's unsq use my Perl sq/unsq
 #  and put it into a directory known by your PATH variable!
 #UNSQ = unsq.pl
@@ -113,7 +113,7 @@
 	$$(BUILDHASH) all.words.$(1).tmp ./ispell/$(1).aff $$@
 
 all.words.$(1).tmp: all.words.$(1).raw ispell/$(1).aff
-	munchlist -l ./ispell/$(1).aff all.words.$(1).raw > $$@
+	../ispell-3.3.02/munchlist -l ./ispell/$(1).aff all.words.$(1).raw > $$@
 
 all.words.$(1).raw: $$(WORDS_$(1)) $$(WORDS_de_all) dicts/abkuerzpunkt.txt
 	sed "s/qq//g;s/\.//;s/\-.*//" dicts/abkuerzpunkt.txt $$(WORDS_$(1)) $$(WORDS_de_all) \
