--- Makefile.orig	Tue Sep  5 23:50:30 2000
+++ Makefile	Sat Mar 31 14:57:58 2001
@@ -8,9 +8,9 @@
 SHELL = /bin/sh
 MAKE = make
 
-CONFIG		=	../../config.sh
-PATHADDER	=	../..
-BUILDHASH	=	../../buildhash
+CONFIG		=	../ispell-3.1/config.sh
+PATHADDER	=	../ispell-3.1
+BUILDHASH	?=	../ispell-3.1/buildhash
 
 # The following variables make it easy to adapt this Makefile to
 # numerous languages.
@@ -269,7 +269,7 @@
 	echo -e 'suffixes\nflag *z:\nY Y Y Y Y   >   YYYYYY' >> $@
 
 norsk.words: norsk.words.sq
-	unsq < norsk.words.sq > norsk.words
+	$(UNSQ) < norsk.words.sq > norsk.words
 
 # The following ugly code munches a part of the base file, keeping the
 # indications of the frequency of the words.  It also removes some
@@ -292,19 +292,23 @@
 # Nasty bug, and very silent.
 	PATH=$(PATHADDER):$$PATH; \
 	export PATH; \
+	TMPDIR=/tmp; \
+	export TMPDIR; \
 	${CATNOHEADER} \
 	  | grep -e '$(subst munched.,,$@)$$' \
 	  | sed -e 's/ .*//' -e 's/-/î/g' -e 's/ \*//' \
 	  | grep '^[${UCH}]' \
-	  | munchlist -v -l ${AFFIXES}.munch \
+	  | $(MUNCHLIST) -v -l ${AFFIXES}.munch \
 	  > munch1.tmp
 	PATH=$(PATHADDER):$$PATH; \
 	export PATH; \
+	TMPDIR=/tmp; \
+	export TMPDIR; \
 	${CATNOHEADER} \
 	  | grep -e '$(subst munched.,,$@)$$' \
 	  | sed -e 's/ .*//' -e 's/-/î/g' -e 's/ \*//' \
 	  | grep -v '^[${UCH}]' \
-	  | munchlist -v -l ${AFFIXES}.munch \
+	  | $(MUNCHLIST) -v -l ${AFFIXES}.munch \
 	  >> munch1.tmp
 	cat munch1.tmp \
 	  | sed -e 's/\(zyzyzy\|ZYZYZY\)/\1\/\` /' \
@@ -459,11 +463,13 @@
 nynorsk.mch: norsk.words ny${AFFIXES}.munch
 	PATH=$(PATHADDER):$$PATH; \
 	export PATH; \
+	TMPDIR=/tmp; \
+	export TMPDIR; \
 	${CATNOHEADER} \
 	  | grep '\*' \
 	  | sed -e 's/ .*//' \
 	  | tr -d '-' \
-	  | munchlist -v -l ny${AFFIXES}.munch \
+	  | $(MUNCHLIST) -v -l ny${AFFIXES}.munch \
 	  | sed -e N -e 's/^\(\([-${CH}]\)*\)er\/\(.*F.*\)\n\1rar\/M$$/\1er\/\3D/' \
 		-e '$$ p' -e '$$ d' -e P -e D \
 	  > $@ 
