--- Makefile	Fri Feb  4 22:09:46 2005
+++ Makefile.noglk	Fri Feb  4 22:09:33 2005
@@ -68,8 +68,9 @@
 # If your make can't handle sinclude or -include, and you're not building
 # for Glk, comment out the include line.
 GLKPATH	= $(GLKDIR)/$(GLK)
-sinclude $(GLKPATH)/Make.$(GLK)
-GLKLIBS	= -L$(GLKPATH) $(GLKLIB) $(LINKLIBS)
+#sinclude $(GLKPATH)/Make.$(GLK)
+#GLKLIBS	= -L$(GLKPATH) $(GLKLIB) $(LINKLIBS)
+GLKLIBS	=
 
 # Always use the Bourne shell.
 SHELL	= /bin/sh
