--- source/Makefile	Sun Jan 21 23:18:36 2007
+++ source/Makefile.noglk	Sun Jan 21 23:21:08 2007
@@ -73,8 +73,9 @@
 # If your make can't handle sinclude or -include, and you're not building
 # for Glk, comment out the include line.
 GLKPATH	= $(GLKDIR)/$(GLK)
-sinclude $(GLKPATH)/Make.$(GLK)
-GLKLIBS	= -L$(GLKPATH) $(GLKLIB) $(LINKLIBS)
+#sinclude $(GLKPATH)/Make.$(GLK)
+#GLKLIBS	= -L$(GLKPATH) $(GLKLIB) $(LINKLIBS)
+GLKLIBS	=
 
 # Always use the Bourne shell.
 SHELL	= /bin/sh
