--- Makefile.orig	Fri Apr 30 08:05:12 2004
+++ Makefile	Fri Feb  4 22:18:07 2005
@@ -68,7 +68,7 @@
 # If your make can't handle sinclude or -include, and you're not building
 # for Glk, comment out the include line.
 GLKPATH	= $(GLKDIR)/$(GLK)
-sinclude $(GLKPATH)/Make.$(GLK)
+.include "$(X11BASE)/share/xglk/Make.$(GLK)"
 GLKLIBS	= -L$(GLKPATH) $(GLKLIB) $(LINKLIBS)
 
 # Always use the Bourne shell.
@@ -121,7 +121,7 @@
 		$(LDLIBS) $(GLKLIBS)
 
 os_glk.o: os_glk.c $(HEADERS)
-	$(CC) $(GLKARCH) $(CFLAGS) -c -o $@ $<
+	$(CC) $(GLKARCH) $(CFLAGS) -DLINUX_GRAPHICS -I${X11BASE}/include/xglk -c -o $@ $<
 
 # Option to build an IFP plugin version of SCARE.
 scare_plugin.c: scare.hdr
