--- tools/assistant/config.cpp
+++ tools/assistant/config.cpp
@@ -73,12 +73,12 @@ Config *Config::loadConfig(const QString
 
     QFile file(profileFileName);
     if (!file.exists()) {
-        qWarning( (QLatin1String("File does not exist: ") + profileFileName).toAscii().constData() );
+        qWarning( "File does not exist: %s", qPrintable(profileFileName) );
         return 0;
     }
     DocuParser *parser = DocuParser::createParser( profileFileName );
     if (!parser) {
-        qWarning( (QLatin1String("Failed to create parser for file: ") + profileFileName).toAscii().constData() );
+        qWarning( "Failed to create parser for file: %s", qPrintable(profileFileName) );
         return 0;
     }
     if (parser->parserVersion() < DocuParser::Qt320) {
@@ -89,7 +89,7 @@ Config *Config::loadConfig(const QString
     parser->parse(&file);
     config->profil = profileParser->profile();
     if (!config->profil) {
-        qWarning( (QLatin1String("Config::loadConfig(), no profile in: ") + profileFileName).toAscii().constData() );
+        qWarning( "Config::loadConfig(), no profile in: %s", qPrintable(profileFileName) );
         return 0;
     }
     config->profil->setProfileType(Profile::UserProfile);
--- tools/assistant/index.cpp
+++ tools/assistant/index.cpp
@@ -180,7 +180,7 @@ void Index::parseDocument( const QString
 {
     QFile file( filename );
     if ( !file.open(QFile::ReadOnly) ) {
-        qWarning( (QLatin1String("can not open file ") + filename).toAscii().constData() );
+        qWarning( "can not open file %s", qPrintable(filename) );
         return;
     }
 
@@ -352,7 +352,7 @@ QString Index::getDocumentTitle( const Q
 
     QFile file( fileName );
     if ( !file.open( QFile::ReadOnly ) ) {
-        qWarning( (QLatin1String("cannot open file ") + fileName).toAscii().constData() );
+        qWarning( "cannot open file %s", qPrintable(fileName) );
         return fileName;
     }
     QTextStream s( &file );
@@ -474,7 +474,7 @@ bool Index::searchForPattern( const QStr
     QString fName = url.toLocalFile();
     QFile file( fName );
     if ( !file.open( QFile::ReadOnly ) ) {
-        qWarning( (QLatin1String("cannot open file ") + fName).toAscii().constData() );
+        qWarning( "cannot open file %s", qPrintable(fName) );
         return false;
     }
 