diff -ruN ../sylpheed-2.0.0beta3.orig/src/compose.c ./src/compose.c
--- ../sylpheed-2.0.0beta3.orig/src/compose.c	Tue Jun 21 08:57:48 2005
+++ ./src/compose.c	Sun Jun 26 14:12:37 2005
@@ -4331,7 +4331,7 @@
 	gtk_container_add(GTK_CONTAINER(container), toolbar);
 	gtk_widget_set_size_request(toolbar, 1, -1);
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_MAIL_SEND);
+	icon_wid = gtk_image_new_from_icon_name("stock_mail-send", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	send_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					   _("Send"),
 					   _("Send message"),
@@ -4340,7 +4340,7 @@
 					   G_CALLBACK(toolbar_send_cb),
 					   compose);
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_MAIL_SEND_QUEUE);
+	icon_wid = stock_pixbuf_widget(container, STOCK_PIXMAP_MAIL_SEND_QUEUE);
 	sendl_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					   _("Send later"),
 					   _("Put into queue folder and send later"),
@@ -4349,7 +4349,7 @@
 					   G_CALLBACK(toolbar_send_later_cb),
 					   compose);
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_MAIL);
+	icon_wid = gtk_image_new_from_icon_name("stock_mail", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	draft_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					    _("Draft"),
 					    _("Save to draft folder"),
@@ -4360,7 +4360,7 @@
 
 	gtk_toolbar_append_space(GTK_TOOLBAR(toolbar));
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_PASTE);
+	icon_wid = gtk_image_new_from_icon_name("stock_insert-file", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	insert_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					     _("Insert"),
 					     _("Insert file"),
@@ -4369,7 +4369,7 @@
 					     G_CALLBACK(toolbar_insert_cb),
 					     compose);
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_MAIL_ATTACH);
+	icon_wid = gtk_image_new_from_icon_name("stock_attach", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	attach_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					     _("Attach"),
 					     _("Attach file"),
@@ -4380,7 +4380,7 @@
 
 	gtk_toolbar_append_space(GTK_TOOLBAR(toolbar));
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_MAIL);
+	icon_wid = gtk_image_new_from_icon_name("stock_mail", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	sig_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					  _("Signature"),
 					  _("Insert signature"),
@@ -4390,7 +4390,7 @@
 
 	gtk_toolbar_append_space(GTK_TOOLBAR(toolbar));
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_MAIL_COMPOSE);
+	icon_wid = gtk_image_new_from_icon_name("stock_mail-compose", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	exteditor_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 						_("Editor"),
 						_("Edit with external editor"),
@@ -4410,7 +4410,7 @@
 
 	gtk_toolbar_append_space(GTK_TOOLBAR(toolbar));
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_ADDRESS_BOOK);
+	icon_wid = gtk_image_new_from_icon_name("stock_addressbook", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	addrbook_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					       _("Address"),
 					       _("Address book"),
diff -ruN ../sylpheed-2.0.0beta3.orig/src/folderview.c ./src/folderview.c
--- ../sylpheed-2.0.0beta3.orig/src/folderview.c	Thu Jun 23 10:58:40 2005
+++ ./src/folderview.c	Sat Jun 25 20:17:04 2005
@@ -490,8 +490,8 @@
 {
 	GtkWidget *treeview = folderview->treeview;
 
-	stock_pixbuf_gdk(treeview, STOCK_PIXMAP_INBOX, &inbox_pixbuf);
-	stock_pixbuf_gdk(treeview, STOCK_PIXMAP_OUTBOX, &outbox_pixbuf);
+	inbox_pixbuf = gtk_icon_theme_load_icon(gtk_icon_theme_get_default(), "stock_inbox", 16, 0, NULL);
+	outbox_pixbuf = gtk_icon_theme_load_icon(gtk_icon_theme_get_default(), "stock_outbox", 16, 0, NULL);
 	stock_pixbuf_gdk(treeview, STOCK_PIXMAP_DIR_CLOSE, &folder_pixbuf);
 	stock_pixbuf_gdk(treeview, STOCK_PIXMAP_DIR_OPEN, &folderopen_pixbuf);
 	stock_pixbuf_gdk(treeview, STOCK_PIXMAP_DIR_NOSELECT,
diff -ruN ../sylpheed-2.0.0beta3.orig/src/mainwindow.c ./src/mainwindow.c
--- ../sylpheed-2.0.0beta3.orig/src/mainwindow.c	Mon Jun 20 10:13:33 2005
+++ ./src/mainwindow.c	Sun Jun 26 14:12:20 2005
@@ -2188,11 +2188,11 @@
 				    GTK_ORIENTATION_HORIZONTAL);
 	gtk_toolbar_set_style(GTK_TOOLBAR(toolbar), GTK_TOOLBAR_BOTH);
 	gtk_toolbar_set_icon_size(GTK_TOOLBAR(toolbar),
-				  GTK_ICON_SIZE_SMALL_TOOLBAR);
+				  GTK_ICON_SIZE_LARGE_TOOLBAR);
 	gtk_container_add(GTK_CONTAINER(container), toolbar);
 	gtk_widget_set_size_request(toolbar, 1, -1);
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_MAIL_RECEIVE);
+	icon_wid = gtk_image_new_from_icon_name("stock_mail-receive", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	get_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					  _("Get"),
 					  _("Incorporate new mail"),
@@ -2200,7 +2200,7 @@
 					  icon_wid,
 					  G_CALLBACK(toolbar_inc_cb),
 					  mainwin);
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_MAIL_RECEIVE_ALL);
+	icon_wid = stock_pixbuf_widget(container, STOCK_PIXMAP_MAIL_RECEIVE_ALL);
 	getall_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					     _("Get all"),
 					     _("Incorporate new mail of all accounts"),
@@ -2211,7 +2211,7 @@
 
 	gtk_toolbar_append_space(GTK_TOOLBAR(toolbar));
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_MAIL_SEND);
+	icon_wid = gtk_image_new_from_icon_name("stock_mail-send", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	send_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					   _("Send"),
 					   _("Send queued message(s)"),
@@ -2222,7 +2222,7 @@
 
 	gtk_toolbar_append_space(GTK_TOOLBAR(toolbar));
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_MAIL_COMPOSE);
+	icon_wid = gtk_image_new_from_icon_name("stock_mail-compose", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	compose_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					      _("Compose"),
 					      _("Compose new message"),
@@ -2231,7 +2231,7 @@
 					      G_CALLBACK(toolbar_compose_cb),
 					      mainwin);
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_MAIL_REPLY);
+	icon_wid = gtk_image_new_from_icon_name("stock_mail-reply", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	reply_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					    _("Reply"),
 					    _("Reply to the message"),
@@ -2249,8 +2249,7 @@
 				  GTK_WIDGET_PTR(reply_combo),
 				  _("Reply to the message"), "Reply");
 
-	icon_wid = stock_pixmap_widget
-		(container, STOCK_PIXMAP_MAIL_REPLY_TO_ALL);
+	icon_wid = gtk_image_new_from_icon_name("stock_mail-reply-to-all", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	replyall_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					       _("Reply all"),
 					       _("Reply to all"),
@@ -2259,7 +2258,7 @@
 					       G_CALLBACK(toolbar_reply_to_all_cb),
 					       mainwin);
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_MAIL_FORWARD);
+	icon_wid = gtk_image_new_from_icon_name("stock_mail-forward", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	fwd_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					  _("Forward"),
 					  _("Forward the message"),
@@ -2280,7 +2279,7 @@
 	gtk_toolbar_append_space(GTK_TOOLBAR(toolbar));
 
 	icon_wid = gtk_image_new_from_stock(GTK_STOCK_DELETE,
-					    GTK_ICON_SIZE_SMALL_TOOLBAR);
+					    GTK_ICON_SIZE_LARGE_TOOLBAR);
 	delete_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					  _("Delete"),
 					  _("Delete the message"),
@@ -2289,8 +2288,7 @@
 					  G_CALLBACK(toolbar_delete_cb),
 					  mainwin);
 
-	icon_wid = gtk_image_new_from_stock(GTK_STOCK_CANCEL,
-					    GTK_ICON_SIZE_SMALL_TOOLBAR);
+	icon_wid = gtk_image_new_from_icon_name("stock_spam", GTK_ICON_SIZE_LARGE_TOOLBAR);
 	junk_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					   _("Junk"),
 					   _("Set as junk mail"),
@@ -2300,7 +2298,7 @@
 					   mainwin);
 
 	icon_wid = gtk_image_new_from_stock(GTK_STOCK_EXECUTE,
-					    GTK_ICON_SIZE_SMALL_TOOLBAR);
+					    GTK_ICON_SIZE_LARGE_TOOLBAR);
 	exec_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					   _("Execute"),
 					   _("Execute marked process"),
@@ -2310,7 +2308,7 @@
 					   mainwin);
 
 	icon_wid = gtk_image_new_from_stock(GTK_STOCK_GO_DOWN,
-					    GTK_ICON_SIZE_SMALL_TOOLBAR);
+					    GTK_ICON_SIZE_LARGE_TOOLBAR);
 	next_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					   _("Next"),
 					   _("Next unread message"),
@@ -2322,7 +2320,7 @@
 #if 0
 	gtk_toolbar_append_space(GTK_TOOLBAR(toolbar));
 
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_PREFERENCES);
+	icon_wid = gtk_image_new_from_stock(GTK_STOCK_PREFERENCES, GTK_ICON_SIZE_LARGE_TOOLBAR);
 	prefs_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					    _("Prefs"),
 					    _("Common preferences"),
@@ -2330,7 +2328,7 @@
 					    icon_wid,
 					    G_CALLBACK(toolbar_prefs_cb),
 					    mainwin);
-	icon_wid = stock_pixmap_widget(container, STOCK_PIXMAP_PROPERTIES);
+	icon_wid = gtk_image_new_from_stock(GTK_STOCK_PROPERTIES, GTK_ICON_SIZE_LARGE_TOOLBAR);
 	account_btn = gtk_toolbar_append_item(GTK_TOOLBAR(toolbar),
 					      _("Account"),
 					      _("Account setting"),
diff -ruN ../sylpheed-2.0.0beta3.orig/src/pixmaps/mail_receive_all.png.uu ./src/pixmaps/mail_receive_all.png.uu
--- ../sylpheed-2.0.0beta3.orig/src/pixmaps/mail_receive_all.png.uu	Thu Jan  1 01:00:00 1970
+++ ./src/pixmaps/mail_receive_all.png.uu	Sun Jun 26 13:47:39 2005
@@ -0,0 +1,23 @@
+begin 644 mail_receive_all.png
+MB5!.1PT*&@H````-24A$4@```!@````8"`8```#@=SWX````!F)+1T0`_P#_
+M`/^@O:>3````"7!(67,```L3```+$P$`FIP8```##TE$051(Q]663T@<5QS'
+M/SONKGO*.M8VM$5*H(<M:(Q;2[V+BX(XN`N!P$)MA5)D<LDF]J)X\-![8<$>
+MI%AH#6'72PB4DDNVB0@2XL&35T>P,<[HFG6R?V9^.61WF74WQ3^]],&7>;SW
+MY?=]W_=^\WL/_N_-=WK@2OC*_A>]D2['<7CY:H^]G9>O2F_*'UZ4U]3Z!OKV
+M_WKQIV2?/9#)V9O2]6GG_F5X?D^_"V@3<95BN<CKDV,*A=>(B`)<K7+^.0</
+M`,4CL'@]VKO7]U5?)SX?+H+2IO#QM8\ZU:OA/>#^.7E-`G?"'>'<K>]N52<4
+M_`$_G]_XC&`HD`.^.2/OVU9G"_`](%]^'97,DP?RV^-?9?R'8?G@$U6`)OP+
+M[S80JBU>\633+XE$@LS]+,\>K5',%Q$SR,;3Y[BN2Z52H5*I4"J5L&V;WY?_
+MX.^'3RGFBSC[;3S,/"*12`#\#*A`T.M$`<0T3=$T3;:WM^7NCW=E<W-3CH^/
+M)9_/R]'141V'AX=B699L;&Q(ZMX=R>5RHFF:K*ZNUEQ$@+#W"!1`1$0,PY"A
+MH2%Q'*<>W"M@69:8IBFF:<K!P8'L[.R(IFF2R61D?7V])M!;=:$HIP]"5566
+MEY>)Q^.$0B%$!-=U&[XUG)R<,#DYR<3$!)%(I.5_U200#`91595T.LW(R`CM
+M[>WPSEY3<%W7B<?C#`P,T-'1<3:!4JG4X"06BQ$(!!J"V[;-]/0TT6B4_O[^
+M]P9_KP-OZ^GI87Q\O"YBVS:ZKM/=W<WP\'!],6<6J#FP+(OY^7G2Z30S,S,D
+MDTG\?C^ZKA.+Q4@FDV2SV3-7U7H6E<ME,0Q#IJ:F1$2D4"B(81BRM;55SY:U
+MM34Q#$,,PY#9V=EZOU46M4S35"I5#^Y%+=#IL9I(*P&?1\`1$>;FYEA86&!W
+M=_=<%\OBXB)C8V,,#@X"7`<,X*A!()5*,3HZ>N';:V5EA:6EI08![Q;=;E74
+M+H"?O*7"YREV[=5]"P.!2US#Y>K*+:#H.Y5)P2K:+B'@`*4J7%^+1X#O/WA,
+8U+:+M\VX-P;4@`4U`````$E%3D2N0F""
+`
+end
diff -ruN ../sylpheed-2.0.0beta3.orig/src/pixmaps/mail_send_queue.png.uu ./src/pixmaps/mail_send_queue.png.uu
--- ../sylpheed-2.0.0beta3.orig/src/pixmaps/mail_send_queue.png.uu	Thu Jan  1 01:00:00 1970
+++ ./src/pixmaps/mail_send_queue.png.uu	Sun Jun 26 13:47:51 2005
@@ -0,0 +1,26 @@
+begin 644 mail_send_queue.png
+MB5!.1PT*&@H````-24A$4@```!@````8"`8```#@=SWX````!F)+1T0`_P#_
+M`/^@O:>3````"7!(67,```L3```+$P$`FIP8```#JDE$051(Q\V5STMC5Q3'
+M/^_EY26=*9JH6)JNW'3JHMVUPBPL&`5=M+&&=B/2A6!!8SM%J_-'.#JT*(ZZ
+M$+J8TDY7;84(6?D#"]8Z()5:<2@F(XYOC-4D)B;OG2[R@R03!UT4>N#P[GWW
+MW._WGG/N.1?^3^)[_X;\E_@2>_J+`-<B4:\,'OF)U,GO?/6IB^N0J%<'_PT%
+MA=OO.*]%HEX%/!W[$\5V$Q$3X%HDY01*[I^:![\X^P,4';!*#,M(U#)5\G9:
+MT9X^X$$QR*MO?`#`R5_?8%GI`LGR9I)[WYX4FYI%X\^!6>`"L-2BDS_P^_V(
+M2(EF5[72;TX2B03Q>)Q8+,;^_CY^OQ_@:\`-Z(!23,#L["R=G9TEFP&LS"F8
+MR1?BFTPF22:31*-1`H$`W=W=^:5JP/'"D=QN-Y.3D[2VMK*XN$@BD<B2'?S,
+M0>0)N@9_'Z1YQ9$-L8B02"0(!`+T]/3@\7CR4/;R_*I9>Y%X/"[A<%A\/I^D
+MT^E\8574_?U]\7J],C\_+UM;6[*VMI9?>SL7)E4K=UO7]8(G[>WMA3#D<Y+)
+M9#`,@_7U=?KZ^G"YW`2#038V-FAH:,!FLV&:IEKINA8\R&LX');FYF:)Q^-R
+M?'PL1T='LKJZ*H.#@S(T-"07%RG)RVD\)G?OCDIM;:UHFO:1HBBUQ6$J$*33
+MZ1*"_OY^\7J]<G9V)CL[._+EG2_D^X</I5B>&T9A/#4U)75U==]5556]69&@
+M&+RWMU=$1(+!H/A\/@F%0O*Q_Q,IE]'14=G>WB[,V]O;GWL\G@\O]:`8/$^V
+MM;4E+2TMLKR\+"(BIFF*B,C"PH*TM;6)98DD4]F0_?#HD30V-HY73/+AX2$3
+M$Q/,S<T5KJG;[<;M=M/4U(1I9HLVG<[@<.CT]P^@::`H%`I3+`O+LFHJ-KOI
+MZ6G&QL:(1")$H]$2=;E<&(:1==FF<GY^#@B[NT^8N'\?)`.`81A8EG54WHL8
+M'AZFHZ.#4"A4L3,Z'`Y65E;HZNK"KFG8-8W'FX\Y^>>$\W@2I_,F`$M+2[%8
+M++91?DT'7U94@*BJ*KJNR\S,C%PF]\;'T\"/3H?S5G&$%,`)O`Z\E:O$2GH;
+MN*/K^J\C(R.)2#@LJ51*8K%SV=O;D_Z!@02P#GP&O$8V-25>Z#FU7?)^5`,-
+MP'O`N_7U];=J:FI<IFDJAF&<1J/176`#6`4V@6=*A0='><D#=2/78VH41:FV
+LV^UUBJ)4B8AJ658JD\E$@2CP##@"3O\%H+J*M#,%!#P`````245.1*Y"8((`
+`
+end
diff -ruN ../sylpheed-2.0.0beta3.orig/src/stock_pixmap.c ./src/stock_pixmap.c
--- ../sylpheed-2.0.0beta3.orig/src/stock_pixmap.c	Mon May  9 11:29:14 2005
+++ ./src/stock_pixmap.c	Sun Jun 26 14:19:51 2005
@@ -28,8 +28,6 @@
 #include "pixmaps/address.xpm"
 #include "pixmaps/book.xpm"
 #include "pixmaps/category.xpm"
-#include "pixmaps/checkbox_off.xpm"
-#include "pixmaps/checkbox_on.xpm"
 #include "pixmaps/clip.xpm"
 #include "pixmaps/complete.xpm"
 #include "pixmaps/continue.xpm"
@@ -40,33 +38,16 @@
 #include "pixmaps/error.xpm"
 #include "pixmaps/forwarded.xpm"
 #include "pixmaps/group.xpm"
-#include "pixmaps/inbox.xpm"
 #include "pixmaps/interface.xpm"
 #include "pixmaps/jpilot.xpm"
 #include "pixmaps/ldap.xpm"
 #include "pixmaps/linewrap.xpm"
 #include "pixmaps/mark.xpm"
 #include "pixmaps/new.xpm"
-#include "pixmaps/outbox.xpm"
 #include "pixmaps/replied.xpm"
-#include "pixmaps/stock_close.xpm"
-#include "pixmaps/stock_down_arrow.xpm"
-#include "pixmaps/stock_exec.xpm"
-#include "pixmaps/stock_mail.xpm"
-#include "pixmaps/stock_mail_attach.xpm"
-#include "pixmaps/stock_mail_compose.xpm"
-#include "pixmaps/stock_mail_forward.xpm"
-#include "pixmaps/stock_mail_receive.xpm"
-#include "pixmaps/stock_mail_receive_all.xpm"
-#include "pixmaps/stock_mail_reply.xpm"
-#include "pixmaps/stock_mail_reply_to_all.xpm"
-#include "pixmaps/stock_mail_send.xpm"
-#include "pixmaps/stock_mail_send_queue.xpm"
-#include "pixmaps/stock_paste.xpm"
-#include "pixmaps/stock_preferences.xpm"
-#include "pixmaps/stock_properties.xpm"
+#include "pixmaps/mail_receive_all.h"
+#include "pixmaps/mail_send_queue.h"
 #include "pixmaps/sylpheed-logo.xpm"
-#include "pixmaps/tb_address_book.xpm"
 #include "pixmaps/trash.xpm"
 #include "pixmaps/unread.xpm"
 #include "pixmaps/vcard.xpm"
@@ -82,6 +63,8 @@
 	GdkPixmap *pixmap;
 	GdkBitmap *mask;
 	GdkPixbuf *pixbuf;
+	const guint8 *pixbuf_data;
+	int pixbuf_data_len;
 };
 
 static StockPixmapData pixmaps[] =
@@ -89,8 +72,6 @@
 	{address_xpm			, NULL, NULL},
 	{book_xpm			, NULL, NULL},
 	{category_xpm			, NULL, NULL},
-	{checkbox_off_xpm		, NULL, NULL},
-	{checkbox_on_xpm		, NULL, NULL},
 	{clip_xpm			, NULL, NULL},
 	{complete_xpm			, NULL, NULL},
 	{continue_xpm			, NULL, NULL},
@@ -101,33 +82,16 @@
 	{error_xpm			, NULL, NULL},
 	{forwarded_xpm			, NULL, NULL},
 	{group_xpm			, NULL, NULL},
-	{inbox_xpm			, NULL, NULL},
 	{interface_xpm			, NULL, NULL},
 	{jpilot_xpm			, NULL, NULL},
 	{ldap_xpm			, NULL, NULL},
 	{linewrap_xpm			, NULL, NULL},
 	{mark_xpm			, NULL, NULL},
 	{new_xpm			, NULL, NULL},
-	{outbox_xpm			, NULL, NULL},
 	{replied_xpm			, NULL, NULL},
-	{stock_close_xpm		, NULL, NULL},
-	{stock_down_arrow_xpm		, NULL, NULL},
-	{stock_exec_xpm			, NULL, NULL},
-	{stock_mail_xpm			, NULL, NULL},
-	{stock_mail_attach_xpm		, NULL, NULL},
-	{stock_mail_compose_xpm		, NULL, NULL},
-	{stock_mail_forward_xpm		, NULL, NULL},
-	{stock_mail_receive_xpm		, NULL, NULL},
-	{stock_mail_receive_all_xpm	, NULL, NULL},
-	{stock_mail_reply_xpm		, NULL, NULL},
-	{stock_mail_reply_to_all_xpm	, NULL, NULL},
-	{stock_mail_send_xpm		, NULL, NULL},
-	{stock_mail_send_queue_xpm	, NULL, NULL},
-	{stock_paste_xpm		, NULL, NULL},
-	{stock_preferences_xpm		, NULL, NULL},
-	{stock_properties_xpm		, NULL, NULL},
+	{NULL				, NULL, NULL, NULL, mail_receive_all, sizeof(mail_receive_all) },
+	{NULL				, NULL, NULL, NULL, mail_send_queue, sizeof(mail_send_queue) },
 	{sylpheed_logo_xpm		, NULL, NULL},
-	{tb_address_book_xpm		, NULL, NULL},
 	{trash_xpm			, NULL, NULL},
 	{unread_xpm			, NULL, NULL},
 	{vcard_xpm			, NULL, NULL},
@@ -197,9 +161,19 @@
 
 	pix_d = &pixmaps[icon];
 
-	if (!pix_d->pixbuf)
-		pix_d->pixbuf = gdk_pixbuf_new_from_xpm_data
-			((const gchar **)pix_d->data);
+	if (!pix_d->pixbuf) {
+		if (pix_d->data)
+			pix_d->pixbuf = gdk_pixbuf_new_from_xpm_data
+				((const gchar **)pix_d->data);
+		else if (pix_d->pixbuf_data)
+			pix_d->pixbuf = gdk_pixbuf_new_from_inline
+				(pix_d->pixbuf_data_len,
+				 pix_d->pixbuf_data,
+				 FALSE,
+				 NULL);
+		else
+			g_return_val_if_reached(-1);
+	}
 	if (!pix_d->pixbuf)
 		return -1;
 
diff -ruN ../sylpheed-2.0.0beta3.orig/src/stock_pixmap.h ./src/stock_pixmap.h
--- ../sylpheed-2.0.0beta3.orig/src/stock_pixmap.h	Mon May  9 11:29:14 2005
+++ ./src/stock_pixmap.h	Sat Jun 25 19:59:24 2005
@@ -28,8 +28,6 @@
 	STOCK_PIXMAP_ADDRESS,
 	STOCK_PIXMAP_BOOK,
 	STOCK_PIXMAP_CATEGORY,
-	STOCK_PIXMAP_CHECKBOX_OFF,
-	STOCK_PIXMAP_CHECKBOX_ON,
 	STOCK_PIXMAP_CLIP,
 	STOCK_PIXMAP_COMPLETE,
 	STOCK_PIXMAP_CONTINUE,
@@ -40,33 +38,16 @@
 	STOCK_PIXMAP_ERROR,
 	STOCK_PIXMAP_FORWARDED,
 	STOCK_PIXMAP_GROUP,
-	STOCK_PIXMAP_INBOX,
 	STOCK_PIXMAP_INTERFACE,
 	STOCK_PIXMAP_JPILOT,
 	STOCK_PIXMAP_LDAP,
 	STOCK_PIXMAP_LINEWRAP,
 	STOCK_PIXMAP_MARK,
 	STOCK_PIXMAP_NEW,
-	STOCK_PIXMAP_OUTBOX,
 	STOCK_PIXMAP_REPLIED,
-	STOCK_PIXMAP_CLOSE,
-	STOCK_PIXMAP_DOWN_ARROW,
-	STOCK_PIXMAP_EXEC,
-	STOCK_PIXMAP_MAIL,
-	STOCK_PIXMAP_MAIL_ATTACH,
-	STOCK_PIXMAP_MAIL_COMPOSE,
-	STOCK_PIXMAP_MAIL_FORWARD,
-	STOCK_PIXMAP_MAIL_RECEIVE,
 	STOCK_PIXMAP_MAIL_RECEIVE_ALL,
-	STOCK_PIXMAP_MAIL_REPLY,
-	STOCK_PIXMAP_MAIL_REPLY_TO_ALL,
-	STOCK_PIXMAP_MAIL_SEND,
 	STOCK_PIXMAP_MAIL_SEND_QUEUE,
-	STOCK_PIXMAP_PASTE,
-	STOCK_PIXMAP_PREFERENCES,
-	STOCK_PIXMAP_PROPERTIES,
 	STOCK_PIXMAP_SYLPHEED_LOGO,
-	STOCK_PIXMAP_ADDRESS_BOOK,
 	STOCK_PIXMAP_TRASH,
 	STOCK_PIXMAP_UNREAD,
 	STOCK_PIXMAP_VCARD,
