From: Alam Arias <alam.gbc@gmail.com>
Date: Fri, 8 Mar 2013 14:27:02 -0500
Subject: added back support for gtkmm-2.4 builds

--- src/devicewidget.cc.orig	2012-12-17 09:50:42 UTC
+++ src/devicewidget.cc
@@ -63,8 +63,8 @@ DeviceWidget::DeviceWidget(BaseObjectTyp
     for (unsigned i = 0; i < PA_CHANNELS_MAX; i++)
         channelWidgets[i] = NULL;
 
-    offsetAdjustment = Gtk::Adjustment::create(0.0, -2000.0, 2000.0, 10.0, 50.0, 0.0);
-    offsetButton->configure(offsetAdjustment, 0, 2);
+    offsetAdjustment = new Gtk::Adjustment(0.0, -2000.0, 2000.0, 10.0, 50.0, 0.0);
+    offsetButton->configure(*offsetAdjustment, 0.0, 2);
 }
 
 void DeviceWidget::init(MainWindow* mainWindow, Glib::ustring deviceType) {
--- src/devicewidget.h.orig	2012-12-17 09:17:56 UTC
+++ src/devicewidget.h
@@ -98,7 +98,7 @@ protected:
     Gtk::HBox *portSelect, *offsetSelect;
     Gtk::ComboBox *portList;
     Glib::RefPtr<Gtk::ListStore> treeModel;
-    Glib::RefPtr<Gtk::Adjustment> offsetAdjustment;
+    Gtk::Adjustment *offsetAdjustment;
 
 private:
     Glib::ustring mDeviceType;
