--- PEAR/Installer.php.orig	Mon Aug 21 10:20:13 2006
+++ PEAR/Installer.php	Mon Aug 21 10:20:53 2006
@@ -1464,7 +1464,8 @@
                 $this->log(0, $e[0]);
             }
         }
-        // {{{ Delete the files
+        // {{{ Delete the files, if not set 'register-only' option
+        if (!isset($options['register-only'])) {
         $this->startFileTransaction();
         PEAR::pushErrorHandling(PEAR_ERROR_RETURN);
         if (PEAR::isError($err = $this->_deletePackageFiles($package, $channel))) {
@@ -1505,6 +1506,7 @@
             }
         }
         // }}}
+        }
 
         $this->configSet('default_channel', $savechannel);
         // Register that the package is no longer installed
