--- src/makefile.std	Tue Aug 22 17:08:56 2000
+++ src/makefile.std	Thu Jul 20 13:19:50 2000
@@ -133,8 +131,7 @@
 ##       to get 256 colours.
 ##
 CFLAGS = -Wall -O2 -fno-strength-reduce -m486 -pipe -g \
-         -D"USE_X11" -D"USE_GCU" -D"USE_TRANSPARENCY" \
-         -D"JP" -D"EUC"
+         -D"USE_X11" -D"USE_GCU" -D"USE_TRANSPARENCY"
 LIBS = -lX11 -lcurses -ltermcap -L/usr/X11R6/lib
 
 #CFLAGS = -Wall -O2 -fno-strength-reduce -m486 -pipe -g \
