--- kcms/CMakeLists.txt	2022-10-01 22:26:31.912154000 +0100
+++ kcms/CMakeLists.txt	2022-10-01 22:26:57.273631000 +0100
@@ -9,7 +9,6 @@
 endif()
 
 add_subdirectory( access )
-add_subdirectory( dateandtime )
 add_subdirectory( ksplash )
 
 add_subdirectory( launch )
--- CMakeLists.txt	2022-10-05 19:44:55.065430000 +0100
+++ CMakeLists.txt	2022-10-05 20:53:30.202917000 +0100
@@ -53,7 +53,6 @@
     Crash
     Plasma
     PlasmaQuick
-    DocTools
     I18n
     KCMUtils
     NewStuff
@@ -76,9 +75,6 @@
     Sonnet
     Package
 )
-if (QT_MAJOR_VERSION EQUAL "5")
-    find_package(KF5 ${KF5_MIN_VERSION} REQUIRED COMPONENTS KDELibs4Support) # kcms/dateandtime
-endif()
 
 find_package(AccountsQt${QT_MAJOR_VERSION} ${ACCOUNTSQT_DEP_VERSION} CONFIG)
 
@@ -244,7 +240,6 @@
 
 add_subdirectory(layout-templates)
 
-add_subdirectory(doc)
 add_subdirectory(runners)
 add_subdirectory(containments)
 add_subdirectory(toolboxes)
@@ -277,4 +272,3 @@
 )
 
 ki18n_install(po)
-kdoctools_install(po)
