*** unix-gcc.mak.orig	Thu Apr  9 18:16:54 1908
--- unix-gcc.mak	Thu Apr  9 18:35:45 1908
***************
*** 82,95 ****
  #   specific stuff that <math.h> typically needs; nevertheless, we expect
  #   gcc to accept ANSI-style function prototypes and function definitions.
  
! CFLAGS=-g -O $(XCFLAGS) -DCOMPFONT -DWMODE -DKANJI
  
  # Define platform flags for ld.
  # SunOS and some others want -X; Ultrix wants -x.
  # SunOS 4.n may need -Bstatic.
  # XLDFLAGS can be set from the command line.
  
! LDFLAGS=$(XLDFLAGS)
  
  # Define any extra libraries to link into the executable.
  # ISC Unix 2.2 wants -linet.
--- 82,96 ----
  #   specific stuff that <math.h> typically needs; nevertheless, we expect
  #   gcc to accept ANSI-style function prototypes and function definitions.
  
! CFLAGS+= $(XCFLAGS) -DCOMPFONT -DWMODE -DKANJI -D_POSIX_SOURCE \
! 	-D_HAVE_PARAM_H -I/usr/local/include
  
  # Define platform flags for ld.
  # SunOS and some others want -X; Ultrix wants -x.
  # SunOS 4.n may need -Bstatic.
  # XLDFLAGS can be set from the command line.
  
! LDFLAGS=$(XLDFLAGS) -L/usr/local/lib
  
  # Define any extra libraries to link into the executable.
  # ISC Unix 2.2 wants -linet.
***************
*** 104,122 ****
  # Note that x_.h expects to find the header files in $(XINCLUDE)/X11,
  # not in $(XINCLUDE).
  
! XINCLUDE=-I/usr/local/X/include
  
  # Define the directory/ies for the X11 library files.
  # This can be null if these files are in the default linker search path.
  
! XLIBDIRS=-L/usr/local/X/lib
  
  # ------ Devices and features ------ #
  
  # Choose the language feature(s) to include.  See gs.mak for details.
  
  FEATURE_DEVS=filter.dev dps.dev level2.dev kanji.dev \
!              kfpcf.dev kfsnf.dev kfzeit.dev kfztbez.dev kfjtex.dev
  
  # Choose the device(s) to include.  See devs.mak for details.
  
--- 105,124 ----
  # Note that x_.h expects to find the header files in $(XINCLUDE)/X11,
  # not in $(XINCLUDE).
  
! XINCLUDE=-I/usr/X11R6/include
  
  # Define the directory/ies for the X11 library files.
  # This can be null if these files are in the default linker search path.
  
! XLIBDIRS=-L/usr/X11R6/lib
  
  # ------ Devices and features ------ #
  
  # Choose the language feature(s) to include.  See gs.mak for details.
  
  FEATURE_DEVS=filter.dev dps.dev level2.dev kanji.dev \
!              kfpcf.dev kfsnf.dev kfzeit.dev kfztbez.dev kfjtex.dev \
! 	     kfvflib.dev
  
  # Choose the device(s) to include.  See devs.mak for details.
  
***************
*** 1552,1558 ****
  # E.g,. to make A4 paper the default, change the second line below this to
  #	$(CCC) -DA4 gdevdjet.c
  gdevcdj.$(OBJ): gdevcdj.c $(PDEVH) $(gdevpcl_h)
! 	$(CCC) gdevcdj.c
  
  djet500c_=gdevdjtc.$(OBJ) $(HPPCL)
  djet500c.dev: $(djet500c_)
--- 1554,1560 ----
  # E.g,. to make A4 paper the default, change the second line below this to
  #	$(CCC) -DA4 gdevdjet.c
  gdevcdj.$(OBJ): gdevcdj.c $(PDEVH) $(gdevpcl_h)
! 	$(CCC) -DA4 gdevcdj.c
  
  djet500c_=gdevdjtc.$(OBJ) $(HPPCL)
  djet500c.dev: $(djet500c_)
