--- tinderd.orig	2010-01-04 21:12:37.000000000 +0100
+++ tinderd	2010-01-04 21:14:28.000000000 +0100
@@ -36,12 +36,14 @@
 	MAIL=$(echo ${ENTRY} | cut -d: -f5)
 
 	if [ -n "${BUILD}" -a -n "${PORT}" -a -n "${ID}" ] ; then
+	    mount -t tmpfs tmpfs ${pb}/${BUILD}
 	    tinderEcho "INFO: Going to build ${PORT} on ${BUILD}"
 	    ${pb}/scripts/tc updateBuildPortsQueueEntryStatus \
 			-i "${ID}" -s PROCESSING
 	    if ! ${pb}/scripts/tc addPort -b "${BUILD}" -d "${PORT}"; then
 	        ${pb}/scripts/tc updateBuildPortsQueueEntryStatus \
 	        -i "${ID}" -s FAIL
+		umount ${pb}/${BUILD}
 	        continue
 	    fi
 
@@ -64,6 +66,7 @@
 	   fi
 
 	   ${pb}/scripts/tc reorgBuildPortsQueue
+	   umount ${pb}/${BUILD}
 
 	else
 	    tinderEcho "INFO: Nothing to do. Sleeping ${TINDERD_SLEEPTIME} seconds."
