From 7eeeb6f26af05322814ecc77c87f09c72ab2216a Mon Sep 17 00:00:00 2001
From: "Heiko Schlittermann (HS12-RIPE)" <hs@schlittermann.de>
Date: Sun, 28 Mar 2021 10:58:46 +0200
Subject: [PATCH 05/23] parse

---
 src/parse.c | 6 +-----
 1 file changed, 1 insertion(+), 5 deletions(-)

diff --git a/src/parse.c b/src/parse.c
index 3ea758ac9..d1bc79039 100644
--- a/src/parse.c
+++ b/src/parse.c
@@ -1402,12 +1402,8 @@ for (;;)
       return FF_ERROR;
       }
 
-    if (is_tainted(filename))
-      {
-      *error = string_sprintf("Tainted name '%s' for included file  not permitted\n",
-       filename);
+    if (*error = is_tainted2(filename, 0, "Tainted name '%s' for included file not permitted\n", filename))
       return FF_ERROR;
-      }
 
     /* Check file name if required */
 
-- 
2.30.2

