From 015fff57c854184f8bce61476c46a2830a97daf8 Mon Sep 17 00:00:00 2001
From: "Heiko Schlittermann (HS12-RIPE)" <hs@schlittermann.de>
Date: Fri, 2 Apr 2021 08:36:24 +0200
Subject: [PATCH 14/23] rf_get_transport

---
 src/routers/rf_get_transport.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/src/routers/rf_get_transport.c b/src/routers/rf_get_transport.c
index 4a43818ff..32bde9ec3 100644
--- a/src/routers/rf_get_transport.c
+++ b/src/routers/rf_get_transport.c
@@ -66,10 +66,8 @@ if (expandable)
       "\"%s\" in %s router: %s", tpname, router_name, expand_string_message);
     return FALSE;
     }
-  if (is_tainted(ss))
+  if (is_tainted2(ss, LOG_MAIN|LOG_PANIC, "Tainted tainted value '%s' from '%s' for transport", ss, tpname))
     {
-    log_write(0, LOG_MAIN|LOG_PANIC,
-      "attempt to use tainted value '%s' from '%s' for transport", ss, tpname);
     addr->basic_errno = ERRNO_BADTRANSPORT;
     /* Avoid leaking info to an attacker */
     addr->message = US"internal configuration error";
-- 
2.30.2

