From 33d5b8e8e4c2f23b4e834e3a095e3c9dd9f0686b Mon Sep 17 00:00:00 2001
From: "Heiko Schlittermann (HS12-RIPE)" <hs@schlittermann.de>
Date: Sun, 25 Apr 2021 18:58:35 +0200
Subject: [PATCH 1/4] Silence the compiler

---
 src/log.c                   | 4 ++--
 src/transports/appendfile.c | 4 ++--
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/log.c b/src/log.c
index c2ef698e7..11d259197 100644
--- a/src/log.c
+++ b/src/log.c
@@ -651,7 +651,7 @@ set_file_path(BOOL *multiple)
 {
 uschar *s;
 int sep = ':';              /* Fixed separator - outside use */
-uschar *ss = *log_file_path ? log_file_path : LOG_FILE_PATH;
+const uschar *ss = *log_file_path ? log_file_path : US LOG_FILE_PATH;
 
 logging_mode = 0;
 while ((s = string_nextinlist(&ss, &sep, log_buffer, LOG_BUFFER_SIZE)))
@@ -1445,7 +1445,7 @@ unlink_log(lt_debug);
 }
 
 void
-open_logs(const char *m)
+open_logs(void)
 {
 set_file_path(NULL);
 open_log(&mainlogfd, lt_main, 0);
diff --git a/src/transports/appendfile.c b/src/transports/appendfile.c
index 706af6dde..c0f4de4c8 100644
--- a/src/transports/appendfile.c
+++ b/src/transports/appendfile.c
@@ -218,7 +218,7 @@ Returns:     OK, FAIL, or DEFER
 */
 
 void
-openlogs();
+open_logs(void);
 
 static int
 appendfile_transport_setup(transport_instance *tblock, address_item *addrlist,
@@ -235,7 +235,7 @@ uid = uid;
 gid = gid;
 
 /* we can't wait until we're not privileged anymore */
-open_logs("appendfile");
+open_logs();
 
 if (ob->expand_maildir_use_size_file)
 	ob->maildir_use_size_file = expand_check_condition(ob->expand_maildir_use_size_file,
-- 
2.30.2

