From 51be321b27825c01829dffd90f11bfff256f7e42 Mon Sep 17 00:00:00 2001
From: Adam Lackorzynski <adam@l4re.org>
Date: Sat, 16 Oct 2021 16:30:07 +0100
Subject: [PATCH] Fix PAM auth.  Bug 2813

---
 src/auths/call_pam.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/auths/call_pam.c b/src/auths/call_pam.c
index 80bb23ec3..03b9be1a8 100644
--- a/src/auths/call_pam.c
+++ b/src/auths/call_pam.c
@@ -88,7 +88,7 @@ for (int i = 0; i < num_msg; i++)
 	arg = US"";
 	pam_arg_ended = TRUE;
 	}
-      reply[i].resp = CS string_copy_malloc(arg); /* PAM frees resp */
+      reply[i].resp = strdup(CCS arg); /* Use libc malloc, PAM frees resp directly*/
       reply[i].resp_retcode = PAM_SUCCESS;
       break;
 
-- 
2.34.1

