--- bin/named/dlz_postgres_driver.c.orig	Mon Jun 10 05:59:08 2002
+++ bin/named/dlz_postgres_driver.c	Mon Jun 10 05:59:45 2002
@@ -510,15 +510,15 @@
 
 		// free dbi->zone string
 	if(dbi->zone != NULL)
-		isc_mem_free(ns_g_mctx, (char *) dbi->zone);
+		isc_mem_free(ns_g_mctx, dbi->zone);
 
 		// free dbi->record string
 	if(dbi->record != NULL)
-		isc_mem_free(ns_g_mctx, (char *) dbi->record);
+		isc_mem_free(ns_g_mctx, dbi->record);
 
 		// free dbi->client string
 	if(dbi->client != NULL)
-		isc_mem_free(ns_g_mctx, (char *) dbi->client);
+		isc_mem_free(ns_g_mctx, dbi->client);
 
 #ifdef ISC_PLATFORM_USETHREADS
 
