--- public/components/all/package.go.orig
+++ public/components/all/package.go
@@ -45,8 +45,6 @@
 	_ "github.com/redpanda-data/connect/v4/public/components/redis"
 	_ "github.com/redpanda-data/connect/v4/public/components/sentry"
 	_ "github.com/redpanda-data/connect/v4/public/components/sftp"
-	_ "github.com/redpanda-data/connect/v4/public/components/snowflake"
-	_ "github.com/redpanda-data/connect/v4/public/components/splunk"
 	_ "github.com/redpanda-data/connect/v4/public/components/sql"
 	_ "github.com/redpanda-data/connect/v4/public/components/statsd"
 	_ "github.com/redpanda-data/connect/v4/public/components/twitter"
