--- public/components/all/package.go.orig	1979-11-30 00:00:00.000000000 +0100
+++ public/components/all/package.go	2024-07-20 15:56:19.212115000 +0200
@@ -21,8 +21,4 @@
 import (
 	// Import all community components.
 	_ "github.com/redpanda-data/connect/v4/public/components/community"
-
-	// Import all enterprise components.
-	_ "github.com/redpanda-data/connect/v4/public/components/snowflake"
-	_ "github.com/redpanda-data/connect/v4/public/components/splunk"
 )
