--- gss-serv.c.orig	Sat Nov  5 02:07:05 2005
+++ gss-serv.c	Thu Feb  2 22:45:37 2006
@@ -134,6 +134,16 @@
 	OM_uint32 offset;
 	OM_uint32 oidl;
 
+#if defined(HEIMDAL)
+	if (strcmp(heimdal_version, "Heimdal 0.6") < 0) {
+		name->length = ename->length;
+		name->value = xmalloc(name->length+1);
+		memcpy(name->value,ename->value,name->length);
+		((char *)name->value)[name->length] = 0;
+		return GSS_S_COMPLETE;
+	}
+#endif
+
 	tok = ename->value;
 
 	/*
