diff -Nuar --exclude '*.rej' servconf.c.orig servconf.c
--- servconf.c.orig	2008-08-23 15:02:47.000000000 -0700
+++ servconf.c	2008-08-23 15:04:21.000000000 -0700
@@ -701,6 +701,7 @@
 	int cmdline = 0, *intptr, value, n;
 	SyslogFacility *log_facility_ptr;
 	LogLevel *log_level_ptr;
+ 	unsigned long lvalue, *longptr;
 	ServerOpCodes opcode;
 	u_short port;
 	u_int i, flags = 0;
@@ -715,6 +716,7 @@
 	if (!arg || !*arg || *arg == '#')
 		return 0;
 	intptr = NULL;
+	longptr = NULL;
 	charptr = NULL;
 	opcode = parse_token(arg, filename, linenum, &flags);
 
@@ -1449,11 +1451,20 @@
 			*intptr = value;
 		break;
 	case sBindTimeout:
-		intptr = (int *) &options->lpk.b_timeout.tv_sec;
-		goto parse_int;
+		longptr = (unsigned long *) &options->lpk.b_timeout.tv_sec;
+parse_ulong:
+		arg = strdelim(&cp);
+		if (!arg || *arg == '\0')
+			fatal("%s line %d: missing integer value.",
+			    filename, linenum);
+		lvalue = atol(arg);
+		if (*activep && *longptr == -1)
+			*longptr = lvalue;
+		break;
+
 	case sSearchTimeout:
-		intptr = (int *) &options->lpk.s_timeout.tv_sec;
-		goto parse_int;
+		longptr = (unsigned long *) &options->lpk.s_timeout.tv_sec;
+		goto parse_ulong;
 		break;
 	case sLdapConf:
 		arg = cp;
