--- servconf.c.orig	Sat Mar 23 11:02:41 2002
+++ servconf.c	Sat Mar 23 11:07:39 2002
@@ -17,12 +17,12 @@
 #endif
 #if defined(KRB5)
 #ifdef HEIMDAL
-#include <krb.h>
+#include <krb5.h>
 #else
 /* Bodge - but then, so is using the kerberos IV KEYFILE to get a Kerberos V
  * keytab */
-#define KEYFILE "/etc/krb5.keytab"
 #endif
+#define KEYFILE "/etc/krb5.keytab"
 #endif
 #ifdef AFS
 #include <kafs.h>
