--- src/Makefile.am.orig
+++ src/Makefile.am
@@ -9,8 +9,8 @@
 bin_PROGRAMS = xmlcopyeditor
 bindir = ${prefix}/bin
 xmlcopyeditordir = ${prefix}/share/xmlcopyeditor
-pixmapdir = /usr/share/pixmaps
-applicationsdir = /usr/share/applications
+pixmapdir = ${prefix}/share/pixmaps
+applicationsdir = ${prefix}/share/applications
 
 # the application source, library search path, and link libraries
 xmlcopyeditor_SOURCES = xmlcopyeditor.cpp associatedialog.cpp casehandler.cpp \
@@ -49,15 +49,6 @@
 	templates/*.lzx \
 	templates/*.xlf \
 	help/*.* help/html/*.* \
-	copying/*.txt copying/boost_shared_ptr/* \
-	copying/Expat/* copying/libxml2/* copying/libxslt/* \
-	copying/PCRE/* copying/Scintilla/* \
-	copying/wxStyledTextCtrl/* \
-	copying/wxStEditor/* \
-	copying/xmlcopyeditor/* \
-	copying/Aspell/* \
-	copying/Apache/LICENSE-2.0 \
-	copying/Apache/LICENSE-2_files/* \
 	po/sk/messages.mo \
 	po/sv/messages.mo \
 	po/fr/messages.mo \
