--- src/xmlcopyeditor.cpp.orig
+++ src/xmlcopyeditor.cpp
@@ -3002,10 +3002,7 @@
 
 		nconv = iconv (
 		            cd,
-#ifdef __WXMSW__
-		            ( const char ** )
-#endif
-		            &docBuffer,
+			    ICONV_CHAR_CAST(&docBuffer),
 		            &docBufferLeft,
 		            &iconvBuffer,
 		            &iconvBufferLeft );
@@ -4617,10 +4614,7 @@
 
 					nconv = iconv (
 					            cd,
-#ifdef __WXMSW__
-					            ( const char ** )
-#endif
-					            &utf8BufferPtr,
+						    ICONV_CHAR_CAST(&utf8BufferPtr),
 					            &utf8BufferLeft,
 					            &iconvBuffer,
 					            &iconvBufferLeft );
