--- src/Makefile.am.orig	2021-11-05 21:58:53 UTC
+++ src/Makefile.am
@@ -9,8 +9,8 @@ noinst_HEADERS = $(srcdir)/*.h
 bin_PROGRAMS = xmlcopyeditor
 bindir = ${prefix}/bin
 xmlcopyeditordir = ${prefix}/share/xmlcopyeditor
-pixmapdir = /usr/share/pixmaps
-applicationsdir = /usr/share/applications
+pixmapdir = ${prefix}/share/pixmaps
+applicationsdir = ${prefix}/share/applications
 
 # the application source, library search path, and link libraries
 xmlcopyeditor_SOURCES = aboutdialog.cpp \
