--- src/xmlcopyeditor.cpp.orig	Thu Dec 20 19:39:44 2007
+++ src/xmlcopyeditor.cpp	Thu Dec 20 19:42:26 2007
@@ -2977,10 +2977,7 @@
 
         nconv = iconv (
                     cd,
-#ifdef __WXMSW__
-                    ( const char ** )
-#endif
-                    &docBuffer,
+                    ICONV_CHAR_CAST(&docBuffer),
                     &docBufferLeft,
                     &iconvBuffer,
                     &iconvBufferLeft );
@@ -4685,10 +4682,7 @@
 
                     nconv = iconv (
                                 cd,
-#ifdef __WXMSW__
-                                ( const char ** )
-#endif
-                                &utf8BufferPtr,
+                                ICONV_CHAR_CAST(&utf8BufferPtr),
                                 &utf8BufferLeft,
                                 &iconvBuffer,
                                 &iconvBufferLeft );
