From 72eb25c26107f31bcf6d8cd02e47ec16a69082ce Mon Sep 17 00:00:00 2001
From: "Alexander V. Chernikov" <melifaro@yandex-team.ru>
Date: Wed, 21 Jan 2015 19:12:33 +0000
Subject: [PATCH 3/3] Locks fix from list.

---
 nest/locks.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git nest/locks.c nest/locks.c
index 7044d6a..8ee5898 100644
--- nest/locks.c
+++ nest/locks.c
@@ -68,7 +68,7 @@ olock_free(resource *r)
 	  DBG("olock: -> %p becomes locked\n", n);
 	  q = SKIP_BACK(struct object_lock, n, n);
 	  rem_node(n);
-	  add_tail_list(&l->waiters, &q->waiters);
+	  add_tail_list(&q->waiters, &l->waiters);
 	  q->state = OLOCK_STATE_EVENT;
 	  add_head(&olock_list, n);
 	  ev_schedule(olock_event);
-- 
2.1.2

