--- tools/designer/src/lib/uilib/formbuilderextra.cpp
+++ tools/designer/src/lib/uilib/formbuilderextra.cpp
@@ -33,9 +33,7 @@ namespace QFormInternal {
 #endif

 void uiLibWarning(const QString &message) {
-    QString prefixedMessage = QLatin1String("Designer: ");
-    prefixedMessage += message;
-    qWarning(prefixedMessage.toUtf8().constData());
+    qWarning("Designer: %s", qPrintable(message));
 }

 QFormBuilderExtra::QFormBuilderExtra() :
--- tools/designer/src/lib/shared/qdesigner_utils.cpp
+++ tools/designer/src/lib/shared/qdesigner_utils.cpp
@@ -40,9 +40,7 @@ namespace qdesigner_internal
 {
     QDESIGNER_SHARED_EXPORT void designerWarning(const QString &message)
     {
-        QString prefixedMessage = QLatin1String("Designer: ");
-        prefixedMessage += message;
-        qWarning(prefixedMessage.toUtf8().constData());
+        qWarning("Designer: %s", qPrintable(message));
     }

     QString EnumType::id() const