Index: etc/pulledpork.conf
===================================================================
--- etc/pulledpork.conf	(revision 268)
+++ etc/pulledpork.conf	(working copy)
@@ -121,12 +121,12 @@
 
 # Define your distro, this is for the precompiled shared object libs!
 # Valid Distro Types:
-# Debian-5-0, Debian-6-0,
-# Ubuntu-8.04, Ubuntu-10-4
-# Centos-4-8, Centos-5-4
+# Debian-6-0,
+# Ubuntu-10-4, Ubuntu-12-04
+# Centos-5-4
 # FC-12, FC-14, RHEL-5-5, RHEL-6-0
-# FreeBSD-7-3, FreeBSD-8-1
-# OpenBSD-4-8
+# FreeBSD-8-1, FreeBSD-9-0
+# OpenBSD-4-8, OpenBSD-5-2, OpenBSD-5-3
 # Slackware-13-1
 distro=FreeBSD-8.1
 
Index: pulledpork.pl
===================================================================
--- pulledpork.pl	(revision 268)
+++ pulledpork.pl	(working copy)
@@ -144,10 +144,10 @@
    -D What Distro are you running on, for the so_rules
       For latest supported options see http://www.snort.org/snort-rules/shared-object-rules
       Valid Distro Types:
-		Debian-5-0, Debian-6-0, Ubuntu-8.04, Ubuntu-10-4
-		Centos-4-8, Centos-5-4,	FC-12, FC-14, RHEL-5-5, RHEL-6-0
-		FreeBSD-7-3, FreeBSD-8-1
-		OpenBSD-4-8
+		Debian-6-0, Ubuntu-10-4, Ubuntu-12-04
+		Centos-5-4, FC-12, FC-14, RHEL-5-5, RHEL-6-0
+		FreeBSD-8-1, FreeBSD-9-0
+		OpenBSD-4-8, OpenBSD-5-2, OpenBSD-5-3
 		Slackware-13-1
    -e Where the enablesid config file lives.
    -E Write ONLY the enabled rules to the output files.
