--- sample.conf.orig	2008-07-01 22:09:31.000000000 +0200
+++ sample.conf	2008-07-05 13:39:47.000000000 +0200
@@ -107,9 +107,12 @@
 
 PageType	htm*
 PageType	cgi
+#PageType	shtml
 #PageType	phtml
 #PageType	php3
+#PageType	php
 #PageType	pl
+#PageType	rb
 
 # PagePrefix allows all requests with a specified prefix to be
 # considered as 'pages'. If you want everything under /documents
@@ -242,7 +245,8 @@
 # is 80 characters, so use multiple lines if needed.
 
 #HTMLHead <META NAME="author" CONTENT="The Webalizer">
-#HTMLHead <META NAME="ROBOTS" CONTENT="NOINDEX, NOFOLLOW">
+HTMLHead <META NAME="ROBOTS" CONTENT="NOINDEX, NOFOLLOW">
+HTMLHead <META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=x-euc-jp">
 
 # HTMLBody defined the HTML code to be inserted, starting with the
 # <BODY> tag.  If not specified, the default is shown below.  If
@@ -545,6 +550,9 @@
 HideURL		*.png
 HideURL		*.PNG
 HideURL		*.ra
+HideURL		*.css
+HideURL		*.CSS
+HideURL		*.ico
 
 # Hiding agents is kind of futile
 #HideAgent	RealPlayer
@@ -564,6 +572,11 @@
 #GroupReferrer	excite.com/     Excite
 #GroupReferrer	infoseek.com/   InfoSeek
 #GroupReferrer	webcrawler.com/ WebCrawler
+#GroupReferrer	yahoo.co.jp/	Yahoo!Japan
+#GroupReferrer	google.co.jp/	GoogleJapan
+#GroupReferrer	infoseek.co.jp/	InfoSeekJapan
+#GroupReferrer	goo.ne.jp/	Goo
+#GroupReferrer	msn.co.jp/	MSNJapan
 
 #GroupUser      root            Admin users
 #GroupUser      admin           Admin users
@@ -702,6 +715,21 @@
 #SearchEngine	alltheweb.com	q=
 #SearchEngine	northernlight.com  qr=
 
+SearchEngine	yahoo.co.jp	p=
+SearchEngine	google.co.jp	q=
+SearchEngine	infoseek.co.jp	qt=
+SearchEngine	msn.co.jp	q=
+# ocn
+SearchEngine	goo.ne.jp	MT=
+SearchEngine	biglobe.ne.jp	q=
+SearchEngine	nifty.com	Text=
+# so-net odn
+SearchEngine	excite.co.jp	search=
+SearchEngine	livedoor.com	q=
+SearchEngine	jp.aol.com	query=
+#SearchEngine .google.        q=
+#SearchEngine bulkfeeds.net   q=
+
 # Normally, search strings are converted to lower case in order to
 # increase accuracy.  The SearchCaseI option allows them to maintain
 # case sensitivity, useful for some sites.  The value can be 'yes'
