--- lib/ohai/plugins/network.rb.orig	2012-10-23 13:08:01.000000000 +0900
+++ lib/ohai/plugins/network.rb	2012-10-23 13:25:46.000000000 +0900
@@ -37,7 +37,8 @@
 
 def find_ip_and_mac(addresses, match = nil)
   ip = nil; mac = nil; ip6 = nil
-  addresses.keys.each do |addr|
+  sorted_addresses = addresses.keys.sort { |x, y| IPAddress("#{x}/#{addresses[x]["netmask"]}").prefix <=> IPAddress("#{y}/#{addresses[y]["netmask"]}").prefix }
+  sorted_addresses.each do |addr|
     if match.nil?
       ip = addr if addresses[addr]["family"].eql?("inet")
     else
