--- acroread.orig	Thu Mar  9 09:14:30 2000
+++ acroread	Thu Mar  9 09:24:53 2000
@@ -305,10 +305,14 @@
   alphaosf)
     LD_LIBRARY_PATH="`prepend "$ACRO_INSTALL_DIR/$ACRO_CONFIG/lib:$ACRO_INSTALL_DIR/$ACRO_CONFIG/lib" "$LD_LIBRARY_PATH"`"
     export LD_LIBRARY_PATH
+    LC_CTYPE="C"
+    export LC_CTYPE
     ;;
   intellinux)
     LD_LIBRARY_PATH="`prepend "$ACRO_INSTALL_DIR/$ACRO_CONFIG/lib:$ACRO_INSTALL_DIR/$ACRO_CONFIG/lib" "$LD_LIBRARY_PATH"`"
     export LD_LIBRARY_PATH
+    LC_CTYPE="C"
+    export LC_CTYPE
     ;;
   intelbsdi)
     ;;
