diff --git gdk/gdkenumtypes.c.template gdk/gdkenumtypes.c.template
index df9edfefa1..5a15d05dd2 100644
--- gdk/gdkenumtypes.c.template
+++ gdk/gdkenumtypes.c.template
@@ -12,9 +12,9 @@
 GType
 @enum_name@_get_type (void)
 {
-  static volatile gsize g_define_type_id__volatile = 0;
+  static volatile GType g_define_type_id__volatile = 0;
 
-  if (g_once_init_enter (&g_define_type_id__volatile))
+  if (g_once_init_enter_pointer (&g_define_type_id__volatile))
     {
       static const G@Type@Value values[] = {
 /*** END value-header ***/
@@ -28,7 +28,7 @@ GType
       };
       GType g_define_type_id =
         g_@type@_register_static (g_intern_static_string ("@EnumName@"), values);
-      g_once_init_leave (&g_define_type_id__volatile, g_define_type_id);
+      g_once_init_leave_pointer (&g_define_type_id__volatile, g_define_type_id);
     }
 
   return g_define_type_id__volatile;
diff --git gtk/a11y/gtkcellaccessibleparent.c gtk/a11y/gtkcellaccessibleparent.c
index bb0590c1ee..1c1475a708 100644
--- gtk/a11y/gtkcellaccessibleparent.c
+++ gtk/a11y/gtkcellaccessibleparent.c
@@ -23,9 +23,9 @@
 GType
 gtk_cell_accessible_parent_get_type (void)
 {
-  static volatile gsize g_define_type_id__volatile = 0;
+  static volatile GType g_define_type_id__volatile = 0;
 
-  if (g_once_init_enter (&g_define_type_id__volatile))
+  if (g_once_init_enter_pointer (&g_define_type_id__volatile))
     {
       GType g_define_type_id =
         g_type_register_static_simple (G_TYPE_INTERFACE,
@@ -36,7 +36,7 @@ gtk_cell_accessible_parent_get_type (void)
                                        NULL,
                                        0);
 
-      g_once_init_leave (&g_define_type_id__volatile, g_define_type_id);
+      g_once_init_leave_pointer (&g_define_type_id__volatile, g_define_type_id);
     }
 
   return g_define_type_id__volatile;
diff --git gtk/gtkcssnodestylecache.c gtk/gtkcssnodestylecache.c
index c5d7562119..9a002207f4 100644
--- gtk/gtkcssnodestylecache.c
+++ gtk/gtkcssnodestylecache.c
@@ -28,9 +28,15 @@ struct _GtkCssNodeStyleCache {
   GHashTable  *children;
 };
 
+#if defined(__CHERI_PURE_CAPABILITY__)
+#define UNPACK_DECLARATION(packed) ((GtkCssNodeDeclaration *) ((guintptr)packed & ~0x3))
+#define UNPACK_FLAGS(packed) ((ptraddr_t)packed & 0x3)
+#define PACK(decl, first_child, last_child) (gpointer) ((guintptr) decl | ((first_child) ? 0x2 : 0) | ((last_child) ? 0x1 : 0))
+#else   // !__CHERI_PURE_CAPABILITY__
 #define UNPACK_DECLARATION(packed) ((GtkCssNodeDeclaration *) (GPOINTER_TO_SIZE (packed) & ~0x3))
 #define UNPACK_FLAGS(packed) (GPOINTER_TO_SIZE (packed) & 0x3)
 #define PACK(decl, first_child, last_child) GSIZE_TO_POINTER (GPOINTER_TO_SIZE (decl) | ((first_child) ? 0x2 : 0) | ((last_child) ? 0x1 : 0))
+#endif  // !__CHERI_PURE_CAPABILITY__
 
 GtkCssNodeStyleCache *
 gtk_css_node_style_cache_new (GtkCssStyle *style)
diff --git gtk/gtkcssstylefuncs.c gtk/gtkcssstylefuncs.c
index 466863114a..086e999f50 100644
--- gtk/gtkcssstylefuncs.c
+++ gtk/gtkcssstylefuncs.c
@@ -67,11 +67,11 @@ register_conversion_function (GType               type,
                               GtkStyleComputeFunc compute)
 {
   if (parse)
-    g_hash_table_insert (parse_funcs, GSIZE_TO_POINTER (type), parse);
+    g_hash_table_insert (parse_funcs, GTYPE_TO_POINTER (type), parse);
   if (print)
-    g_hash_table_insert (print_funcs, GSIZE_TO_POINTER (type), print);
+    g_hash_table_insert (print_funcs, GTYPE_TO_POINTER (type), print);
   if (compute)
-    g_hash_table_insert (compute_funcs, GSIZE_TO_POINTER (type), compute);
+    g_hash_table_insert (compute_funcs, GTYPE_TO_POINTER (type), compute);
 }
 
 static void
@@ -1082,10 +1082,10 @@ _gtk_css_style_funcs_parse_value (GValue       *value,
   gtk_css_style_funcs_init ();
 
   func = g_hash_table_lookup (parse_funcs,
-                              GSIZE_TO_POINTER (G_VALUE_TYPE (value)));
+                              GTYPE_TO_POINTER (G_VALUE_TYPE (value)));
   if (func == NULL)
     func = g_hash_table_lookup (parse_funcs,
-                                GSIZE_TO_POINTER (g_type_fundamental (G_VALUE_TYPE (value))));
+                                GTYPE_TO_POINTER (g_type_fundamental (G_VALUE_TYPE (value))));
 
   if (func == NULL)
     {
@@ -1115,10 +1115,10 @@ _gtk_css_style_funcs_print_value (const GValue *value,
   gtk_css_style_funcs_init ();
 
   func = g_hash_table_lookup (print_funcs,
-                              GSIZE_TO_POINTER (G_VALUE_TYPE (value)));
+                              GTYPE_TO_POINTER (G_VALUE_TYPE (value)));
   if (func == NULL)
     func = g_hash_table_lookup (print_funcs,
-                                GSIZE_TO_POINTER (g_type_fundamental (G_VALUE_TYPE (value))));
+                                GTYPE_TO_POINTER (g_type_fundamental (G_VALUE_TYPE (value))));
 
   if (func == NULL)
     {
@@ -1162,10 +1162,10 @@ _gtk_css_style_funcs_compute_value (GtkStyleProviderPrivate *provider,
   gtk_css_style_funcs_init ();
 
   func = g_hash_table_lookup (compute_funcs,
-                              GSIZE_TO_POINTER (target_type));
+                              GTYPE_TO_POINTER (target_type));
   if (func == NULL)
     func = g_hash_table_lookup (compute_funcs,
-                                GSIZE_TO_POINTER (g_type_fundamental (target_type)));
+                                GTYPE_TO_POINTER (g_type_fundamental (target_type)));
 
   if (func)
     return func (provider, style, parent_style, specified);
diff --git gtk/gtkprivatetypebuiltins.c.template gtk/gtkprivatetypebuiltins.c.template
index 2565208bfc..b38d8c2203 100644
--- gtk/gtkprivatetypebuiltins.c.template
+++ gtk/gtkprivatetypebuiltins.c.template
@@ -14,9 +14,9 @@
 GType
 _@enum_name@_get_type (void)
 {
-  static volatile gsize g_define_type_id__volatile = 0;
+  static volatile GType g_define_type_id__volatile = 0;
 
-  if (g_once_init_enter (&g_define_type_id__volatile))
+  if (g_once_init_enter_pointer (&g_define_type_id__volatile))
     {
       static const G@Type@Value values[] = {
 /*** END value-header ***/
@@ -30,7 +30,7 @@ _@enum_name@_get_type (void)
       };
       GType g_define_type_id =
         g_@type@_register_static (g_intern_static_string ("@EnumName@"), values);
-      g_once_init_leave (&g_define_type_id__volatile, g_define_type_id);
+      g_once_init_leave_pointer (&g_define_type_id__volatile, g_define_type_id);
     }
 
   return g_define_type_id__volatile;
diff --git gtk/gtkselection.c gtk/gtkselection.c
index 32d9de88bf..248380fd82 100644
--- gtk/gtkselection.c
+++ gtk/gtkselection.c
@@ -3345,8 +3345,11 @@ gtk_selection_bytes_per_item (gint format)
       return sizeof (short);
       break;
     case 32:
+#if defined(__CHERI_PURE_CAPABILITY__)
+      return sizeof (void *);
+#else
       return sizeof (long);
-      break;
+#endif
     default:
       g_assert_not_reached();
     }
diff --git gtk/gtktypebuiltins.c.template gtk/gtktypebuiltins.c.template
index f4d748b7b9..9799cd6a5b 100644
--- gtk/gtktypebuiltins.c.template
+++ gtk/gtktypebuiltins.c.template
@@ -13,9 +13,9 @@
 GType
 @enum_name@_get_type (void)
 {
-  static volatile gsize g_define_type_id__volatile = 0;
+  static volatile GType g_define_type_id__volatile = 0;
 
-  if (g_once_init_enter (&g_define_type_id__volatile))
+  if (g_once_init_enter_pointer (&g_define_type_id__volatile))
     {
       static const G@Type@Value values[] = {
 /*** END value-header ***/
@@ -29,7 +29,7 @@ GType
       };
       GType g_define_type_id =
         g_@type@_register_static (g_intern_static_string ("@EnumName@"), values);
-      g_once_init_leave (&g_define_type_id__volatile, g_define_type_id);
+      g_once_init_leave_pointer (&g_define_type_id__volatile, g_define_type_id);
     }
 
   return g_define_type_id__volatile;
diff --git gtk/gtkwidget.c gtk/gtkwidget.c
index b6e00115bb..10b7c1d3ba 100644
--- gtk/gtkwidget.c
+++ gtk/gtkwidget.c
@@ -16916,7 +16916,7 @@ get_auto_child_hash (GtkWidget *widget,
     }
 
   auto_child_hash =
-    g_hash_table_lookup (auto_children, GSIZE_TO_POINTER (type));
+    g_hash_table_lookup (auto_children, GTYPE_TO_POINTER (type));
 
   if (!auto_child_hash && create)
     {
@@ -16926,7 +16926,7 @@ get_auto_child_hash (GtkWidget *widget,
 					       (GDestroyNotify)g_object_unref);
 
       g_hash_table_insert (auto_children,
-			   GSIZE_TO_POINTER (type),
+			   GTYPE_TO_POINTER (type),
 			   auto_child_hash);
     }
 
diff --git gtk/inspector/signals-list.c gtk/inspector/signals-list.c
index 523856eac5..250f50caf9 100644
--- gtk/inspector/signals-list.c
+++ gtk/inspector/signals-list.c
@@ -76,10 +76,10 @@ get_types (GObject *object, guint *length)
   type = ((GTypeInstance*)object)->g_class->g_type;
   while (type)
     {
-      g_hash_table_add (seen, GSIZE_TO_POINTER (type));
+      g_hash_table_add (seen, GTYPE_TO_POINTER (type));
       iface = g_type_interfaces (type, NULL);
       for (i = 0; iface[i]; i++)
-        g_hash_table_add (seen, GSIZE_TO_POINTER (iface[i]));
+        g_hash_table_add (seen, GTYPE_TO_POINTER (iface[i]));
       g_free (iface);
       type = g_type_parent (type);
     }
diff --git gtk/inspector/statistics.c gtk/inspector/statistics.c
index 42b315c49d..403ceaba0d 100644
--- gtk/inspector/statistics.c
+++ gtk/inspector/statistics.c
@@ -93,7 +93,7 @@ add_type_count (GtkInspectorStatistics *sl, GType type)
   for (i = 0; i < n_children; i++)
     cumulative += add_type_count (sl, children[i]);
 
-  data = g_hash_table_lookup (sl->priv->counts, GSIZE_TO_POINTER (type));
+  data = g_hash_table_lookup (sl->priv->counts, GTYPE_TO_POINTER (type));
   if (!data)
     {
       data = g_new0 (TypeData, 1);
@@ -107,7 +107,7 @@ add_type_count (GtkInspectorStatistics *sl, GType type)
                           COLUMN_SELF_DATA, data->self,
                           COLUMN_CUMULATIVE_DATA, data->cumulative,
                           -1);
-      g_hash_table_insert (sl->priv->counts, GSIZE_TO_POINTER (type), data);
+      g_hash_table_insert (sl->priv->counts, GTYPE_TO_POINTER (type), data);
     }
 
   self = g_type_get_instance_count (type);
diff --git testsuite/gtk/objects-finalize.c testsuite/gtk/objects-finalize.c
index 24540e313f..338d5c9564 100644
--- testsuite/gtk/objects-finalize.c
+++ testsuite/gtk/objects-finalize.c
@@ -49,7 +49,7 @@ check_finalized (gpointer data,
 static void
 test_finalize_object (gconstpointer data)
 {
-  GType test_type = GPOINTER_TO_SIZE (data);
+  GType test_type = GPOINTER_TO_TYPE (data);
   GObject *object;
 
   object = g_object_new (test_type, NULL);
@@ -125,7 +125,7 @@ main (int argc, char **argv)
 	{
 	  gchar *test_path = g_strdup_printf ("/FinalizeObject/%s", g_type_name (all_types[i]));
 
-	  g_test_add_data_func (test_path, GSIZE_TO_POINTER (all_types[i]), test_finalize_object);
+	  g_test_add_data_func (test_path, GTYPE_TO_POINTER (all_types[i]), test_finalize_object);
 
 	  g_free (test_path);
 	}
