--- src/multitrack.c	2007/05/30 20:22:08	1.214
+++ src/multitrack.c	2007/05/31 19:15:24	1.215

@@ -5357,6 +5358,9 @@
   if (prefs->open_maximised) {
     gtk_window_maximize (GTK_WINDOW(multi->window));
   }
+
+  if (event_list!=NULL) mt_show_current_frame(multi);
+
   return TRUE;
 }



--- src/events.c	2007/05/30 05:50:22	1.138
+++ src/events.c	2007/05/31 19:15:23	1.139
@@ -3460,6 +3460,7 @@
 	mainw->unordered_blocks=TRUE;
 	if (on_multitrack_activate (NULL, (gpointer)mainw->event_list)) {
 	  mainw->event_list=NULL;
+	  new_clip=TRUE;
 	}
 	else render_choice=RENDER_CHOICE_PREVIEW;
 	break;
