--- smsd/Makefile.orig	Sun Nov 13 16:00:45 2005
+++ smsd/Makefile	Sun Nov 13 16:00:54 2005
@@ -7,14 +7,14 @@
 # $Id: Makefile,v 1.33 2005/08/10 07:33:43 ja Exp $
 #
 
-#GLIBCFLAGS = glib-config --cflags
-#GLIBLDLIBS = glib-config --libs
+GLIBCFLAGS = glib-config --cflags
+GLIBLDLIBS = glib-config --libs
 
 # Comment out above lines and uncomment below lines if you want compile smsd
 # with glib2 instead of glib1.
 #
-GLIBCFLAGS = pkg-config --cflags glib-2.0
-GLIBLDLIBS = pkg-config --libs glib-2.0
+#GLIBCFLAGS = pkg-config --cflags glib-2.0
+#GLIBLDLIBS = pkg-config --libs glib-2.0
 
 TOPDIR=..
 include $(TOPDIR)/Makefile.global
