--- Makefile.PL.orig	Mon Sep 27 21:43:08 2004
+++ Makefile.PL	Mon Sep 27 21:43:33 2004
@@ -1,15 +1,15 @@
-use 5.008;
+use 5.005;
 use ExtUtils::MakeMaker;
 # See lib/ExtUtils/MakeMaker.pm for details of how to influence
 # the contents of the Makefile that is written.
 
-eval { require 5.6.0 }
-    or die <<'EOD';
-###
-## I have not tested this module in versions of perl lower than 5.8.0.
-## However, 5.6.0 should be fine.
-###
-EOD
+#eval { require 5.6.0 }
+#    or die <<'EOD';
+####
+### I have not tested this module in versions of perl lower than 5.8.0.
+### However, 5.6.0 should be fine.
+####
+#EOD
 
 WriteMakefile(
     'NAME'		=> 'Qmail::Envelope',
--- Envelope.pm.orig	Fri Apr 16 17:53:14 2004
+++ Envelope.pm	Mon Sep 27 21:51:28 2004
@@ -1,7 +1,8 @@
 package Qmail::Envelope;
 use strict;
 
-our $VERSION = '0.52';
+use vars qw/$VERSION/;
+$VERSION = '0.52';
 
 
 ## constructor
