--- crengine/src/cp_stats.cpp.orig	2021-02-01 20:29:51 UTC
+++ crengine/src/cp_stats.cpp
@@ -2266,8 +2266,8 @@ cp_stat_t cp_stat_table[] = {
 {ch_stat_cp850_es2,dbl_ch_stat_cp850_es2,"cp850","es"}, 
 {ch_stat_utf8_ee1,dbl_ch_stat_utf8_ee1,"utf8","ee"}, 
 {ch_stat_cp775_ee1,dbl_ch_stat_cp775_ee1,"cp775","ee"}, 
-{ch_stat_cp1253_gr1,dbl_ch_stat_cp1253_gr1,"cp1253","gr"}, 
-{ch_stat_cp737_gr1,dbl_ch_stat_cp737_gr1,"cp737","gr"}, 
+{ch_stat_cp1253_gr1,dbl_ch_stat_cp1253_gr1,"cp1253","el"}, 
+{ch_stat_cp737_gr1,dbl_ch_stat_cp737_gr1,"cp737","el"}, 
 {ch_stat_cp1257_lit1,dbl_ch_stat_cp1257_lit1,"cp1257","lit"}, 
 {ch_stat_cp1257_lat1,dbl_ch_stat_cp1257_lat1,"cp1257","lat"}, 
 {ch_stat_cp1250_sr1,dbl_ch_stat_cp1250_sr1,"cp1250","sr"},
--- crengine/Tools/langstat/langstat.cpp.orig	2021-02-01 20:29:51 UTC
+++ crengine/Tools/langstat/langstat.cpp
@@ -71,8 +71,8 @@ void MakeStats()
    MakeStatsForFile( BASE_DIR "es2-utf8.txt", "utf8", "ee", 2, f, list );
    MakeStatsForFile( BASE_DIR "ee3-cp775.txt", "cp775", "ee", 1, f, list );
    MakeStatsForFile( BASE_DIR "es3-cp1257.txt", "cp1257", "ee", 2, f, list );
-   MakeStatsForFile( BASE_DIR "gr1-cp1253.txt", "cp1253", "gr", 1, f, list );
-   MakeStatsForFile( BASE_DIR "gr1-cp737.txt", "cp737", "gr", 1, f, list );
+   MakeStatsForFile( BASE_DIR "gr1-cp1253.txt", "cp1253", "el", 1, f, list );
+   MakeStatsForFile( BASE_DIR "gr1-cp737.txt", "cp737", "el", 1, f, list );
    MakeStatsForFile( BASE_DIR "lit1-cp1257.txt", "cp1257", "lit", 1, f, list );
    MakeStatsForFile( BASE_DIR "lat1-cp1257.txt", "cp1257", "lat", 1, f, list );
     
