--- stlport/stl/config/user_config.h	Sun Dec  2 19:53:27 2007
+++ stlport/stl/config/user_config.h	Sun Dec  2 19:53:41 2007
@@ -285,9 +285,9 @@
  * but after link phase and optimization you will only experiment benefit if you use
  * many container with pointer types.
  */
-/*
+
 #define _STLP_USE_PTR_SPECIALIZATIONS 1
-*/
+
 
 /*
  * To achieve many different optimizations within the template implementations STLport
