--- stlport/stl/config/user_config.h	2008-07-30 22:57:38.719239317 +0200
+++ stlport/stl/config/user_config.h	2008-07-30 22:58:06.413639084 +0200
@@ -298,9 +298,9 @@
  * to use this feature at STLport built time you will have to define the
  * STLP_BUILD_BOOST_PATH enrironment variable with the value of the boost library path.
  */
-/*
+
 #define _STLP_USE_BOOST_SUPPORT 1
-*/
+
 
 /*==========================================================*/
 
