--- Makefile
+++ Makefile
@@ -241,9 +241,11 @@
   ifeq ($(OS_VENDOR), Apple)
     OTHER_LDLIBS += -liconv
   else
-    ifneq ($(OS_NAME), netbsd)
-      CPPFLAGS += -I$(PACKAGE_PATH)/include
-      OTHER_LDLIBS += -L$(PACKAGE_PATH)/lib -liconv
+    ifneq ($(OS_NAME), freebsd)
+      ifneq ($(OS_NAME), netbsd)
+        CPPFLAGS += -I$(PACKAGE_PATH)/include
+        OTHER_LDLIBS += -L$(PACKAGE_PATH)/lib -liconv
+      endif
     endif
   endif
 endif
--- jdk/make/java/instrument/Makefile
+++ jdk/make/java/instrument/Makefile
@@ -112,10 +112,12 @@
     LDFLAGS += -Wl,--no-whole-archive
   endif
 
-  ICONV_PATH = $(PACKAGE_PATH)
+  ifneq ($(OS_VENDOR), FreeBSD)
+    ICONV_PATH = $(PACKAGE_PATH)
 # Use CPPFLAGS instead of OTHER_INCLUDES to force this last
-  CPPFLAGS += -I$(ICONV_PATH)/include
-  OTHER_LDLIBS += -L$(ICONV_PATH)/lib -liconv
+    CPPFLAGS += -I$(ICONV_PATH)/include
+    OTHER_LDLIBS += -L$(ICONV_PATH)/lib -liconv
+  endif
 else
   LDFLAGS += -L $(LIBDIR)/$(LIBARCH)/jli
   OTHER_LDLIBS += -ljli
--- jdk/make/java/npt/Makefile
+++ jdk/make/java/npt/Makefile
@@ -75,9 +75,11 @@
   ifeq ($(OS_VENDOR), Apple)
     OTHER_LDLIBS += -liconv
   else
-    ifneq ($(OS_NAME), netbsd)
-      CPPFLAGS += -I$(PACKAGE_PATH)/include
-      OTHER_LDLIBS += -L$(PACKAGE_PATH)/lib -liconv
+    ifneq ($(OS_NAME), freebsd)
+      ifneq ($(OS_NAME), netbsd)
+        CPPFLAGS += -I$(PACKAGE_PATH)/include
+        OTHER_LDLIBS += -L$(PACKAGE_PATH)/lib -liconv
+      endif
     endif
   endif
 endif
--- jdk/make/sun/splashscreen/Makefile
+++ jdk/make/sun/splashscreen/Makefile
@@ -74,9 +74,13 @@
     CFLAGS += -DPNG_NO_MMX_CODE
     ICONV_PATH = $(PACKAGE_PATH)
     CPPFLAGS += -I$(OPENWIN_HOME)/include \
-                -I$(OPENWIN_HOME)/include/X11/extensions \
-                -I$(ICONV_PATH)/include
-    OTHER_LDLIBS += -L$(OPENWIN_LIB) -lX11 -lXext -L$(ICONV_PATH)/lib -liconv $(LIBM) -pthread
+                -I$(OPENWIN_HOME)/include/X11/extensions
+    OTHER_LDLIBS += -L$(OPENWIN_LIB) -lX11 -lXext $(LIBM)
+    ifneq ($(OS_NAME), freebsd)
+      CPPFLAGS += -I$(ICONV_PATH)/include
+      OTHER_LDLIBS += -L$(ICONV_PATH)/lib -liconv
+    endif
+    OTHER_LDLIBS += -pthread
   else
     CPPFLAGS += -I$(OPENWIN_HOME)/include -I$(OPENWIN_HOME)/include/X11/extensions
     OTHER_LDLIBS += -L$(OPENWIN_LIB) -lX11 -lXext $(LIBM) -lpthread
