--- plugins/bittorrent/core.h.orig	2014-07-28 18:35:44 UTC
+++ plugins/bittorrent/core.h
@@ -271,7 +271,7 @@ namespace BitTorrent
 				const QVector<bool>& files = QVector<bool> (),
 				LeechCraft::TaskParameters params = LeechCraft::NoParameters);
 		void KillTask (int);
-		void RemoveTorrent (int, int opt = libtorrent::session::none);
+		void RemoveTorrent (int, int opt = 0);
 		void PauseTorrent (int);
 		void ResumeTorrent (int);
 		void ForceReannounce (int);
--- plugins/bittorrent/core.cpp.orig	2014-07-28 18:35:44 UTC
+++ plugins/bittorrent/core.cpp
@@ -230,9 +230,11 @@ namespace BitTorrent
 			setLoggingSettings ();
 			tcpPortRangeChanged ();
 
+#if LIBTORRENT_VERSION_NUM < 010000
 			if (XmlSettingsManager::Instance ()->
 					property ("EnableMetadata").toBool ())
 				Session_->add_extension (&libtorrent::create_metadata_plugin);
+#endif
 			if (XmlSettingsManager::Instance ()->
 					property ("EnablePEX").toBool ())
 				Session_->add_extension (&libtorrent::create_ut_pex_plugin);
@@ -1932,10 +1934,17 @@ namespace BitTorrent
 			atp.paused = pause;
 			atp.duplicate_is_error = true;
 #endif
+
+#if LIBTORRENT_VERSION_NUM >= 010000
+			std::copy (resumeData.constData (),
+					resumeData.constData () + resumeData.size (),
+					std::back_inserter (atp.resume_data));
+#else
 			atp.resume_data = new std::vector<char>;
 			std::copy (resumeData.constData (),
 					resumeData.constData () + resumeData.size (),
 					std::back_inserter (*atp.resume_data));
+#endif
 
 			handle = Session_->add_torrent (atp);
 			if (XmlSettingsManager::Instance ()->property ("ResolveCountries").toBool ())
--- plugins/bittorrent/torrentplugin.cpp.orig	2014-07-28 18:35:44 UTC
+++ plugins/bittorrent/torrentplugin.cpp
@@ -555,7 +555,7 @@ namespace LeechCraft
 
 				confirm.exec ();
 
-				int roptions = libtorrent::session::none;
+				int roptions = 0;
 				if (confirm.clickedButton () == deleteTorrentsAndFilesButton)
 					roptions |= libtorrent::session::delete_files;
 				else if (confirm.clickedButton () == deleteTorrentsButton)
--- plugins/bittorrent/torrenttab.cpp.orig	2014-07-28 18:35:44 UTC
+++ plugins/bittorrent/torrenttab.cpp
@@ -450,7 +450,7 @@ namespace BitTorrent
 
 		confirm.exec ();
 
-		int roptions = libtorrent::session::none;
+		int roptions = 0;
 		if (confirm.clickedButton () == deleteTorrentsAndFilesButton)
 			roptions |= libtorrent::session::delete_files;
 		else if (confirm.clickedButton () == deleteTorrentsButton)
