--- scrotwm.c.orig	2011-07-01 12:27:56.000000000 +0200
+++ scrotwm.c	2011-07-01 12:29:49.000000000 +0200
@@ -3891,14 +3891,16 @@
 					" -nb $bar_color"
 					" -nf $bar_font_color"
 					" -sb $bar_border"
-					" -sf $bar_color",	0);
+					" -sf $bar_color"
+					" -b",	0);
 	setconfspawn("uniconify",	"dmenu"
 					" -i"
 					" -fn $bar_font"
 					" -nb $bar_color"
 					" -nf $bar_font_color"
 					" -sb $bar_border"
-					" -sf $bar_color",	0);
+					" -sf $bar_color"
+					" -b",	0);
 }
 
 /* key bindings */
