--- scrotwm.c	13 Aug 2011 20:26:02 -0000
+++ scrotwm.c	13 Aug 2011 22:56:08 -0000
@@ -3984,14 +3984,16 @@
 					" -nb $bar_color"
 					" -nf $bar_font_color"
 					" -sb $bar_border"
-					" -sf $bar_color",	0);
+					" -sf $bar_color"
+					" -b",	0);
 	setconfspawn("uniconify",	"dmenu"
 					" -i"
 					" -fn $bar_font"
 					" -nb $bar_color"
 					" -nf $bar_font_color"
 					" -sb $bar_border"
-					" -sf $bar_color",	0);
+					" -sf $bar_color"
+					" -b",	0);
 }
 
 /* key bindings */
