--- scrotwm.c.orig	2009-06-29 03:46:08.000000000 +0200
+++ scrotwm.c	2009-06-28 10:07:24.000000000 +0200
@@ -199,7 +200,7 @@
 char			*spawn_lock[] = { "xlock", NULL };
 char			*spawn_initscr[] = { "initscreen.sh", NULL };
 char			*spawn_menu[] = { "dmenu_run", "-fn", NULL, "-nb", NULL,
-			    "-nf", NULL, "-sb", NULL, "-sf", NULL, NULL };
+			    "-nf", NULL, "-sb", NULL, "-sf", NULL, "-b", NULL };
 
 #define SWM_MENU_FN	(2)
 #define SWM_MENU_NB	(4)
