From 194ebeb4eb2efff32864452a52c2edfb4915b89a Mon Sep 17 00:00:00 2001
From: Alexander Nusov <alexander.nusov@nfvexpress.com>
Date: Sun, 22 Jan 2017 13:31:51 +0300
Subject: [PATCH] fixed floating IP list retrieving

---
 openstack_dashboard/api/nova.py | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/openstack_dashboard/api/nova.py b/openstack_dashboard/api/nova.py
index 4a82cbc11..8a3936bad 100644
--- a/openstack_dashboard/api/nova.py
+++ b/openstack_dashboard/api/nova.py
@@ -415,8 +415,7 @@ class FloatingIpManager(network_base.FloatingIpManager):
 
     def list(self, all_tenants=False):
         return [FloatingIp(fip) for fip in
-                self.client.floating_ips.list(
-                    all_tenants=all_tenants)]
+                self.client.floating_ips.list()]
 
     def get(self, floating_ip_id):
         return FloatingIp(self.client.floating_ips.get(floating_ip_id))
-- 
2.11.0

