Index: semantic-idle.el
===================================================================
RCS file: /cvsroot/cedet/cedet/semantic/semantic-idle.el,v
retrieving revision 1.33
retrieving revision 1.35
diff -u -r1.33 -r1.35
--- semantic/semantic-idle.el	30 Jun 2005 01:31:20 -0000	1.33
+++ semantic/semantic-idle.el	12 Sep 2006 01:16:12 -0000	1.35
@@ -1,10 +1,10 @@
 ;;; semantic-idle.el --- Schedule parsing tasks in idle time
 
-;;; Copyright (C) 2003, 2004, 2005 Eric M. Ludlam
+;;; Copyright (C) 2003, 2004, 2005, 2006 Eric M. Ludlam
 
 ;; Author: Eric M. Ludlam <zappo@gnu.org>
 ;; Keywords: syntax
-;; X-RCS: $Id: semantic-idle.el,v 1.33 2005/06/30 01:31:20 zappo Exp $
+;; X-RCS: $Id: semantic-idle.el,v 1.35 2006/09/12 01:16:12 zappo Exp $
 
 ;; This file is not part of GNU Emacs.
 
@@ -20,8 +20,8 @@
 
 ;; You should have received a copy of the GNU General Public License
 ;; along with GNU Emacs; see the file COPYING.  If not, write to the
-;; Free Software Foundation, Inc., 59 Temple Place - Suite 330,
-;; Boston, MA 02111-1307, USA.
+;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+;; Boston, MA 02110-1301, USA.
 
 ;;; Commentary:
 ;;
@@ -286,12 +286,9 @@
   (when (zerop (recursion-depth))
     (unwind-protect
         (semantic-safe "idle error: %S"
-          ;; Disable the auto parse timer while re-parsing
-          (semantic-idle-scheduler-kill-timer)
           ;; Handle re-parsing and other scheduled services
           (save-match-data (semantic-idle-core-handler)))
-      ;; Enable again the auto parse timer
-      (semantic-idle-scheduler-setup-timer))))
+      )))
 
 ;;; REPARSING
 ;;
