--- Makefile.orig	Wed Feb  4 16:39:44 1998
+++ Makefile	Wed Feb  4 16:42:02 1998
@@ -13,11 +13,11 @@
 COLUMNS		= 80
 
 BINDIR		= /usr/local/bin
-MANDIR		= /usr/local/man/man1
+MANDIR		= /usr/local/man/ja/cat1
 MANEXT		= .1
 
-NEWSSPOOL	= /usr/spool/news
-NEWSLIB		= /usr/lib/news
+NEWSSPOOL	= /var/spool/gnspool
+NEWSLIB		= /usr/local/news/lib
 
 NEWSRC		= .newsrc
 SIGNATURE	= .signature
@@ -30,7 +30,7 @@
 MAILER		= /usr/lib/sendmail -t
 
 CC		= cc
-OPTCFLAGS	= -O
+OPTCFLAGS	= -UHAS_MALLOC_H
 OPTGNOBJS	= 
 OPTGNSPOOLOBJS	= 
 OPTLFLAGS	= 
@@ -38,7 +38,9 @@
 
 SHELL		= /bin/sh
 DEFINES		= -DUNIX
-HEADER_DEFINES	=
+HEADER_DEFINES = -DHAS_STRING_H -DHAS_STDLIB_H -DHAS_MALLOC_H -DHAS_UNISTD_H \
+	-DHAS_NETDB_H -DHAS_SYS_IOCTL_H -DHAS_SYS_SOCKET_H -DHAS_NETINET_IN_H \
+	-DHAS_ARPA_INET_H
 CFLAGS		= $(DEFINES) $(HEADER_DEFINES) $(OPTCFLAGS)
 LFLAGS		= $(OPTLFLAGS)
 LIBS		= $(OPTLIBS)
