--- Source/NSFileManager.m.orig	Sat Jul 10 05:19:56 2004
+++ Source/NSFileManager.m	Tue Sep 14 16:56:38 2004
@@ -567,6 +567,12 @@
   NSDictionary	*needChown = nil;
 #endif
 
+  if (path == nil) // don't try to do anything when path is nil
+    {
+      ASSIGN(_lastError, @"Could not create directory - name is nil");
+      return NO;
+    }
+
   /* This is consitent with MacOSX - just return NO for an invalid path. */
   if ([path length] == 0)
     return NO;
@@ -599,7 +605,6 @@
     }
 
 #else
-
   /*
    * If there is no file owner specified, and we are running setuid to
    * root, then we assume we need to change ownership to correct user.
@@ -693,6 +698,12 @@
   while (cur < len);
 
 #endif /* !MINGW */
+
+  // dont change when we have no attibutes
+  if (attributes == nil)
+    {
+      return YES;
+    }
 
   // change attributes of last directory
   if ([attributes count] == 0)
