--- binsrc/cached_resources/res_to_c.awk.orig	2011-10-26 20:53:59.000000000 +0700
+++ binsrc/cached_resources/res_to_c.awk	2012-06-16 16:36:20.000000000 +0700
@@ -30,13 +30,7 @@
 	}
 	{
 	  fun = $0
-	  # was: gsub ( /\\/, "\\\\", fun)
-	  q = "\\\\"
-	  if (PROCINFO["version"] ~ /^4/)
-	    gsub ( q, q q, fun)
-	  else
-	    gsub ( q, q, fun)
-
+          gsub ( "\\\\", "&&", fun)
 	  gsub ( /"/, "\\\"", fun)
 	  gsub ( /\$/, "\\044", fun)
 	  gsub ( /.*/, "\"&\\n\",", fun)
--- libsrc/Wi/sql_to_c.awk.orig	2011-10-26 20:54:20.000000000 +0700
+++ libsrc/Wi/sql_to_c.awk	2012-06-16 16:41:30.000000000 +0700
@@ -343,13 +343,7 @@
 
                 # does escape the symbols
 		fun = $0
-		q = "\\\\"
-		if (PROCINFO["version"] ~ /^4/)
-		    gsub ( q, q q, fun)
-		else
-		    gsub ( q, q, fun)
-
-	        #WAS: gsub ( /\\/, "\\\\", fun)
+		gsub ( "\\\\", "&&", fun)
 
 		# remove whitespace except when there is just a semicolon
 		if ((in_xsl_mode == 0) && (in_xsd_mode == 0))
