--- plugins/check_dig.c.orig	2015-07-30 21:40:06 UTC
+++ plugins/check_dig.c
@@ -94,8 +94,8 @@ main (int argc, char **argv)
   timeout_interval_dig = ceil((double) timeout_interval / (double) number_tries);
 
   /* get the command to run */
-  xasprintf (&command_line, "%s %s %s -p %d @%s %s %s +tries=%d +time=%d",
-  	PATH_TO_DIG, dig_args, query_transport, server_port, dns_server, query_address, record_type, number_tries, timeout_interval_dig);
+  xasprintf (&command_line, "%s %s %s -p %d @%s %s %s",
+  	PATH_TO_DIG, dig_args, query_transport, server_port, dns_server, query_address, record_type);
 
   alarm (timeout_interval);
   gettimeofday (&tv, NULL);
