
$FreeBSD$

--- tools/clang/lib/Basic/Targets.cpp.orig
+++ tools/clang/lib/Basic/Targets.cpp
@@ -3957,6 +3957,9 @@
     PointerWidth = PointerAlign = 64;
     LongDoubleWidth = LongDoubleAlign = 128;
     LongDoubleFormat = &llvm::APFloat::IEEEquad;
+    // FIXME: Ugly ABI-Breaking hack until the back end supports 128-bit floats
+    LongDoubleWidth = LongDoubleAlign = 64;
+    LongDoubleFormat = &llvm::APFloat::IEEEdouble;
     SuitableAlign = 128;
   }
   virtual bool setABI(const std::string &Name) {
