--- kicker/taskbar/taskbar.cpp	(revision 570839)
+++ kicker/taskbar/taskbar.cpp	(revision 570840)
@@ -177,14 +177,14 @@ QSize TaskBar::sizeHint( KPanelExtension
 
         if (containerCount() == 0)
         {
-            actualMax = maxSize.height();
+            actualMax = minButtonHeight;
         }
 
         if (actualMax > maxSize.height())
         {
             return maxSize;
         }
-        return QSize( maxSize.width(), maxSize.height() );
+        return QSize( maxSize.width(), actualMax );
     }
     else
     {
