------------------------------------------------------------------------
r621124 | mueller | 2007-01-08 11:33:10 +0100 (Mon, 08 Jan 2007) | 2 lines

add fix for CVE-2006-6811

------------------------------------------------------------------------
Index: ksirc/iocontroller.cpp
===================================================================
--- ksirc/iocontroller.cpp	(revision 621123)
+++ ksirc/iocontroller.cpp	(revision 621124)
@@ -266,13 +266,13 @@ void KSircIOController::stdout_read(KPro
         if(ksopts->autoCreateWin == TRUE && line[0] != '`' && line[1] != '#' && line[1] != '&' && noticeCreate) {
 	    //kdDebug(5008) << "Creating window for: " << qsname << " because of: " << line.data() << endl;
 	    ksircproc->new_toplevel(KSircChannel(ksircproc->serverName(), qsname));
-	    assert(ksircproc->TopList[qsname] != 0x0);
         }
-        else{
+        if (!ksircproc->TopList[qsname]) {
             qsname = "!default";
             if(line[0] == '`')
                 qsname = "!discard";
         }
+        assert(ksircproc->TopList[qsname]);
     }
 
     ksircproc->TopList[qsname]->sirc_receive(line);
