--- src/gui/painting/qprintengine_pdf_p.h
+++ src/gui/painting/qprintengine_pdf_p.h
@@ -148,7 +148,11 @@ private:
     void writePage();
 
     int addXrefEntry(int object, bool printostr = true);
-    void xprintf(const char* fmt, ...);
+    void xprintf(const char* fmt, ...)
+#if defined(Q_CC_GNU) && !defined(__INSURE__)
+        __attribute__ ((format (printf, 2, 3)))
+#endif
+        ;
     inline void write(const QByteArray &data) {
         stream->writeRawData(data.constData(), data.size());
         streampos += data.size();
--- src/gui/painting/qprintengine_pdf.cpp
+++ src/gui/painting/qprintengine_pdf.cpp
@@ -386,9 +386,8 @@ int QPdfEnginePrivate::addConstantAlphaO
         object = addXrefEntry(-1);
         QByteArray alphaDef;
         QPdf::ByteStream s(&alphaDef);
-        s << "<< /ca " << (alpha/qreal(255.)) << ">>\n";
-        xprintf(alphaDef.constData());
-        xprintf("endobj\n");
+        s << "<< /ca " << (alpha/qreal(255.)) << ">>";
+        xprintf("%s\nendobj\n", alphaDef.constData());
     }
     currentPage->graphicStates.append(object);
     return object;