--- cmd/platlibs.mk.orig	2009-08-31 17:56:52.000000000 +0200
+++ cmd/platlibs.mk	2009-08-31 17:53:35.000000000 +0200
@@ -80,7 +80,7 @@
 endif
 endif
 
-SQLITE=-lsqlite3
+SQLITE=-lnsssqlite3
 
 ifdef NSS_DISABLE_DBM
 DBMLIB = $(NULL)
@@ -129,7 +129,7 @@
 	$(DIST)/lib/$(LIB_PREFIX)nssb.$(LIB_SUFFIX) \
 	$(PKIXLIB) \
 	$(DBMLIB) \
-	$(DIST)/lib/$(LIB_PREFIX)sqlite3.$(LIB_SUFFIX) \
+	$(DIST)/lib/$(LIB_PREFIX)nsssqlite3.$(LIB_SUFFIX) \
 	$(DIST)/lib/$(LIB_PREFIX)nssutil3.$(LIB_SUFFIX) \
 	$(NSPR_LIB_DIR)/$(NSPR31_LIB_PREFIX)plc4.$(LIB_SUFFIX) \
 	$(NSPR_LIB_DIR)/$(NSPR31_LIB_PREFIX)plds4.$(LIB_SUFFIX) \
--- lib/sqlite/manifest.mn.orig	2009-08-31 17:54:32.000000000 +0200
+++ lib/sqlite/manifest.mn	2009-08-31 17:54:44.000000000 +0200
@@ -38,9 +38,9 @@
 
 MODULE = nss
 
-LIBRARY_NAME = sqlite
+LIBRARY_NAME = nsssqlite
 LIBRARY_VERSION = 3
-MAPFILE = $(OBJDIR)/sqlite.def
+MAPFILE = $(OBJDIR)/nsssqlite.def
 DEFINES += -DTHREADSAFE=1
 
 EXPORTS = \
--- lib/softoken/config.mk.orig	2009-08-31 17:55:03.000000000 +0200
+++ lib/softoken/config.mk	2009-08-31 17:55:15.000000000 +0200
@@ -57,7 +57,7 @@
 	-L$(DIST)/lib \
 	-L$(NSSUTIL_LIB_DIR) \
 	-lnssutil3 \
-	-lsqlite3 \
+	-lnsssqlite3 \
 	-L$(NSPR_LIB_DIR) \
 	-lplc4 \
 	-lplds4 \
@@ -66,7 +66,7 @@
 else # ! NS_USE_GCC
 
 EXTRA_SHARED_LIBS += \
-	$(DIST)/lib/sqlite3.lib \
+	$(DIST)/lib/nsssqlite3.lib \
 	$(DIST)/lib/nssutil3.lib \
 	$(NSPR_LIB_DIR)/$(NSPR31_LIB_PREFIX)plc4.lib \
 	$(NSPR_LIB_DIR)/$(NSPR31_LIB_PREFIX)plds4.lib \
@@ -83,7 +83,7 @@
 	-L$(DIST)/lib \
 	-L$(NSSUTIL_LIB_DIR) \
 	-lnssutil3 \
-	-lsqlite3 \
+	-lnsssqlite3 \
 	-L$(NSPR_LIB_DIR) \
 	-lplc4 \
 	-lplds4 \
