commit 7342af72793989b72a2a7ff80bdcaeb81dee4459
Author: Kevin Bowling <kevin.bowling@kev009.com>
Date:   Tue Dec 13 22:37:47 2022 -0700

    Don't use time_t for timestamps

diff --git backend/flipperzero/assetmanifest.cpp backend/flipperzero/assetmanifest.cpp
index d8c7c84..3b599d1 100644
--- backend/flipperzero/assetmanifest.cpp
+++ backend/flipperzero/assetmanifest.cpp
@@ -36,7 +36,7 @@ int AssetManifest::version() const
     return m_version;
 }
 
-time_t AssetManifest::timestamp() const
+qint64 AssetManifest::timestamp() const
 {
     return m_timestamp;
 }
diff --git backend/flipperzero/assetmanifest.h backend/flipperzero/assetmanifest.h
index 1cb922e..8f2724e 100644
--- backend/flipperzero/assetmanifest.h
+++ backend/flipperzero/assetmanifest.h
@@ -25,7 +25,7 @@ public:
     AssetManifest(const QByteArray &text);
 
     int version() const;
-    time_t timestamp() const;
+    qint64 timestamp() const;
     FileNode *tree() const;
 
 private:
@@ -36,7 +36,7 @@ private:
     bool parseDirectory(const QStringList &tokens);
 
     int m_version;
-    time_t m_timestamp;
+    qint64 m_timestamp;
     QSharedPointer<FileNode> m_root;
 };
 
diff --git backend/flipperzero/factoryinfo.cpp backend/flipperzero/factoryinfo.cpp
index 25b29a5..4df59bb 100644
--- backend/flipperzero/factoryinfo.cpp
+++ backend/flipperzero/factoryinfo.cpp
@@ -204,7 +204,7 @@ void FactoryInfo::parseV2(const QByteArray &data)
     m_region = (Region)otp->region;
 }
 
-time_t FactoryInfo::date() const
+qint64 FactoryInfo::date() const
 {
     return m_date;
 }
diff --git backend/flipperzero/factoryinfo.h backend/flipperzero/factoryinfo.h
index 8cdca86..eeaa9c3 100644
--- backend/flipperzero/factoryinfo.h
+++ backend/flipperzero/factoryinfo.h
@@ -23,7 +23,7 @@ public:
     uint8_t target() const;
     uint8_t body() const;
     uint8_t connect() const;
-    time_t date() const;
+    qint64 date() const;
 
     const QString &name() const;
 
@@ -44,7 +44,7 @@ private:
     uint8_t m_target;
     uint8_t m_body;
     uint8_t m_connect;
-    time_t m_date;
+    qint64 m_date;
     QString m_name;
 
     Color m_color;
diff --git backend/flipperzero/radiomanifest.cpp backend/flipperzero/radiomanifest.cpp
index 3088955..943e176 100644
--- backend/flipperzero/radiomanifest.cpp
+++ backend/flipperzero/radiomanifest.cpp
@@ -32,7 +32,7 @@ RadioManifest::Header::Header(const QJsonValue &json)
     }
 
     if(obj.contains(QStringLiteral("timestamp"))) {
-        m_timestamp = (time_t)obj.value(QStringLiteral("timestamp")).toInt();
+        m_timestamp = obj.value(QStringLiteral("timestamp")).toInt();
     } else {
         throw std::runtime_error("Failed to read manifest timestamp");
     }
@@ -43,7 +43,7 @@ int RadioManifest::Header::version() const
     return m_version;
 }
 
-time_t RadioManifest::Header::timestamp() const
+qint64 RadioManifest::Header::timestamp() const
 {
     return m_timestamp;
 }
diff --git backend/flipperzero/radiomanifest.h backend/flipperzero/radiomanifest.h
index f71fed4..04575bb 100644
--- backend/flipperzero/radiomanifest.h
+++ backend/flipperzero/radiomanifest.h
@@ -19,11 +19,11 @@ public:
         Header(const QJsonValue &json);
 
         int version() const;
-        time_t timestamp() const;
+        qint64 timestamp() const;
 
     private:
         int m_version;
-        time_t m_timestamp;
+        qint64 m_timestamp;
     };
 
     class Condition {
