diff -ru lib.old/Net/Server/Mail/ESMTP/8BITMIME.pm lib/Net/Server/Mail/ESMTP/8BITMIME.pm
--- lib.old/Net/Server/Mail/ESMTP/8BITMIME.pm	Tue Jun 25 15:53:21 2002
+++ lib/Net/Server/Mail/ESMTP/8BITMIME.pm	Thu Feb 19 15:27:11 2004
@@ -1,6 +1,5 @@
 package Net::Server::Mail::ESMTP::8BITMIME;
 
-use 5.006;
 use strict;
-use base qw(Net::Server::Mail::ESMTP::Extension);
+BEGIN {require Net::Server::Mail::ESMTP::Extension;@Net::Server::Mail::ESMTP::8BITMIME::ISA = ('Net::Server::Mail::ESMTP::Extension')};
 
diff -ru lib.old/Net/Server/Mail/ESMTP/Extension.pm lib/Net/Server/Mail/ESMTP/Extension.pm
--- lib.old/Net/Server/Mail/ESMTP/Extension.pm	Thu Nov  7 18:53:51 2002
+++ lib/Net/Server/Mail/ESMTP/Extension.pm	Thu Feb 19 15:27:13 2004
@@ -1,6 +1,5 @@
 package Net::Server::Mail::ESMTP::Extension;
 
-use 5.006;
 use strict;
 
 =pod
diff -ru lib.old/Net/Server/Mail/ESMTP/PIPELINING.pm lib/Net/Server/Mail/ESMTP/PIPELINING.pm
--- lib.old/Net/Server/Mail/ESMTP/PIPELINING.pm	Mon Jan  6 12:21:43 2003
+++ lib/Net/Server/Mail/ESMTP/PIPELINING.pm	Thu Feb 19 15:27:16 2004
@@ -1,6 +1,5 @@
 package Net::Server::Mail::ESMTP::PIPELINING;
 
-use 5.006;
 use strict;
-use base 'Net::Server::Mail::ESMTP::Extension';
+BEGIN {require Net::Server::Mail::ESMTP::Extension;@Net::Server::Mail::ESMTP::PIPELINING::ISA = ('Net::Server::Mail::ESMTP::Extension')};
 use constant GROUP_COMMANDS => [qw(RSET MAIL SEND SOML SAML RCPT)];
diff -ru lib.old/Net/Server/Mail/ESMTP.pm lib/Net/Server/Mail/ESMTP.pm
--- lib.old/Net/Server/Mail/ESMTP.pm	Sun Apr  6 17:33:02 2003
+++ lib/Net/Server/Mail/ESMTP.pm	Thu Feb 19 15:32:21 2004
@@ -1,9 +1,8 @@
 package Net::Server::Mail::ESMTP;
 
-use 5.006;
 use strict;
 use Carp;
-use base qw(Net::Server::Mail::SMTP);
+BEGIN {require Net::Server::Mail::SMTP;@Net::Server::Mail::ESMTP::ISA = ('Net::Server::Mail::SMTP')};
 
 =pod
 
diff -ru lib.old/Net/Server/Mail/LMTP.pm lib/Net/Server/Mail/LMTP.pm
--- lib.old/Net/Server/Mail/LMTP.pm	Fri Mar 14 16:25:03 2003
+++ lib/Net/Server/Mail/LMTP.pm	Thu Feb 19 15:26:59 2004
@@ -1,6 +1,5 @@
 package Net::Server::Mail::LMTP;
 
-use 5.006;
 use strict;
-use base qw(Net::Server::Mail::ESMTP);
+BEGIN {require Net::Server::Mail::ESMTP;@Net::Server::Mail::LMTP::ISA = ('Net::Server::Mail::ESMTP')};
 
diff -ru lib.old/Net/Server/Mail/SMTP.pm lib/Net/Server/Mail/SMTP.pm
--- lib.old/Net/Server/Mail/SMTP.pm	Sun Apr  6 17:36:40 2003
+++ lib/Net/Server/Mail/SMTP.pm	Thu Feb 19 15:32:24 2004
@@ -1,8 +1,7 @@
 package Net::Server::Mail::SMTP;
 
-use 5.006;
 use strict;
-use base 'Net::Server::Mail';
+BEGIN {require Net::Server::Mail;@Net::Server::Mail::SMTP::ISA = ('Net::Server::Mail')};
 
 =pod
 
--- lib/Net/Server/Mail.pm.orig	Wed Jan  5 19:50:35 2005
+++ lib/Net/Server/Mail.pm	Mon Jan 24 20:22:04 2005
@@ -1,6 +1,5 @@
 package Net::Server::Mail;
 
-use 5.006;
 use strict;
 use Sys::Hostname;
 use IO::Select;
@@ -502,7 +501,7 @@
         else
         {
             next unless defined;
-            $rv = $self->{process_operation}($self, $_);
+            $rv = $self->{process_operation}->($self, $_);
         }
         # if $rv is defined, we have to close the connection
         return $rv if defined $rv;
@@ -521,7 +520,7 @@
     }
     else
     {
-        return $self->{process_operation}($self, $operation);
+        return $self->{process_operation}->($self, $operation);
     }
 }
 
