diff --git a/src/rgw/rgw_main.cc b/src/rgw/rgw_main.cc
index 44609741ed..ebbbd12c01 100644
--- src/rgw/rgw_main.cc
+++ src/rgw/rgw_main.cc
@@ -565,8 +565,8 @@ int main(int argc, const char **argv)
   // add a watcher to respond to realm configuration changes
   RGWPeriodPusher pusher(store);
   RGWFrontendPauser pauser(fes, implicit_tenant_context, &pusher);
-  std::optional<RGWRealmReloader> reloader(std::in_place, store,
-                                           service_map_meta, &pauser);
+  auto reloader = std::make_unique<RGWRealmReloader>(store,
+                                                    service_map_meta, &pauser);
 
   RGWRealmWatcher realm_watcher(g_ceph_context, store->svc.zone->get_realm());
   realm_watcher.add_watcher(RGWRealmNotify::Reload, *reloader);
