--- src/ref.c.orig	Wed Aug 17 17:27:40 2005
+++ src/ref.c	Wed Aug 17 17:27:53 2005
@@ -263,7 +263,7 @@
 read_new_value (CELLREF row, CELLREF col, char *form, char *val)
 {
   unsigned char *new_bytes;
-  extern double __plinf, __neinf, __nan;
+  extern double __plinf, __neinf, __naN_;
 
   cur_row = row;
   cur_col = col;
@@ -348,7 +348,7 @@
 	  else if (!stricmp (nname, val))
 	    {
 	      SET_TYP (my_cell, TYP_FLT);
-	      my_cell->cell_flt = __nan;
+	      my_cell->cell_flt = __naN_;
 	    }
 	  else
 	    {
