--- src/Makefile.orig	Sat Apr  5 19:43:25 1997
+++ src/Makefile	Sat Apr  5 19:51:26 1997
@@ -124,11 +124,11 @@
 # flags for debugging:
 # CFLAGS = -g -I../include
 
-CFLAGS = -O2 -I../include
-LFLAGS = -L/usr/X11/lib -L$(QTDIR)/lib
+CFLAGS = -pipe -O -I../include
+LFLAGS = -L/usr/X11R6/lib -L$(QTDIR)/lib
 
 # For USER_SOUNDS support, include -I/usr/X11/include
-CXXFLAGS = $(CFLAGS) -I. -I$(QTDIR)/include -I/usr/X11/include
+CXXFLAGS = $(CFLAGS) -I. -I/usr/X11R6/include -I/usr/X11R6/include/X11/qt
 
 # Set the WINSRC, WINOBJ, and WINLIB lines to correspond to your desired
 # combination of windowing systems.  Also set windowing systems in config.h.
@@ -213,7 +213,7 @@
 # IRIX 4.0.x needs -lsun if NIS (YP) is being used for passwd file lookup
 # LIBS = -lsun
 #
-LIBS =
+LIBS = -lgnuregex
 
 # make NetHack
 GAME     = nethack
