Index: kmymoney2/dialogs/kenterscheduledialog.cpp
===================================================================
RCS file: /cvsroot/kmymoney2/kmymoney2/kmymoney2/dialogs/kenterscheduledialog.cpp,v
retrieving revision 1.11.2.1
retrieving revision 1.11.2.2
diff -c -r1.11.2.1 -r1.11.2.2
*** kmymoney2/dialogs/kenterscheduledialog.cpp	31 Jan 2004 12:53:42 -0000	1.11.2.1
--- kmymoney2/dialogs/kenterscheduledialog.cpp	28 Dec 2004 22:41:52 -0000	1.11.2.2
***************
*** 994,1001 ****
  
    if (date < firstDate || date > nextDate)
    {
!     KMessageBox::error(this, i18n("The date must lie in the range %1 to %2").arg(firstDate.addDays(1).toString()).arg(nextDate.toString()));
!     return false;
    }
  
    return true;
--- 994,1005 ----
  
    if (date < firstDate || date > nextDate)
    {
! //    KMessageBox::error(this, i18n("The date must lie in the range %1 to %2").arg(firstDate.addDays(1).toString()).arg(nextDate.toString()));
! 
!     // let the user continue anyway as the calculation is not always correct
!     // proper fix will be available in 0.7/0.9.
!     if (KMessageBox::warningContinueCancel(this, i18n("The date must lie in the range %1 to %2").arg(firstDate.addDays(1).toString()).arg(nextDate.toString())) != KMessageBox::Continue)
!       return false;
    }
  
    return true;
