*** jp-printers/gdevdmpr/dmp_site.ps.orig	Tue May 27 09:57:00 1997
--- jp-printers/gdevdmpr/dmp_site.ps	Sun Jul  6 00:56:56 1997
***************
*** 1,6 ****
  BeginConfig
  %%%%%%%%%%%%%%%%%%%%%% Standard user cusomizations %%%%%%%%%%%%%%%%%%%%%
! /printer      (epsimage.src)
  %               horizontal / vertical
  % /resolution   [ 160.0 180.0 ]
  %               horizontal / vertical
--- 1,6 ----
  BeginConfig
  %%%%%%%%%%%%%%%%%%%%%% Standard user cusomizations %%%%%%%%%%%%%%%%%%%%%
! /printer      (/usr/local/share/ghostscript/5.50vflib/escp_24.src)
  %               horizontal / vertical
  % /resolution   [ 160.0 180.0 ]
  %               horizontal / vertical
