--- gdevepag.mak.orig	Tue Mar  9 09:45:25 1999
+++ gdevepag.mak	Tue Mar  9 09:51:22 1999
@@ -1,6 +1,7 @@
 ### ------- Epson ESC/Page printer device ----------------- ###
-epag_ = gdevepag.$(OBJ)
-gdevepag.$(OBJ): gdevepag.c $(PDEVH)
+epag_ = $(GLOBJ)gdevepag.$(OBJ) $(GLOBJ)gdevprn.$(OBJ)
+$(GLOBJ)gdevepag.$(OBJ): $(GLSRC)gdevepag.c $(PDEVH)
+	$(GLCC) $(GLO_)gdevepag.$(OBJ) $(C_) $(GLSRC)gdevepag.c
 
 epag.dev: $(epag_) page.dev
 	$(SETPDEV) epag $(epag_)
