--- gdevmjc.mak.orig	Fri Jun  4 01:44:35 1999
+++ gdevmjc.mak	Fri Jun  4 01:53:04 1999
@@ -10,19 +10,25 @@
 ##
 MJCSRC = gdevmjc.c mjmtx2.c mjbksep.c mjhtocmy.c mjvtbl.c \
 	mjgrnsp.c mjgrnsp2.c mjespdtc.c mjespdtm.c mjespdty.c mjbarrie.c
+MJCSRC2 = ${MJCSRC:S/^/$(GLOBJ)/g}
 
-mj700v2c_=gdevmjc.$(OBJ) $(HPPCL)
+all:
+	echo $(MJCSRC2)
+	echo $(SETPDEV) mjc180 $(mj700v2c_)
 
-# To make A4 paper the default, change the second line below this to
-gdevmjc.$(OBJ): $(MJCSRC) $(PDEVH) $(gdevpcl_h)
-	$(CCC) -DA4 gdevmjc.c
+mj700v2c_=$(GLOBJ)gdevmjc.$(OBJ) $(HPPCL)
 
-mjc180.dev: $(mj700v2c_)
-	$(SHP)gssetdev mjc180 $(mj700v2c_)
-mjc360.dev: $(mj700v2c_)
-	$(SHP)gssetdev mjc360 $(mj700v2c_)
-mjc720.dev: $(mj700v2c_)
-	$(SHP)gssetdev mjc720 $(mj700v2c_)
-mj500c.dev: $(mj700v2c_)
-	$(SHP)gssetdev mj500c $(mj700v2c_)
+# To make A4 paper the default, change the second line below this to
+#	$(CC) -DA4 -o gdevmjc.o gdevmjc.c
+$(GLOBJ)gdevmjc.$(OBJ): $(GLSRC)gdevmjc.c $(PDEVH) $(gdevpcl_h)
+	$(GLCC) $(GLO_)gdevmjc.$(OBJ) $(C_) $(GLSRC)gdevmjc.c
+	
 
+mjc180.dev: $(mj700v2c_) page.dev
+	$(SETPDEV) mjc180 $(mj700v2c_)
+mjc360.dev: $(mj700v2c_) page.dev
+	$(SETPDEV) mjc360 $(mj700v2c_)
+mjc720.dev: $(mj700v2c_) page.dev
+	$(SETPDEV) mjc720 $(mj700v2c_)
+mj500c.dev: $(mj700v2c_) page.dev
+	$(SETPDEV) mj500c $(mj700v2c_)
