*** mjc.dev.orig	Sat Nov  9 22:51:15 1996
--- mjc.dev	Sat Nov  9 22:53:25 1996
***************
*** 17,27 ****
  	$(CCC) gdevmjc.c
  
  mjc180.dev: $(mj700v2c_)
! 	$(SHP)gssetdev mjc180 $(mj700v2c_)
  mjc360.dev: $(mj700v2c_)
! 	$(SHP)gssetdev mjc360 $(mj700v2c_)
  mjc720.dev: $(mj700v2c_)
! 	$(SHP)gssetdev mjc720 $(mj700v2c_)
  mj500c.dev: $(mj700v2c_)
! 	$(SHP)gssetdev mj500c $(mj700v2c_)
  
--- 17,27 ----
  	$(CCC) gdevmjc.c
  
  mjc180.dev: $(mj700v2c_)
! 	$(SETPDEV) mjc180 $(mj700v2c_)
  mjc360.dev: $(mj700v2c_)
! 	$(SETPDEV) mjc360 $(mj700v2c_)
  mjc720.dev: $(mj700v2c_)
! 	$(SETPDEV) mjc720 $(mj700v2c_)
  mj500c.dev: $(mj700v2c_)
! 	$(SETPDEV) mj500c $(mj700v2c_)
  
