*** mjc.dev.orig	Sun Nov 15 23:30:39 1998
--- mjc.dev	Sun Nov 15 23:43:17 1998
***************
*** 9,27 ****
  #
  ##
  
! mj700v2c_=gdevmjc.$(OBJ) $(HPPCL)
  
  # To make A4 paper the default, change the second line below this to
! #	$(CCC) -DA4 gdevmjc.c
! gdevmjc.$(OBJ): gdevmjc.c $(PDEVH) $(gdevpcl_h)
! 	$(CCC) gdevmjc.c
  
! mjc180.dev: $(mj700v2c_)
! 	$(SHP)gssetdev mjc180 $(mj700v2c_)
! mjc360.dev: $(mj700v2c_)
! 	$(SHP)gssetdev mjc360 $(mj700v2c_)
! mjc720.dev: $(mj700v2c_)
! 	$(SHP)gssetdev mjc720 $(mj700v2c_)
! mj500c.dev: $(mj700v2c_)
! 	$(SHP)gssetdev mj500c $(mj700v2c_)
  
--- 9,28 ----
  #
  ##
  
! mj700v2c_=$(GLOBJ)gdevmjc.$(OBJ) $(HPPCL)
  
  # To make A4 paper the default, change the second line below this to
! #	$(CC) -DA4 -o gdevmjc.o gdevmjc.c
! $(GLOBJ)gdevmjc.$(OBJ): $(GLSRC)gdevmjc.c $(PDEVH) $(gdevpcl_h)
! 	$(GLCC) $(GLO_)gdevmjc.$(OBJ) $(C_) $(GLSRC)gdevmjc.c
! 	
  
! mjc180.dev: $(mj700v2c_) page.dev
! 	$(SETPDEV) mjc180 $(mj700v2c_)
! mjc360.dev: $(mj700v2c_) page.dev
! 	$(SETPDEV) mjc360 $(mj700v2c_)
! mjc720.dev: $(mj700v2c_) page.dev
! 	$(SETPDEV) mjc720 $(mj700v2c_)
! mj500c.dev: $(mj700v2c_) page.dev
! 	$(SETPDEV) mj500c $(mj700v2c_)
  
