--- app/config/core.php.orig	Thu Jan  3 17:07:25 2008
+++ app/config/core.php	Thu Jan  3 17:09:11 2008
@@ -56,7 +56,7 @@
  *
  * And uncomment the App.baseUrl below:
  */
-	//Configure::write('App.baseUrl', env('SCRIPT_NAME'));
+	Configure::write('App.baseUrl', env('SCRIPT_NAME'));
 /**
  * Uncomment the define below to use CakePHP admin routes.
  *
@@ -226,4 +226,4 @@
  * 					);
  */
 	Cache::config('default', array('engine' => 'File'));
-?>
\ No newline at end of file
+?>
