--- a/cc/private/toolchain/freebsd_cc_toolchain_config.bzl	2022-08-28 16:31:59.271838000 +0100
+++ b/cc/private/toolchain/freebsd_cc_toolchain_config.bzl	2022-08-28 16:32:10.096301000 +0100
@@ -159,7 +159,7 @@
             ),
             flag_set(
                 actions = all_cpp_compile_actions + [ACTION_NAMES.lto_backend],
-                flag_groups = [flag_group(flags = ["-std=c++0x"])],
+                flag_groups = [flag_group(flags = ["-std=c++17"])],
             ),
         ],
     )
