--- a/build_config/BUILD.bazel
+++ b/build_config/BUILD.bazel
@@ -43,6 +43,11 @@ config_setting(
     values = {"cpu": "k8"},
 )
 
+config_setting(
+    name = "freebsd_x86_64",
+    values = {"cpu": "freebsd"},
+)
+
 config_setting(
     name = "linux_arm",
     values = {"cpu": "arm"},
@@ -357,6 +362,7 @@ selects.config_setting_group(
     match_any = [
         ":android_x86",
         ":android_x86_64",
+        ":freebsd_x86_64",
         ":ios_x86_64",
         ":linux_k8",
         ":macos_x86_64",
@@ -372,6 +378,7 @@ selects.config_setting_group(
     name = "x86_64",
     match_any = [
         ":android_x86_64",
+        ":freebsd_x86_64",
         ":ios_x86_64",
         ":linux_k8",
         ":macos_x86_64",
