diff --git CMakeLists.txt CMakeLists.txt
index 2deada7c..d0df78b7 100644
--- CMakeLists.txt
+++ CMakeLists.txt
@@ -132,9 +132,13 @@ option(DBUS_DISABLE_ASSERT "Disable assertion checking" OFF)
 option(DBUS_ENABLE_STATS "enable bus daemon usage statistics" OFF)
 option(ENABLE_TRADITIONAL_ACTIVATION "Enable traditional activation (without using systemd)" ON)
 
+find_package(PkgConfig)
+
 if(DBUS_LINUX)
     add_auto_option(ENABLE_SYSTEMD "build with systemd at_console support" AUTO)
-    include(FindPkgConfig)
+    if (NOT PKG_CONFIG_FOUND)
+        message(SEND_ERROR "pkg-config not found, this is required on Linux systems")
+    endif()
     pkg_check_modules(SYSTEMD libsystemd>=209)
     if(NOT SYSTEMD_FOUND)
         pkg_check_modules(SYSTEMD libsystemd-login>=32 libsystemd-daemon>=32 libsystemd-journal>=32)
@@ -652,6 +656,10 @@ if(DBUS_BUILD_TESTS)
     endif()
 endif()
 
+if(APPLE)
+    option(DBUS_ENABLE_LAUNCHD "Use launchd autolaunch" ON)
+endif()
+
 set(DBUS_LIBRARIES dbus-1)
 set(DBUS_INTERNAL_LIBRARIES dbus-internal)
 
diff --git cmake/config.h.cmake cmake/config.h.cmake
index b8d98fb2..12a7b7f9 100644
--- cmake/config.h.cmake
+++ cmake/config.h.cmake
@@ -64,6 +64,7 @@
 #  define DBUS_ENABLE_CHECKS 1
 #endif
 #cmakedefine DBUS_ENABLE_EMBEDDED_TESTS 1
+#cmakedefine DBUS_ENABLE_LAUNCHD 1
 #cmakedefine DBUS_ENABLE_MODULAR_TESTS 1
 #cmakedefine DBUS_USE_OUTPUT_DEBUG_STRING 1
 
diff --git dbus/CMakeLists.txt dbus/CMakeLists.txt
index 42442cce..1d4ee7ea 100644
--- dbus/CMakeLists.txt
+++ dbus/CMakeLists.txt
@@ -65,6 +65,9 @@ if(UNIX)
         dbus-transport-unix.c
         dbus-server-unix.c
     )
+    if(DBUS_ENABLE_LAUNCHD)
+        set(DBUS_LIB_SOURCES ${DBUS_LIB_SOURCES} dbus-server-launchd.c)
+    endif()
 else(UNIX)
     set(DBUS_LIB_SOURCES ${DBUS_LIB_SOURCES}
         dbus-transport-win.c
diff --git dbus/dbus-internals.h dbus/dbus-internals.h
index c7967d24..dd694569 100644
--- dbus/dbus-internals.h
+++ dbus/dbus-internals.h
@@ -277,11 +277,13 @@ _dbus_assert_error_xor_bool (const DBusError *error,
  */
 
 #define _DBUS_ALIGN_VALUE(this, boundary) \
-  (( ((uintptr_t)(this)) + (((uintptr_t)(boundary)) -1)) & (~(((uintptr_t)(boundary))-1)))
+  (( ((uintptr_t)(this)) + (((size_t)(boundary)) -1)) & (~(((size_t)(boundary))-1)))
 
 #define _DBUS_ALIGN_ADDRESS(this, boundary) \
   ((void*)_DBUS_ALIGN_VALUE(this, boundary))
 
+#define _DBUS_IS_ALIGNED(this, boundary) \
+  ((size_t)(this) % (boundary) == 0)
 
 DBUS_PRIVATE_EXPORT
 char*       _dbus_strdup                (const char  *str);
diff --git dbus/dbus-marshal-recursive.h dbus/dbus-marshal-recursive.h
index c3b96819..1b313a4b 100644
--- dbus/dbus-marshal-recursive.h
+++ dbus/dbus-marshal-recursive.h
@@ -44,10 +44,10 @@ struct DBusTypeReader
                                  * where we don't have another way to tell
                                  */
   dbus_uint32_t array_len_offset : 3; /**< bytes back from start_pos that len ends */
-  const DBusString *type_str;   /**< string containing signature of block */
   int type_pos;                 /**< current position in signature */
-  const DBusString *value_str;  /**< string containing values of block */
   int value_pos;                /**< current position in values */
+  const DBusString *type_str;   /**< string containing signature of block */
+  const DBusString *value_str;  /**< string containing values of block */
 
   const DBusTypeReaderClass *klass; /**< the vtable for the reader */
   union
@@ -71,10 +71,10 @@ struct DBusTypeWriter
 
   dbus_uint32_t enabled : 1; /**< whether to write values */
 
-  DBusString *type_str; /**< where to write typecodes (or read type expectations) */
   int type_pos;         /**< current pos in type_str */
-  DBusString *value_str; /**< where to write values */
   int value_pos;         /**< next position to write */
+  DBusString *type_str; /**< where to write typecodes (or read type expectations) */
+  DBusString *value_str; /**< where to write values */
 
   union
   {
diff --git dbus/dbus-memory.c dbus/dbus-memory.c
index ce177bea..7c311b39 100644
--- dbus/dbus-memory.c
+++ dbus/dbus-memory.c
@@ -117,7 +117,7 @@ static DBusAtomic n_blocks_outstanding = {0};
 /** size of the information about the block stored in guard mode */
 #define GUARD_INFO_SIZE 8
 /** size of the GUARD_VALUE-filled padding after the header info  */
-#define GUARD_START_PAD 16
+#define GUARD_START_PAD 24
 /** size of the GUARD_VALUE-filled padding at the end of the block */
 #define GUARD_END_PAD 16
 /** size of stuff at start of block */
@@ -125,6 +125,8 @@ static DBusAtomic n_blocks_outstanding = {0};
 /** total extra size over the requested allocation for guard stuff */
 #define GUARD_EXTRA_SIZE (GUARD_START_OFFSET + GUARD_END_PAD)
 
+_DBUS_STATIC_ASSERT(GUARD_START_OFFSET % _DBUS_ALIGNOF(max_align_t) == 0);
+
 static void
 _dbus_initialize_malloc_debug (void)
 {
@@ -388,6 +390,8 @@ check_guards (void       *free_block,
     }
 }
 
+_DBUS_STATIC_ASSERT(GUARD_START_OFFSET + GUARD_END_PAD == GUARD_EXTRA_SIZE);
+
 static void*
 set_guards (void       *real_block,
             size_t      requested_bytes,
 
diff --git dbus/dbus-mempool.c dbus/dbus-mempool.c
index 737cf201..c9db1e59 100644
--- dbus/dbus-mempool.c
+++ dbus/dbus-mempool.c
@@ -65,12 +65,6 @@ struct DBusFreedElement
   DBusFreedElement *next; /**< next element of the free list */
 };
 
-/**
- * The dummy size of the variable-length "elements"
- * field in DBusMemBlock
- */
-#define ELEMENT_PADDING 4
-
 /**
  * Typedef for DBusMemBlock so the struct can recursively
  * point to itself.
@@ -87,13 +81,12 @@ struct DBusMemBlock
                         *   only saved so we can free all the blocks
                         *   when we free the mem pool.
                         */
-
-  /* this is a long so that "elements" is aligned */
   long used_so_far;     /**< bytes of this block already allocated as elements. */
-  
-  unsigned char elements[ELEMENT_PADDING]; /**< the block data, actually allocated to required size */
+  _Alignas(_DBUS_ALIGNOF(max_align_t)) unsigned char elements[]; /**< the block data, actually allocated to required size */
 };
 
+_DBUS_STATIC_ASSERT(offsetof(struct DBusMemBlock, elements) % _DBUS_ALIGNOF(max_align_t) == 0);
+
 /**
  * Internals fields of DBusMemPool
  */
@@ -102,10 +95,10 @@ struct DBusMemPool
   int element_size;                /**< size of a single object in the pool */
   int block_size;                  /**< size of most recently allocated block */
   unsigned int zero_elements : 1;  /**< whether to zero-init allocated elements */
+  int allocated_elements;          /**< Count of outstanding allocated elements */
 
   DBusFreedElement *free_elements; /**< a free list of elements to recycle */
   DBusMemBlock *blocks;            /**< blocks of memory from malloc() */
-  int allocated_elements;          /**< Count of outstanding allocated elements */
 };
 
 /** @} */
@@ -161,7 +154,7 @@ _dbus_mem_pool_new (int element_size,
   /* align the element size to a pointer boundary so we won't get bus
    * errors under other architectures.  
    */
-  pool->element_size = _DBUS_ALIGN_VALUE (element_size, sizeof (void *));
+  pool->element_size = _DBUS_ALIGN_VALUE (element_size, _DBUS_ALIGNOF(max_align_t));
 
   pool->zero_elements = zero_elements != FALSE;
 
@@ -217,6 +210,7 @@ _dbus_mem_pool_free (DBusMemPool *pool)
 void*
 _dbus_mem_pool_alloc (DBusMemPool *pool)
 {
+  // FIXME: set bounds here
 #ifdef DBUS_ENABLE_EMBEDDED_TESTS
   if (_dbus_disable_mem_pools ())
     {
@@ -230,8 +224,7 @@ _dbus_mem_pool_alloc (DBusMemPool *pool)
        * should vanish)
        */
       
-      alloc_size = sizeof (DBusMemBlock) - ELEMENT_PADDING +
-        pool->element_size;
+      alloc_size = sizeof (DBusMemBlock) + pool->element_size;
       
       if (pool->zero_elements)
         block = dbus_malloc0 (alloc_size);
@@ -246,6 +239,7 @@ _dbus_mem_pool_alloc (DBusMemPool *pool)
 
           VALGRIND_MEMPOOL_ALLOC (pool, (void *) &block->elements[0],
               pool->element_size);
+	  _dbus_assert(_DBUS_IS_ALIGNED(&block->elements[0], _DBUS_ALIGNOF(max_align_t)));
           return (void*) &block->elements[0];
         }
       else
@@ -271,7 +265,7 @@ _dbus_mem_pool_alloc (DBusMemPool *pool)
             memset (element, '\0', pool->element_size);
 
           pool->allocated_elements += 1;
-
+	  _dbus_assert(_DBUS_IS_ALIGNED(element, _DBUS_ALIGNOF(max_align_t)));
           return element;
         }
       else
@@ -296,7 +290,7 @@ _dbus_mem_pool_alloc (DBusMemPool *pool)
                                  pool->element_size) == 0);
                 }
 
-              alloc_size = sizeof (DBusMemBlock) - ELEMENT_PADDING + pool->block_size;
+              alloc_size = sizeof (DBusMemBlock) + pool->block_size;
 
 #ifdef DBUS_ENABLE_EMBEDDED_TESTS
               /* We save/restore the counter, so that memory pools won't
@@ -313,6 +307,7 @@ _dbus_mem_pool_alloc (DBusMemPool *pool)
                 block = dbus_malloc0 (alloc_size);
               else
                 block = dbus_malloc (alloc_size);
+	      _dbus_assert(_DBUS_IS_ALIGNED(block, _DBUS_ALIGNOF(max_align_t)));
 
 #ifdef DBUS_ENABLE_EMBEDDED_TESTS
               _dbus_set_fail_alloc_counter (saved_counter);
@@ -326,14 +321,19 @@ _dbus_mem_pool_alloc (DBusMemPool *pool)
               block->next = pool->blocks;
               pool->blocks = block;          
             }
-      
+
+	  _dbus_assert(_DBUS_IS_ALIGNED(pool->blocks->used_so_far, _DBUS_ALIGNOF(max_align_t)));
+
           element = &pool->blocks->elements[pool->blocks->used_so_far];
-          
+	  _dbus_assert(_DBUS_IS_ALIGNED(pool->blocks->elements, _DBUS_ALIGNOF(max_align_t)));
+
           pool->blocks->used_so_far += pool->element_size;
+	  _dbus_assert(_DBUS_IS_ALIGNED(pool->blocks->used_so_far, _DBUS_ALIGNOF(max_align_t)));
 
           pool->allocated_elements += 1;
 
           VALGRIND_MEMPOOL_ALLOC (pool, element, pool->element_size);
+	  _dbus_assert(_DBUS_IS_ALIGNED(element, _DBUS_ALIGNOF(max_align_t)));
           return element;
         }
     }
diff --git dbus/dbus-message.c dbus/dbus-message.c
index 19a43750..0ae09b13 100644
--- dbus/dbus-message.c
+++ dbus/dbus-message.c
@@ -136,6 +136,7 @@ struct DBusMessageRealIter
   } u; /**< the type writer or reader that does all the work */
 };
 
+#ifndef __CHERI_PURE_CAPABILITY__ /* No backwards ABI compat required */
 /**
  * Layout of a DBusMessageIter on the stack in dbus 1.10.0. This is no
  * longer used, but for ABI compatibility we need to assert that the
@@ -158,6 +159,7 @@ typedef struct
   int pad2;
   void *pad3;
 } DBusMessageIter_1_10_0;
+#endif
 
 static void
 get_const_signature (DBusHeader        *header,
@@ -2069,16 +2071,21 @@ _dbus_message_iter_init_common (DBusMessage         *message,
   _DBUS_STATIC_ASSERT (sizeof (DBusMessageRealIter) <= sizeof (DBusMessageIter));
   _DBUS_STATIC_ASSERT (_DBUS_ALIGNOF (DBusMessageRealIter) <=
       _DBUS_ALIGNOF (DBusMessageIter));
+#ifndef __CHERI_PURE_CAPABILITY__
   /* A failure of these two assertions would indicate that we've broken
    * ABI on this platform since 1.10.0. */
   _DBUS_STATIC_ASSERT (sizeof (DBusMessageIter_1_10_0) ==
       sizeof (DBusMessageIter));
   _DBUS_STATIC_ASSERT (_DBUS_ALIGNOF (DBusMessageIter_1_10_0) ==
       _DBUS_ALIGNOF (DBusMessageIter));
+#endif
   /* If this static assertion fails, it means the DBusMessageIter struct
    * is not "packed", which might result in "iter = other_iter" not copying
    * every byte. */
   _DBUS_STATIC_ASSERT (sizeof (DBusMessageIter) ==
+#if __SIZEOF_POINTER__ == 16
+      2 * sizeof(int) +
+#endif
       4 * sizeof (void *) + sizeof (dbus_uint32_t) + 9 * sizeof (int));
 
   /* Since the iterator will read or write who-knows-what from the
diff --git dbus/dbus-message.h dbus/dbus-message.h
index da2f2d9d..a3455e70 100644
--- dbus/dbus-message.h
+++ dbus/dbus-message.h
@@ -71,7 +71,7 @@ struct DBusMessageIter
   int dummy9;           /**< Don't use this */
   int dummy10;          /**< Don't use this */
   int dummy11;          /**< Don't use this */
-  int pad1;             /**< Don't use this */
+  int pad1[sizeof(void*) > 8 ? (sizeof(void*) - sizeof(int)) / sizeof(int) : 1]; /**< Don't use this */
   void *pad2;           /**< Don't use this */
   void *pad3;           /**< Don't use this */
 };
@@ -93,7 +93,7 @@ struct DBusMessageIter
   0, /* dummy9 */ \
   0, /* dummy10 */ \
   0, /* dummy11 */ \
-  0, /* pad1 */ \
+  {0}, /* pad1 */ \
   NULL, /* pad2 */ \
   NULL /* pad3 */ \
 }
diff --git test/internals/misc-internals.c test/internals/misc-internals.c
index a1777bb2..0defb6aa 100644
--- test/internals/misc-internals.c
+++ test/internals/misc-internals.c
@@ -679,8 +679,8 @@ _dbus_server_test (const char *test_data_dir _DBUS_GNUC_UNUSED)
     "tcp:host=localhost,port=1234",
     "tcp:host=localhost,port=1234;tcp:port=5678",
 #ifdef DBUS_UNIX
-    "unix:path=./boogie",
-    "tcp:port=1234;unix:path=./boogie",
+    "unix:path=/tmp/boogie",
+    "tcp:port=1234;unix:path=/tmp/boogie",
 #endif
   };
 
diff --git test/test-utils-glib.c test/test-utils-glib.c
index bcd83d11..8f25cf46 100644
--- test/test-utils-glib.c
+++ test/test-utils-glib.c
@@ -553,7 +553,7 @@ become_other_user (TestUser user,
 }
 
 /* Undo the effect of a successful call to become_other_user() */
-static void
+static _DBUS_GNUC_NORETURN void
 back_to_root (void)
 {
 #if defined(HAVE_GETRESUID) && defined(HAVE_SETRESUID) && defined(__linux__)
@@ -562,6 +562,7 @@ back_to_root (void)
 #else
   g_error ("become_other_user() cannot succeed on this platform");
 #endif
+  abort();
 }
 
 /*
diff --git tools/dbus-launch-x11.c tools/dbus-launch-x11.c
index 48ab51d8..e59353bf 100644
--- tools/dbus-launch-x11.c
+++ tools/dbus-launch-x11.c
@@ -306,7 +306,7 @@ x11_get_address (char **paddress, pid_t *pid, long *wid)
   int format;
   unsigned long items;
   unsigned long after;
-  char *data;
+  void *data;
 
   *paddress = NULL;
 
