--- ../../xdelta3-3.0.7/testing/regtest.cc	2013-05-13 04:54:35.000000000 +0000
+++ testing/regtest.cc	2013-05-20 02:32:05.243969745 +0000
@@ -761,7 +761,7 @@
   ExtFile d01, d12, d23;
   Options options;
   options.encode_srcwin_maxsz = 
-    std::max(spec0.Size(), options.encode_srcwin_maxsz);
+    std::max((unsigned long long)spec0.Size(), (unsigned long long)options.encode_srcwin_maxsz);
 
   spec0.WriteTmpFile(&f0);
   spec1.WriteTmpFile(&f1);
