--- server/protocol.c.orig	Tue Jun 29 08:21:28 2004
+++ server/protocol.c	Tue Jun 29 08:21:50 2004
@@ -1248,7 +1248,7 @@
      * We can only set a C-L in the response header if we haven't already
      * sent any buckets on to the next output filter for this request.
      */
-    if (ctx->data_sent == 0 && eos) {
+    if (ctx->data_sent == 0 && eos && !r->header_only) {
         ap_set_content_length(r, r->bytes_sent);
     }
 
