--- pat.c.orig	2011-03-13 23:15:42.000000000 -0700
+++ pat.c	2011-03-14 00:14:40.000000000 -0700
@@ -486,6 +486,7 @@ void cPatFilter::Process(u_short Pid, u_
                                  SI::ISO639LanguageDescriptor *ld = (SI::ISO639LanguageDescriptor *)d;
                                  strn0cpy(lang, I18nNormalizeLanguageCode(ld->languageCode), MAXLANGCODE1);
                                  }
+                                 IsAc3 = true;
                                  break;
                             default: ;
                             }
