$FreeBSD$

This is applied if WITH_DEBUG is unset.

--- src/osd/sdl/sdl.mak	2011-06-21 19:58:18.000000000 -0500
+++ src/osd/sdl/sdl.mak	2011-06-21 20:01:07.000000000 -0500
@@ -139,7 +139,7 @@
 ifeq ($(TARGETOS),freebsd)
 BASE_TARGETOS = unix
 SYNC_IMPLEMENTATION = tc
-DEFS += -DNO_AFFINITY_NP
+DEFS += -DNO_AFFINITY_NP -DNO_DEBUGGER
 LIBS += -lutil
 # /usr/local/include is not considered a system include directory
 # on FreeBSD. GL.h resides there and throws warnings
@@ -213,7 +213,7 @@
 ifndef GTK_INSTALL_ROOT
 NO_DEBUGGER = 1
 else
-DEBUGOBJS = $(SDLOBJ)/debugwin.o $(SDLOBJ)/dview.o $(SDLOBJ)/debug-sup.o $(SDLOBJ)/debug-intf.o
+DEBUGOBJS = $(SDLOBJ)/debugwin.o
 LIBS += -lgtk-win32-2.0 -lgdk-win32-2.0 -lgmodule-2.0 -lglib-2.0 -lgobject-2.0 \
 	-lpango-1.0 -latk-1.0 -lgdk_pixbuf-2.0
 CCOMFLAGS += -mms-bitfields
@@ -315,7 +315,7 @@
 ifeq ($(BASE_TARGETOS),unix)
 
 DEFS += -DSDLMAME_UNIX
-DEBUGOBJS = $(SDLOBJ)/debugwin.o $(SDLOBJ)/dview.o $(SDLOBJ)/debug-sup.o $(SDLOBJ)/debug-intf.o
+DEBUGOBJS = $(SDLOBJ)/debugwin.o
 LIBGL = -lGL
 ifeq ($(NO_X11),1)
 NO_DEBUGGER = 1
@@ -484,9 +484,9 @@
 
 # the new debugger relies on GTK+ in addition to the base SDLMAME needs
 # Non-X11 builds can not use the debugger
-INCPATH += `pkg-config --cflags-only-I gtk+-2.0` `pkg-config --cflags-only-I gconf-2.0`
-CCOMFLAGS += `pkg-config --cflags-only-other gtk+-2.0` `pkg-config --cflags-only-other gconf-2.0`
-LIBS += `pkg-config --libs gtk+-2.0` `pkg-config --libs gconf-2.0`
+#INCPATH += `pkg-config --cflags-only-I gtk+-2.0` `pkg-config --cflags-only-I gconf-2.0`
+#CCOMFLAGS += `pkg-config --cflags-only-other gtk+-2.0` `pkg-config --cflags-only-other gconf-2.0`
+#LIBS += `pkg-config --libs gtk+-2.0` `pkg-config --libs gconf-2.0`
 #CCOMFLAGS += -DGTK_DISABLE_DEPRECATED
 
 # some systems still put important things in a different prefix
