--- gresolver.pl.orig	Sat Oct 23 07:47:39 2004
+++ gresolver.pl	Sat Oct 23 07:47:53 2004
@@ -160,7 +160,7 @@
 	### build the dig query:
 	my $cmd = join(' ', (
 		$dig,
-		'+'.($recursive		== 1 ? '' : 'no').'recursive',
+		'+'.($recursive		== 1 ? '' : 'no').'recurse',
 		'+'.($additional	== 1 ? '' : 'no').'additional',
 		'+'.($authoritative	== 1 ? '' : 'no').'authority',
 		'+'.($trace		== 1 ? '' : 'no').'trace',
