diff -urN variants/nethack-3.4.2-deet.orig/include/config.h variants/nethack-3.4.2-deet/include/config.h
--- variants/nethack-3.4.2-deet.orig/include/config.h	Sun Nov 21 00:09:45 2004
+++ variants/nethack-3.4.2-deet/include/config.h	Sun Nov 21 00:10:51 2004
@@ -178,7 +178,7 @@
 /* #define COMPRESS "/usr/bin/compress"	*/	/* Lempel-Ziv compression */
 /* #define COMPRESS_EXTENSION ".Z" */		/* compress's extension */
 /* An example of one alternative you might want to use: */
-#define COMPRESS "/bin/gzip"  			/* FSF gzip compression */
+#define COMPRESS "/usr/bin/gzip"  			/* FSF gzip compression */
 #define COMPRESS_EXTENSION ".gz"  		/* normal gzip extension */
 #endif
 
diff -urN variants/nethack-3.4.2-deet.orig/include/unixconf.h variants/nethack-3.4.2-deet/include/unixconf.h
--- variants/nethack-3.4.2-deet.orig/include/unixconf.h	Sun Nov 21 00:09:41 2004
+++ variants/nethack-3.4.2-deet/include/unixconf.h	Sun Nov 21 00:11:32 2004
@@ -19,13 +19,13 @@
  */
 
 /* define exactly one of the following four choices */
-/* #define BSD 1 */	/* define for 4.n/Free/Open/Net BSD  */
+#define BSD 1 		/* define for 4.n/Free/Open/Net BSD  */
 			/* also for relatives like SunOS 4.x, DG/UX, and */
 			/* older versions of Linux */
 /* #define ULTRIX */	/* define for Ultrix v3.0 or higher (but not lower) */
 			/* Use BSD for < v3.0 */
 			/* "ULTRIX" not to be confused with "ultrix" */
-#define SYSV		/* define for System V, Solaris 2.x, newer versions */
+/* #define SYSV	*/	/* define for System V, Solaris 2.x, newer versions */
 			/* of Linux */
 /* #define HPUX */	/* Hewlett-Packard's Unix, version 6.5 or higher */
 			/* use SYSV for < v6.5 */
@@ -37,7 +37,7 @@
 /* #define NETWORK */	/* if running on a networked system */
 			/* e.g. Suns sharing a playground through NFS */
 /* #define SUNOS4 */	/* SunOS 4.x */
-#define LINUX		/* Another Unix clone */
+/* #define LINUX */	/* Another Unix clone */
 /* #define CYGWIN32 */	/* Unix on Win32 -- use with case sensitive defines */
 /* #define GENIX */	/* Yet Another Unix Clone */
 /* #define HISX */	/* Bull Unix for XPS Machines */
@@ -86,7 +86,7 @@
 /* #define VPIX_MUSIC */	/* play real music through speaker on systems
 				   with built-in VPIX support */
 
-#define LINUXFONT		/* include "linuxfont" option to use
+/* #define LINUXFONT */		/* include "linuxfont" option to use
 				   special linux console font (tty, not X11) */
 #ifdef LINUXFONT
 #define LINUXFONT_NAME "linuxhack.font"	/* installed in HACKDIR */
diff -urN variants/nethack-3.4.2-deet.orig/sys/unix/Makefile.src variants/nethack-3.4.2-deet/sys/unix/Makefile.src
--- variants/nethack-3.4.2-deet.orig/sys/unix/Makefile.src	Sun Nov 21 00:09:45 2004
+++ variants/nethack-3.4.2-deet/sys/unix/Makefile.src	Sun Nov 21 00:15:30 2004
@@ -31,7 +31,7 @@
 # For Systos users compiling on the ST, you'll either need a bourne shell
 # clone or you'll need to do make depend, etc. by hand. In either case,
 # the line below probably needs changing
-SHELL=/bin/sh
+# SHELL=/bin/sh
 # for Atari
 # SHELL=E:/GEMINI2/MUPFEL.TTP
 
@@ -61,7 +61,7 @@
 #	if you get setcgtty() warnings during execution, you are feeding gcc
 #		a non-ANSI <sys/ioctl.h> -- either run fixincludes on it or use
 #		-traditional in CFLAGS
-CC = gcc
+# CC = gcc
 #
 #	For Bull DPX/2 systems at B.O.S. 2.0 or higher use the following:
 #
@@ -125,14 +125,14 @@
 # flags for Linux
 #   compile normally
 
-COMMON_CFLAGS = -DNETHACK_3_4_2_DEET -fomit-frame-pointer -I../include `sdl-config --cflags`
+COMMON_CFLAGS = -DNETHACK_3_4_2_DEET -fomit-frame-pointer -I../include -I/usr/include/gnu `${SDL_CONFIG} --cflags` -I/usr/X11R6/include
 
 STD_CFLAGS = -O2 -ffast-math $(COMMON_CFLAGS)
 DBG_CFLAGS = -g -Wall -DNOEGNUDDEBUG $(COMMON_CFLAGS)
 
-CFLAGS = $(STD_CFLAGS)
+CFLAGS += $(STD_CFLAGS)
 
-LFLAGS = -L/usr/X11R6/lib
+LFLAGS = -L/usr/X11R6/lib -lgnuregex
 
 #   OR compile backwards compatible a.out format
 # CFLAGS = -O2 -b i486-linuxaout -fomit-frame-pointer -I../include
@@ -165,8 +165,8 @@
 # The Qt and Be window systems are written in C++, while the rest of
 # NetHack is standard C.  If using Qt, uncomment the LINK line here to get
 # the C++ libraries linked in.
-CXXFLAGS = $(CFLAGS) -I. -I$(QTDIR)/include
-CXX=g++
+CXXFLAGS += $(CFLAGS) -I. -I$(QTDIR)/include
+#CXX=g++
 #LINK=g++
 #	For cross-compiling, eg. with gcc on Linux (see also CC further up):
 #CXX=arm-linux-g++
@@ -222,7 +222,10 @@
 
 # Files for Entro-P's OpenGL 3D interface :: noeGNUd
 
-WINNOEGNUDLIB = `sdl-config --libs` -lSDL_image -lSDL_mixer -lGL -lGLU
+WINNOEGNUDLIB = `${SDL_CONFIG} --libs` -lSDL_image -lSDL_mixer -lGL -lGLU
+
+.c.o:
+	${CC} ${CFLAGS} -c $< -o $@
 
 NOEGNUD = ../win/noegnud/noegnud
 
diff -urN variants/nethack-3.4.2-deet.orig/sys/unix/Makefile.top variants/nethack-3.4.2-deet/sys/unix/Makefile.top
--- variants/nethack-3.4.2-deet.orig/sys/unix/Makefile.top	Sun Nov 21 00:09:45 2004
+++ variants/nethack-3.4.2-deet/sys/unix/Makefile.top	Sun Nov 21 00:15:58 2004
@@ -14,7 +14,7 @@
 # MAKE = make
 
 # make NetHack
-PREFIX	 	= /usr/local
+#PREFIX	 	= /usr/local
 GAMEUI   	= noegnud-0.8.0
 GAMEVARIANT	= nethack-3.4.2-deet
 GAMESHORT	= noegnud-$(GAMEVARIANT)
@@ -38,7 +38,7 @@
 # therefore there should not be anything in GAMEDIR that you want to keep
 # (if there is, you'll have to do the installation by hand or modify the
 # instructions)
-GAMEDIR		= $(PREFIX)/lib/$(GAMEUI)/$(GAMEVARIANT)
+GAMEDIR		= $(PREFIX)/share/$(GAMEUI)/$(GAMEVARIANT)
 VARDIR  = $(GAMEDIR)
 SHELLDIR	= $(PREFIX)/bin
 
diff -urN variants/nethack-3.4.2-deet.orig/sys/unix/Makefile.utl variants/nethack-3.4.2-deet/sys/unix/Makefile.utl
--- variants/nethack-3.4.2-deet.orig/sys/unix/Makefile.utl	Sun Nov 21 00:09:45 2004
+++ variants/nethack-3.4.2-deet/sys/unix/Makefile.utl	Sun Nov 21 00:16:28 2004
@@ -15,7 +15,7 @@
 
 # if you are using gcc as your compiler,
 #	uncomment the CC definition below if it's not in your environment
-CC = gcc
+# CC = gcc
 #
 #	For Bull DPX/2 systems at B.O.S. 2.0 or higher use the following:
 #
@@ -72,8 +72,8 @@
 
 # flags for Linux
 #   compile normally
-CFLAGS = -O2 -fomit-frame-pointer -I../include
-LFLAGS = -L/usr/X11R6/lib
+#CFLAGS = -O2 -fomit-frame-pointer -I../include
+#LFLAGS = -L/usr/X11R6/lib
 #   OR compile backwards compatible a.out format
 # CFLAGS = -O2 -b i486-linuxaout -fomit-frame-pointer -I../include
 # LFLAGS = -b i486-linuxaout -L/usr/X11R6/lib
@@ -89,7 +89,7 @@
 # flags for debugging:
 # CFLAGS = -g -I../include
 
-CFLAGS = -O -I../include
+CFLAGS += -O -I../include
 LFLAGS =
 
 LIBS =
