--- docs/_ext/djangodocs.py.orig	2010-11-21 05:00:41.000000000 +0800
+++ docs/_ext/djangodocs.py	2011-12-24 22:37:54.866639407 +0800
@@ -127,12 +127,14 @@
 
     # Don't use border=1, which docutils does by default.
     def visit_table(self, node):
+        self._table_row_index = 0 # Needed by Sphinx
         self.body.append(self.starttag(node, 'table', CLASS='docutils'))
 
     # <big>? Really?
     def visit_desc_parameterlist(self, node):
         self.body.append('(')
         self.first_param = 1
+        self.param_separator = node.child_text_separator
 
     def depart_desc_parameterlist(self, node):
         self.body.append(')')
