--- sshit.orig	Sun Aug  5 10:56:39 2007
+++ sshit	Sun Aug  5 10:59:03 2007
@@ -317,7 +317,7 @@
 
 while (<>) {
 	chomp;
-	if (/(failed|Invalid user) .*from (\d+\.\d+\.\d+\.\d+|[\da-fA-F:]+)/i ) { # IPv4 & IPv6
+	if (/(failed|error|Invalid user) .*from (\d+\.\d+\.\d+\.\d+|([a-z-]+\.)+[a-z]{2,4}|[\da-fA-F:]+)/i ) { # IPv4 & IPv6
 		$ip = $2;
 		if ($list{$ip}{name}) {
 			if ($list{$ip}{n} >= $MAX_COUNT) {
