--- configure.orig	Tue May 20 05:22:53 2003
+++ configure	Tue May 27 23:48:05 2003
@@ -2902,7 +2902,7 @@
 # Makefiles dir in the previous location, reset this variable to
 # 'Makefiles' and regenerate configure. Note this doesn't change all
 # occurances.
-MAKEFILES_SUFFIX=Library/Makefiles
+MAKEFILES_SUFFIX=Makefiles
 
 
 # HOST_INSTALL is the name of the install program in config.make so set it up
