--- npm-better-sqlite3-7.5.1-afdbd49cf18d0046d7385ca4d23da20c66f00d50/node_modules/better-sqlite3/deps/sqlite3.gyp.orig	2022-07-02 14:34:04 UTC
+++ npm-better-sqlite3-7.5.1-afdbd49cf18d0046d7385ca4d23da20c66f00d50/node_modules/better-sqlite3/deps/sqlite3.gyp
@@ -95,6 +95,14 @@
             ]
           }
         },
+        'OS == "freebsd"', {
+          'link_settings': {
+            'libraries': [
+              # This statically links libcrypto, whereas -lcrypto would dynamically link it
+              '/usr/local/lib/libcrypto.a'
+            ]
+          }
+        },
         { # Linux
           'link_settings': {
             'libraries': [
