--- npm-@signalapp-better-sqlite3-8.4.3-7ffa8d03d2a12543247936bfb7b9f74cdbc6fe9b-integrity/node_modules/@signalapp/better-sqlite3/deps/sqlite3.gyp.orig	2022-07-02 14:34:04 UTC
+++ npm-@signalapp-better-sqlite3-8.4.3-7ffa8d03d2a12543247936bfb7b9f74cdbc6fe9b-integrity/node_modules/@signalapp/better-sqlite3/deps/sqlite3.gyp
@@ -100,6 +100,15 @@
             ]
           }
         },
+        'OS == "freebsd"', {
+          'link_settings': {
+            'libraries': [
+              # This statically links libcrypto, whereas -lcrypto would dynamically link it
+              '/usr/local/lib/libcrypto.a',
+              '/usr/local/lib/libsignal_tokenizer.a'
+             ]
+           }
+         },
         { # Linux
           'link_settings': {
             'libraries': [
