--- npm-@signalapp-better-sqlite3-8.0.3-c279ca3f1d1e7f6554d2a629d717e96e1ec45809-integrity/node_modules/@signalapp/better-sqlite3/deps/sqlite3.gyp.orig	2022-07-02 14:34:04 UTC
+++ npm-@signalapp-better-sqlite3-8.0.3-c279ca3f1d1e7f6554d2a629d717e96e1ec45809-integrity/node_modules/@signalapp/better-sqlite3/deps/sqlite3.gyp
@@ -92,6 +92,14 @@
             'libraries': [
               # This statically links libcrypto, whereas -lcrypto would dynamically link it
               '<(SHARED_INTERMEDIATE_DIR)/sqlite3/OpenSSL-mac-<(target_arch)/libcrypto.a'
+            ]
+          }
+        },
+        'OS == "freebsd"', {
+          'link_settings': {
+            'libraries': [
+              # This statically links libcrypto, whereas -lcrypto would dynamically link it
+              '/usr/local/lib/libcrypto.a'
             ]
           }
         },
