--- node_modules/@signalapp/better-sqlite3/deps/sqlite3.gyp.orig	2022-07-02 14:34:04 UTC
+++ node_modules/@signalapp/better-sqlite3/deps/sqlite3.gyp
@@ -90,6 +90,13 @@
             ]
           }
         },
+        'OS == "freebsd"', {
+          'link_settings': {
+            'libraries': [
+              '/usr/local/lib/libsignal_sqlcipher_extension.a',
+             ]
+           }
+         },
         { # Linux
           'link_settings': {
             'libraries': [
