--- Makefile.system.orig	2019-02-06 00:50:04.658572000 +0800
+++ Makefile.system	2019-02-06 00:51:30.598745000 +0800
@@ -9,6 +9,16 @@
 TOPDIR = .
 endif
 
+ifeq ($(ARCH), amd64)
+override ARCH=x86_64
+else ifeq ($(ARCH), powerpc64)
+override ARCH=power
+else ifeq ($(ARCH), i386)
+override ARCH=x86
+else ifeq ($(ARCH), aarch64)
+override ARCH=arm64
+endif
+
 NETLIB_LAPACK_DIR = $(TOPDIR)/lapack-netlib
 
 # Default C compiler
