diff -ruN Python/thread_pthread.h.orig Python/thread_pthread.h
--- Python/thread_pthread.h.orig	Wed Feb  5 01:26:59 2003
+++ Python/thread_pthread.h	Wed Feb  5 01:28:32 2003
@@ -128,6 +128,10 @@
 {
 	pthread_t th;
 	int success;
+    pthread_attr_t pta;
+    pthread_attr_init(&pta);
+    pthread_attr_setstacksize(&pta, (1<<17));
+
 #ifdef PTHREAD_SYSTEM_SCHED_SUPPORTED
 	pthread_attr_t attrs;
 #endif
@@ -156,7 +160,7 @@
 #ifdef PTHREAD_SYSTEM_SCHED_SUPPORTED
 				 &attrs,
 #else
-				 (pthread_attr_t*)NULL,
+				 (pthread_attr_t*)&pta,
 #endif
 				 (void* (*)(void *))func,
 				 (void *)arg
