--- lib/chef/platform/provider_mapping.rb.orig	2013-11-01 12:12:31.000000000 +0900
+++ lib/chef/platform/provider_mapping.rb	2013-11-01 12:12:47.000000000 +0900
@@ -60,7 +60,7 @@
           :freebsd => {
             :default => {
               :group   => Chef::Provider::Group::Pw,
-              :package => Chef::Provider::Package::Freebsd,
+              :package => Chef::Provider::Package::PkgUpgrade,
               :service => Chef::Provider::Service::Freebsd,
               :user    => Chef::Provider::User::Pw,
               :cron    => Chef::Provider::Cron
