--- tmac/eqnrc.orig	Thu Mar 16 10:18:38 2000
+++ tmac/eqnrc	Fri Apr 28 20:59:24 2000
@@ -53,6 +53,7 @@
 ifdef ascii ! define n %1% !
 ifdef latin1 ! define n %1% !
 ifdef koi8-r ! define n %1% !
+ifdef nippon ! define n %1% !
 ifdef n !
 set nroff 1
 !
--- tmac/troffrc.orig	Thu Mar 16 10:18:38 2000
+++ tmac/troffrc	Fri Apr 28 20:59:24 2000
@@ -14,6 +14,7 @@
 .do ds troffrc!ascii tmac.tty
 .do ds troffrc!latin1 tmac.tty
 .do ds troffrc!koi8-r tmac.tty
+.do ds troffrc!nippon tmac.tty
 .do ds troffrc!lj4 tmac.lj4
 .do ds troffrc!html tmac.html
 .do if d troffrc!\*[.T] \
