--- template/Makefile.orig	2009-11-15 19:42:51.000000000 +0100
+++ template/Makefile	2009-11-15 19:43:01.000000000 +0100
@@ -17,8 +17,10 @@
 
 install:
 	@echo "Installing templates ..."
-	@for template_var in README $(PLATFORM).sh; do \
-            cp $${template_var} $(INSTALL_TEMPL)/$${template_var}; \
+	@for template_var in $(PLATFORM).sh; do \
+            cp $${template_var} $(INSTALL_BIN)/gw6c-$${template_var}; \
         done
+	@mkdir -p $(ETCDIR)
+	@cp README $(ETCDIR)
 
 clean: 
--- Makefile.orig	2009-11-15 19:50:40.000000000 +0100
+++ Makefile	2009-11-15 19:50:57.000000000 +0100
@@ -167,17 +167,14 @@
 	@mkdir -p $(INSTALL_DIR)
 	@mkdir -p $(INSTALL_BIN)
 	@mkdir -p $(INSTALL_MAN)
-	@mkdir -p $(INSTALL_TEMPL)
 
 	@for dir in ${SUBDIRS}; do \
 	    $(MAKE) -C $$dir install || exit 1; \
 	done
 
 	@cp $(TARGET) $(INSTALL_BIN)
-	@cp $(BIN_DIR)/gw6c.conf.sample $(INSTALL_BIN)
-	@[ -f $(INSTALL_BIN)/gw6c.conf ] || { \
-	    cp $(INSTALL_BIN)/gw6c.conf.sample $(INSTALL_BIN)/gw6c.conf; \
-	}
+	@mkdir -p $(ETCDIR)
+	@cp $(BIN_DIR)/gw6c.conf.sample $(ETCDIR)
 
 
 # This makefile target will clean the build tree of the Gateway6 Client.
