--- curses/cursesscreen.C.orig	2011-04-04 11:58:28.000000000 +0000
+++ curses/cursesscreen.C	2011-06-27 19:41:38.000000000 +0000
@@ -73,7 +73,9 @@
 	while (input_buf.size() > 0)
 	{
 
-		char *inbuf=&input_buf[0], *outbuf=&winput_buf[winput_cnt];
+		//char *inbuf=&input_buf[0], *outbuf=&winput_buf[winput_cnt];
+		const char *inbuf=&input_buf[0];
+		char *outbuf=&winput_buf[winput_cnt];
 		size_t inbytesleft=input_buf.size(),
 			outbytesleft=winput_buf.size()-winput_cnt;
 
