--- ../gettext-runtime/intl/localealias.c.orig	2023-03-20 16:40:19.034959000 +0000
+++ ../gettext-runtime/intl/localealias.c	2023-03-20 16:42:04.286791000 +0000
@@ -337,14 +337,14 @@
 		      if (new_pool == NULL)
 			goto out;
 
-		      if (__builtin_expect (string_space != new_pool, 0))
+		      if (1)
 			{
 			  size_t i;
 
 			  for (i = 0; i < nmap; i++)
 			    {
-			      map[i].alias += new_pool - string_space;
-			      map[i].value += new_pool - string_space;
+			      map[i].alias = new_pool + (map[i].alias - string_space);
+			      map[i].value = new_pool + (map[i].value - string_space);
 			    }
 			}
 
