--- soffice.orig	Fri Aug 25 12:05:30 2000
+++ soffice	Fri Aug 25 12:00:54 2000
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/compat/linux/bin/sh
 #
 # StarOffice startup script
 # (c) 1999 Sun Microsystems, Inc.
@@ -7,6 +7,9 @@
 # STAR_PROFILE_LOCKING_DISABLED=1
 # export STAR_PROFILE_LOCKING_DISABLED
 #
+
+# fix coredump on remote DISPLAY (IRIX)
+export SAL_IGNOREXERRORS=1
 
 # uncomment this to remote start soffice on hostname  
 # SO_REMOTE_START=rsh
