--- channels/chan_sip.c.orig	2008-04-09 11:39:09.000000000 +0300
+++ channels/chan_sip.c	2008-04-09 11:40:28.000000000 +0300
@@ -2998,6 +2998,9 @@
 			p->t38.state = T38_LOCAL_DIRECT;
 			if (option_debug)
 				ast_log(LOG_DEBUG,"T38State change to %d on channel %s\n", p->t38.state, ast->name);
+		} else if (!strcasecmp(ast_var_name(current), "SIP_FORCE_CALLID")) {
+			
+			ast_string_field_set(p, callid, ast_var_value(current));
 		}
 
 	}
