diff --git a/nova/conf/network.py b/nova/conf/network.py
index 3bb4cd9..eb57041 100644
--- a/nova/conf/network.py
+++ b/nova/conf/network.py
@@ -704,6 +704,30 @@ Related options:
 """),
 ]
 
+freebsd_net_opts = [
+    cfg.StrOpt("freebsdnet_interface_driver",
+            default="nova.network.freebsd.FreeBSDBridgeInterfaceDriver",
+            help="""
+This is the class used as the ethernet device driver for freebsdnet bridge
+operations. The default value should be all you need for most cases, but if you
+wish to use a customized class, set this option to the full dot-separated
+import path for that class.
+
+Possible values:
+
+    Any string representing a dot-separated class path that Nova can import.
+"""),
+    cfg.StrOpt("freebsdnet_ovs_integration_bridge",
+            default="br-int",
+            help="""
+The name of the Open vSwitch bridge that is used with freebsdnet when connecting
+with Open vSwitch."
+
+Possible values:
+
+    Any string representing a valid bridge name.
+"""),
+]
 
 ldap_dns_opts = [
     cfg.StrOpt('ldap_dns_url',
@@ -766,12 +790,13 @@ by using this option.
                      'some rpc network calls will be sent directly to host.'),
 ]
 
-ALL_DEFAULT_OPTS = (linux_net_opts + network_opts + ldap_dns_opts
+ALL_DEFAULT_OPTS = (linux_net_opts + freebsd_net_opts + network_opts + ldap_dns_opts
                    + rpcapi_opts + driver_opts)
 
 
 def register_opts(conf):
     conf.register_opts(linux_net_opts)
+    conf.register_opts(freebsd_net_opts)
     conf.register_opts(network_opts)
     conf.register_opts(ldap_dns_opts)
     conf.register_opts(driver_opts)

