diff --git a/nova/virt/block_device.py b/nova/virt/block_device.py
index f6ed6a3..6539fca 100644
--- a/nova/virt/block_device.py
+++ b/nova/virt/block_device.py
@@ -228,7 +228,12 @@ class DriverVolumeBlockDevice(DriverBlockDevice):
             {k: v for k, v in six.iteritems(self._bdm_obj)
              if k in self._new_fields | set(['delete_on_termination'])}
         )
-        self['mount_device'] = self._bdm_obj.device_name
+
+        device_name = self._bdm_obj.device_name
+        if self['disk_bus'] == 'xen':
+            device_name = device_name.replace("/dev/vd", "/dev/xvd")
+        self['mount_device'] = device_name
+
         try:
             self['connection_info'] = jsonutils.loads(
                 self._bdm_obj.connection_info)
