diff --git a/nova/virt/libvirt/driver.py b/nova/virt/libvirt/driver.py
index 6d0e41ad2c..ebd1a6fdad 100644
--- a/nova/virt/libvirt/driver.py
+++ b/nova/virt/libvirt/driver.py
@@ -4321,7 +4321,7 @@ class LibvirtDriver(driver.ComputeDriver):
 
     def _create_consoles(self, virt_type, guest, instance, flavor, image_meta,
                          caps):
-        if virt_type in ("qemu", "kvm"):
+        if virt_type in ("qemu", "kvm", "xen"):
             # Create the serial console char devices
             self._create_serial_console_devices(guest, instance, flavor,
                                                 image_meta)
