--- src/osd/sdl/sdl.mak.orig	2009-03-05 20:03:50.000000000 +0100
+++ src/osd/sdl/sdl.mak	2009-03-05 20:06:22.000000000 +0100
@@ -264,10 +264,10 @@
 # the new debugger relies on GTK+ in addition to the base SDLMAME needs
 # Non-X11 builds can not use the debugger
 ifndef NO_X11
-OSDCOREOBJS += $(SDLOBJ)/debugwin.o $(SDLOBJ)/dview.o $(SDLOBJ)/debug-sup.o $(SDLOBJ)/debug-intf.o
-CFLAGS += `pkg-config --cflags gtk+-2.0` `pkg-config --cflags gconf-2.0` 
-LIBS += `pkg-config --libs gtk+-2.0` `pkg-config --libs gconf-2.0`
-CFLAGS += -DGTK_DISABLE_DEPRECATED
+OSDCOREOBJS += $(SDLOBJ)/debugwin.o
+#CFLAGS += `pkg-config --cflags gtk+-2.0` `pkg-config --cflags gconf-2.0` 
+#LIBS += `pkg-config --libs gtk+-2.0` `pkg-config --libs gconf-2.0`
+#CFLAGS += -DGTK_DISABLE_DEPRECATED
 else
 OSDCOREOBJS += $(SDLOBJ)/debugwin.o
 endif # NO_X11
