--- youtube-dl.orig	2008-10-12 16:49:43.000000000 +0400
+++ youtube-dl	2008-10-12 16:54:22.000000000 +0400
@@ -309,6 +309,20 @@
 						retcode = self.trouble('ERROR: postprocessing: %s' % str(err))
 						continue
 
+					# Convert to avi if needed
+					if self._params['useavi']:
+						avi_filename = '%s.avi' % filename
+ 
+						try:
+							os.system('%%CONV2AVI_CMD%% "%s" "%s" > %s 2> %s' % (filename, avi_filename, os.path.devnull, os.path.devnull))
+							self.to_stdout('Video file converted to %s' % avi_filename)
+						except OSError, e:
+							retcode = self.trouble('ERROR: unable to convert file: %s' % e.strerror)
+
+						try:
+							os.unlink(filename)
+						except OSError, e:
+							retcode = self.trouble('ERROR: unable to remove downloaded file: %s' % e.strerror)
 				break
 			if not suitable_found:
 				retcode = self.trouble('ERROR: no suitable InfoExtractor: %s' % url)
@@ -859,6 +873,8 @@
 				action='store_true', dest='useliteral', help='use literal title in file name', default=False)
 		parser.add_option('-n', '--netrc',
 				action='store_true', dest='usenetrc', help='use .netrc authentication data', default=False)
+		parser.add_option('-a', '--avi',
+				action='store_true', dest='useavi', help='convert file to MPEG4 format', default=False)
 		parser.add_option('-g', '--get-url',
 				action='store_true', dest='geturl', help='simulate, quiet but print URL', default=False)
 		parser.add_option('-e', '--get-title',
@@ -915,6 +931,7 @@
 				or u'%(id)s.%(ext)s'),
 			'ignoreerrors': opts.ignoreerrors,
 			'ratelimit': opts.ratelimit,
+			'useavi': opts.useavi,
 			})
 		fd.add_info_extractor(youtube_pl_ie)
 		fd.add_info_extractor(metacafe_ie)
