--- youtube-dl.orig	2008-06-09 09:36:33.000000000 +0400
+++ youtube-dl	2008-06-09 09:43:10.000000000 +0400
@@ -216,6 +216,7 @@
 cmdl_parser.add_option('-t', '--title', action='store_true', dest='use_title', help='use title in file name')
 cmdl_parser.add_option('-l', '--literal', action='store_true', dest='use_literal', help='use literal title in file name')
 cmdl_parser.add_option('-n', '--netrc', action='store_true', dest='use_netrc', help='use .netrc authentication data')
+cmdl_parser.add_option('-a', '--avi', action='store_true', dest='use_avi', help='convert file to MPEG4 format')
 cmdl_parser.add_option('-g', '--get-url', action='store_true', dest='get_url', help='print final video URL only')
 cmdl_parser.add_option('-2', '--title-too', action='store_true', dest='get_title', help='used with -g, print title too')
 cmdl_parser.add_option('-f', '--format', dest='video_format', metavar='FORMAT', help='append &fmt=FORMAT to the URL')
@@ -420,6 +421,22 @@
 	
 	except KeyboardInterrupt:
 		sys.exit('\n')
-	
+
+	# Convert to avi if needed
+	avi_filename = '%s.avi' % video_filename
+
+	try:
+		os.system('%%CONV2AVI_CMD%% "%s" "%s" > %s 2> %s' % (video_filename, avi_filename, os.path.devnull, os.path.devnull))
+		cond_print('Video file converted to %s\n' % avi_filename)
+	except OSError:
+		sys.exit('Error: unable to convert file.\n')
+	except KeyboardInterrupt:
+		sys.exit('\n')
+
+	try:
+		os.unlink(video_filename)
+	except OSError:
+		sys.exit('Error: unable to remove downloaded file.\n')
+
 # Finish
 sys.exit()
