--- modules/log-prelude/Makefile.in.orig    Sun Jan 14 14:09:10 2007
+++ modules/log-prelude/Makefile.in Sun Jan 14 14:09:21 2007
@@ -216,7 +216,7 @@
 sysconfdir = @sysconfdir@
 target_alias = @target_alias@
 AUTOMAKE_OPTIONS = foreign
-AM_CPPFLAGS = -I/usr/include/libprelude -I$(top_srcdir)/nepenthes-core/include -I$(top_srcdir)/nepenthes-core/src -pipe -D _GNU_SOURCE $(CPPFLAG_PRELUDE)
+AM_CPPFLAGS = -I/usr/local/include/libprelude -I$(top_srcdir)/nepenthes-core/include -I$(top_srcdir)/nepenthes-core/src -pipe -D _GNU_SOURCE
 AM_CXXFLAGS = -Wall
 AM_LDFLAGS = $(LDFLAG_PRELUDE) ${LIB_PRELUDE}
 pkglib_LTLIBRARIES = logprelude.la
