--- CMakeLists.txt	2023-06-20 14:32:10.000000000 +0100
+++ CMakeLists.txt	2023-08-25 13:20:44.773802000 +0100
@@ -28,7 +28,6 @@
     Config
     ConfigWidgets
     CoreAddons
-    DocTools
     I18n
     KCMUtils
     KIO
@@ -64,8 +63,6 @@
 add_subdirectory(src)
 
 include(ECMOptionalAddSubdirectory)
-ecm_optional_add_subdirectory( doc )
-kdoctools_install(po)
 
 # add clang-format target for all our real source files
 file(GLOB_RECURSE ALL_CLANG_FORMAT_SOURCE_FILES *.cpp *.h)
