--- src/corelib/global/qglobal.h.orig	2009-09-07 13:00:30.000000000 +0200
+++ src/corelib/global/qglobal.h	2009-09-14 12:46:30.965296640 +0200
@@ -1594,8 +1594,16 @@
 #ifdef QT3_SUPPORT
 Q_CORE_EXPORT QT3_SUPPORT void qSystemWarning(const char *msg, int code = -1);
 #endif /* QT3_SUPPORT */
-Q_CORE_EXPORT void qErrnoWarning(int code, const char *msg, ...);
-Q_CORE_EXPORT void qErrnoWarning(const char *msg, ...);
+Q_CORE_EXPORT void qErrnoWarning(int code, const char *msg, ...)
+#if defined(Q_CC_GNU) && !defined(__INSURE__)
+    __attribute__ ((format (printf, 2, 3)))
+#endif
+    ;
+Q_CORE_EXPORT void qErrnoWarning(const char *msg, ...)
+#if defined(Q_CC_GNU) && !defined(__INSURE__)
+    __attribute__ ((format (printf, 1, 2)))
+#endif
+    ;
 
 #if (defined(QT_NO_DEBUG_OUTPUT) || defined(QT_NO_TEXTSTREAM)) && !defined(QT_NO_DEBUG_STREAM)
 #define QT_NO_DEBUG_STREAM
--- src/corelib/tools/qbytearray.h.orig	2009-09-07 13:00:30.000000000 +0200
+++ src/corelib/tools/qbytearray.h	2009-09-14 12:44:29.801547129 +0200
@@ -95,8 +95,16 @@
 Q_CORE_EXPORT int qstrnicmp(const char *, const char *, uint len);
 
 // implemented in qvsnprintf.cpp
-Q_CORE_EXPORT int qvsnprintf(char *str, size_t n, const char *fmt, va_list ap);
-Q_CORE_EXPORT int qsnprintf(char *str, size_t n, const char *fmt, ...);
+Q_CORE_EXPORT int qvsnprintf(char *str, size_t n, const char *fmt, va_list ap)
+#if defined(Q_CC_GNU) && !defined(__INSURE__)
+        __attribute__ ((format (printf, 3, 0)))
+#endif
+    ;
+Q_CORE_EXPORT int qsnprintf(char *str, size_t n, const char *fmt, ...)
+#if defined(Q_CC_GNU) && !defined(__INSURE__)
+        __attribute__ ((format (printf, 3, 4)))
+#endif
+    ;
 
 #ifdef QT3_SUPPORT
 inline QT3_SUPPORT void *qmemmove(void *dst, const void *src, uint len)
