--- tp/Texinfo/XS/gnulib/lib/stdint.in.h.orig	2022-10-06 15:01:42 UTC
+++ tp/Texinfo/XS/gnulib/lib/stdint.in.h
@@ -306,7 +306,7 @@ typedef gl_uint_fast32_t gl_uint_fast16_t;
    uintptr_t to avoid conflicting declarations of system functions like
    _findclose in <io.h>.  */
 # if !((defined __KLIBC__ && defined _INTPTR_T_DECLARED) \
-       || defined __MINGW32__)
+       || defined __MINGW32__ || defined __CHERI_PURE_CAPABILITY__)
 #  undef intptr_t
 #  undef uintptr_t
 #  ifdef _WIN64
