From dcaf03d1f0d6d12689f604347d169bfc732af69a Mon Sep 17 00:00:00 2001
From: David Lukes <dafydd.lukes@gmail.com>
Date: Thu, 27 Aug 2020 12:57:27 +0200
Subject: [PATCH] Fix #199

---
 sys/praat.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/sys/praat.cpp b/sys/praat.cpp
index 4c359e012..4f57f16b4 100644
--- sys/praat.cpp
+++ sys/praat.cpp
@@ -1078,6 +1078,7 @@ void praat_init (conststring32 title, int argc, char **argv)
 	while (praatP.argumentNumber < argc && argv [praatP.argumentNumber] [0] == '-') {
 		if (strequ (argv [praatP.argumentNumber], "-")) {
 			praatP.hasCommandLineInput = true;
+			praatP.argumentNumber += 1;
 		} else if (strequ (argv [praatP.argumentNumber], "--open")) {
 			foundTheOpenOption = true;
 			praatP.argumentNumber += 1;
