--- gdevmjc.mak.~1~	Thu Sep  5 16:01:29 1996
+++ gdevmjc.mak	Sun Jun  6 23:58:16 1999
@@ -18,11 +18,11 @@
 	$(CCC) -DA4 gdevmjc.c
 
 mjc180.dev: $(mj700v2c_)
-	$(SHP)gssetdev mjc180 $(mj700v2c_)
+	$(SETPDEV) mjc180 $(mj700v2c_)
 mjc360.dev: $(mj700v2c_)
-	$(SHP)gssetdev mjc360 $(mj700v2c_)
+	$(SETPDEV) mjc360 $(mj700v2c_)
 mjc720.dev: $(mj700v2c_)
-	$(SHP)gssetdev mjc720 $(mj700v2c_)
+	$(SETPDEV) mjc720 $(mj700v2c_)
 mj500c.dev: $(mj700v2c_)
-	$(SHP)gssetdev mj500c $(mj700v2c_)
+	$(SETPDEV) mj500c $(mj700v2c_)
 
