--- plugins/libwin32/loader/Makefile.orig	Fri Dec 28 10:36:49 2001
+++ plugins/libwin32/loader/Makefile	Fri Dec 28 10:35:55 2001
@@ -534,8 +534,8 @@
 # so we eliminate .rel.text section from library
 # it works this way too
 # we actually don't need this line
-#stubs.lo: stubs.s
-#	$(CC) -c $(srcdir)/stubs.s -o stubs.lo
+stubs.lo: stubs.s
+	$(CC) -c $(srcdir)/stubs.s -o stubs.lo
 
 #win32.lo: win32.c
 #	$(CC) -O2 -fno-inline -fno-strict-aliasing $(DEFS) $(WINFLAGS) -c $< -o $@
