--- menuselect-tree.orig	2011-02-21 09:56:29.000000000 +0100
+++ menuselect-tree	2011-02-21 09:56:46.000000000 +0100
@@ -777,7 +777,6 @@
 		<member name="RADIO_RELAX" displayname="Relax DTMF for Radio Applications">
 		</member>
 		<member name="G711_NEW_ALGORITHM" displayname="Use the NEW ulaw/alaw codecs (slower, but cleaner)">
-			<defaultenabled>no</defaultenabled>
 		</member>
 		<member name="G711_REDUCED_BRANCHING" displayname="New ulaw/alaw codec, reduced branching (might help it run faster in some architectures)">
 			<depend>G711_NEW_ALGORITHM</depend>
