--- ./addons/.moduleinfo.orig	2011-04-21 22:02:46.000000000 +0200
+++ ./addons/.moduleinfo	2011-04-26 20:00:53.552445281 +0200
@@ -1,14 +1,12 @@
 <category name="MENUSELECT_ADDONS" displayname="Add-ons (See README-addons.txt)" remove_on_change="addons/modules.link">
 <member name="app_mysql" displayname="Simple Mysql Interface" remove_on_change="addons/app_mysql.o addons/app_mysql.so">
 	<depend>mysqlclient</depend>
-	<defaultenabled>no</defaultenabled>
 </member>
 <member name="app_saycountpl" displayname="Say polish counting words" remove_on_change="addons/app_saycountpl.o addons/app_saycountpl.so">
 	<defaultenabled>no</defaultenabled>
 </member>
 <member name="cdr_mysql" displayname="MySQL CDR Backend" remove_on_change="addons/cdr_mysql.o addons/cdr_mysql.so">
 	<depend>mysqlclient</depend>
-	<defaultenabled>no</defaultenabled>
 </member>
 <member name="chan_mobile" displayname="Bluetooth Mobile Device Channel Driver" remove_on_change="addons/chan_mobile.o addons/chan_mobile.so">
 	<depend>bluetooth</depend>
@@ -22,6 +20,5 @@
 </member>
 <member name="res_config_mysql" displayname="MySQL RealTime Configuration Driver" remove_on_change="addons/res_config_mysql.o addons/res_config_mysql.so">
 	<depend>mysqlclient</depend>
-	<defaultenabled>no</defaultenabled>
 </member>
 </category>
