diff -Nru openjdk.orig/jdk/src/share/classes/java/beans/Introspector.java openjdk/jdk/src/share/classes/java/beans/Introspector.java
--- jdk/src/share/classes/java/beans/Introspector.java	2013-04-16 14:35:31.707279166 +0100
+++ jdk/src/share/classes/java/beans/Introspector.java	2013-04-16 14:39:23.715018865 +0100
@@ -102,7 +102,7 @@
     public final static int IGNORE_ALL_BEANINFO        = 3;
 
     // Static Caches to speed up introspection.
-    private static final WeakCache<Class<?>, Method[]> declaredMethodCache = new WeakCache<>();
+    private static final WeakCache<Class<?>, Method[]> declaredMethodCache = new WeakCache<Class<?>, Method[]>();
 
     private Class beanClass;
     private BeanInfo explicitBeanInfo;
diff -Nru openjdk.orig/jdk/src/share/classes/java/beans/ThreadGroupContext.java openjdk/jdk/src/share/classes/java/beans/ThreadGroupContext.java
--- jdk/src/share/classes/java/beans/ThreadGroupContext.java	2013-04-16 14:35:31.707279166 +0100
+++ jdk/src/share/classes/java/beans/ThreadGroupContext.java	2013-04-16 14:40:01.243623831 +0100
@@ -25,9 +25,6 @@
 
 package java.beans;
 
-import com.sun.beans.finder.BeanInfoFinder;
-import com.sun.beans.finder.PropertyEditorFinder;
-
 import java.awt.GraphicsEnvironment;
 import java.util.HashMap;
 import java.util.Map;
@@ -42,7 +39,7 @@
  */
 final class ThreadGroupContext {
 
-    private static final Map<ThreadGroup, ThreadGroupContext> contexts = new WeakHashMap<>();
+    private static final Map<ThreadGroup, ThreadGroupContext> contexts = new WeakHashMap<ThreadGroup, ThreadGroupContext>();
 
     /**
      * Returns the appropriate {@code AppContext} for the caller,
@@ -66,8 +63,6 @@
     private volatile Boolean isGuiAvailable;
 
     private Map<Class<?>, BeanInfo> beanInfoCache;
-    private BeanInfoFinder beanInfoFinder;
-    private PropertyEditorFinder propertyEditorFinder;
 
 
     boolean isDesignTime() {
@@ -99,7 +94,7 @@
 
     BeanInfo putBeanInfo(Class<?> type, BeanInfo info) {
         if (this.beanInfoCache == null) {
-            this.beanInfoCache = new WeakHashMap<>();
+            this.beanInfoCache = new WeakHashMap<Class<?>, BeanInfo>();
         }
         return this.beanInfoCache.put(type, info);
     }
@@ -116,18 +111,4 @@
         }
     }
 
-
-    synchronized BeanInfoFinder getBeanInfoFinder() {
-        if (this.beanInfoFinder == null) {
-            this.beanInfoFinder = new BeanInfoFinder();
-        }
-        return this.beanInfoFinder;
-    }
-
-    synchronized PropertyEditorFinder getPropertyEditorFinder() {
-        if (this.propertyEditorFinder == null) {
-            this.propertyEditorFinder = new PropertyEditorFinder();
-        }
-        return this.propertyEditorFinder;
-    }
 }
