diff -Nru openjdk.orig/jdk/src/share/lib/security/java.security openjdk/jdk/src/share/lib/security/java.security
--- jdk/src/share/lib/security/java.security	2013-04-16 14:23:48.623949921 +0100
+++ jdk/src/share/lib/security/java.security	2013-04-16 14:26:07.318184299 +0100
@@ -128,7 +128,9 @@
 # corresponding RuntimePermission ("accessClassInPackage."+package) has
 # been granted.
 package.access=sun.,\
-               com.sun.xml.internal.,\
+               com.sun.xml.internal.bind.,\
+               com.sun.xml.internal.org.jvnet.staxex.,\
+               com.sun.xml.internal.ws.,\
                com.sun.imageio.,\
                com.sun.istack.internal.,\
                com.sun.jmx.
@@ -144,7 +146,9 @@
 # checkPackageDefinition.
 #
 package.definition=sun.,\
-                   com.sun.xml.internal.,\
+                   com.sun.xml.internal.bind.,\
+                   com.sun.xml.internal.org.jvnet.staxex.,\
+                   com.sun.xml.internal.ws.,\
                    com.sun.imageio.,\
                    com.sun.istack.internal.,\
                    com.sun.jmx.
diff -Nru openjdk.orig/jdk/src/share/lib/security/java.security-solaris openjdk/jdk/src/share/lib/security/java.security-solaris
--- jdk/src/share/lib/security/java.security-solaris	2013-04-16 14:23:48.623949921 +0100
+++ jdk/src/share/lib/security/java.security-solaris	2013-04-16 14:26:30.082551058 +0100
@@ -129,7 +129,9 @@
 # corresponding RuntimePermission ("accessClassInPackage."+package) has
 # been granted.
 package.access=sun.,\
-               com.sun.xml.internal.,\
+               com.sun.xml.internal.bind.,\
+               com.sun.xml.internal.org.jvnet.staxex.,\
+               com.sun.xml.internal.ws.,\
                com.sun.imageio.
                com.sun.istack.internal.,\
                com.sun.jmx.
@@ -145,7 +147,9 @@
 # checkPackageDefinition.
 #
 package.definition=sun.,\
-                   com.sun.xml.internal.,\
+                   com.sun.xml.internal.bind.,\
+                   com.sun.xml.internal.org.jvnet.staxex.,\
+                   com.sun.xml.internal.ws.,\
                    com.sun.imageio.
                    com.sun.istack.internal.,\
                    com.sun.jmx.
diff -Nru openjdk.orig/jdk/src/share/lib/security/java.security-windows openjdk/jdk/src/share/lib/security/java.security-windows
--- jdk/src/share/lib/security/java.security-windows	2013-04-16 14:23:48.623949921 +0100
+++ jdk/src/share/lib/security/java.security-windows	2013-04-16 14:26:51.170890824 +0100
@@ -129,7 +129,9 @@
 # corresponding RuntimePermission ("accessClassInPackage."+package) has
 # been granted.
 package.access=sun.,\
-               com.sun.xml.internal.,\
+               com.sun.xml.internal.bind.,\
+               com.sun.xml.internal.org.jvnet.staxex.,\
+               com.sun.xml.internal.ws.,\
                com.sun.imageio.
                com.sun.istack.internal.,\
                com.sun.jmx.
@@ -145,7 +147,9 @@
 # checkPackageDefinition.
 #
 package.definition=sun.,\
-                   com.sun.xml.internal.,\
+                   com.sun.xml.internal.bind.,\
+                   com.sun.xml.internal.org.jvnet.staxex.,\
+                   com.sun.xml.internal.ws.,\
                    com.sun.imageio.
                    com.sun.istack.internal.,\
                    com.sun.jmx.
