Changes required to build on gcc5-aux

--- source/xml/sax/xml-sax-input_sources-streams-sockets.ads.orig	2011-07-22 20:16:17 UTC
+++ source/xml/sax/xml-sax-input_sources-streams-sockets.ads
@@ -63,7 +63,7 @@ package XML.SAX.Input_Sources.Streams.So
 
 private
 
-   type Socket_Input_Source is new Stream_Input_Source with record
+   type Socket_Input_Source is limited new Stream_Input_Source with record
       Socket : GNAT.Sockets.Socket_Type;
    end record;
 
--- source/xml/sax/xml/xml-sax-simple_readers.ads.orig	2012-10-08 15:25:42 UTC
+++ source/xml/sax/xml/xml-sax-simple_readers.ads
@@ -303,7 +303,7 @@ private
    type Simple_Shared_Locator is tagged;
    type Simple_Shared_Locator_Access is access all Simple_Shared_Locator'Class;
 
-   type SAX_Simple_Reader is new Ada.Finalization.Limited_Controlled
+   type SAX_Simple_Reader is limited new Ada.Finalization.Limited_Controlled
      and XML.SAX.Readers.SAX_Reader with
    record
       --  Handlers
--- source/league/league-environment_variables.ads.orig	2012-08-13 07:41:05 UTC
+++ source/league/league-environment_variables.ads
@@ -50,8 +50,7 @@ with League.Strings;
 package League.Environment_Variables is
 
    type Environment_Variable_Set is tagged private
-     with Iterator_Element  => League.Strings.Universal_String,
-          Constant_Indexing => Value;
+     with Iterator_Element  => League.Strings.Universal_String;
 
    procedure Clear (Self : in out Environment_Variable_Set'Class);
 
