diff -ruN common/src/mbus.c work.tom/rat-4.2.25/common/src/mbus.c
--- common/src/mbus.c	Mon Jun 16 18:57:59 2003
+++ common/src/mbus.c	Sun Jan  1 02:02:07 2006
@@ -265,9 +265,9 @@
 	mb_bufpos = mb_buffer + MBUS_AUTH_LEN;
       /* monster kludge */
       if(*dst == '(')
-	    sprintf(mb_bufpos, "\nmbus/1.0 %6d %ld%03ld %c (%s) %s ", seqnum, ts.tv_sec,ts.tv_usec/1000, reliable, src, dst);
+	    sprintf(mb_bufpos, "\nmbus/1.0 %6d %jd%03ld %c (%s) %s ", seqnum, (intmax_t)ts.tv_sec,ts.tv_usec/1000, reliable, src, dst);
       else
-	    sprintf(mb_bufpos, "\nmbus/1.0 %6d %ld%03ld %c (%s) (%s) ", seqnum, ts.tv_sec,ts.tv_usec/1000, reliable, src, dst);
+	    sprintf(mb_bufpos, "\nmbus/1.0 %6d %jd%03ld %c (%s) (%s) ", seqnum, (intmax_t)ts.tv_sec,ts.tv_usec/1000, reliable, src, dst);
 
       mb_bufpos += strlen(mb_bufpos);
 	if (ackseq == -1) {
diff -ruN rat/mbus_engine.c work.tom/rat-4.2.25/rat/mbus_engine.c
--- rat/mbus_engine.c	Mon Jan 12 12:14:38 2004
+++ rat/mbus_engine.c	Sun Jan  1 02:04:57 2006
@@ -604,11 +604,11 @@
 #endif
 			gettimeofday(&t, NULL);
 			gethostname(hname, 64);
-			sprintf(fname, "rat-%p-%ld.%06ld-%s-%s.log", sp, t.tv_sec, (long)t.tv_usec, hname, uname);
+			sprintf(fname, "rat-%p-%jd.%06ld-%s-%s.log", sp, (intmax_t)t.tv_sec, (long)t.tv_usec, hname, uname);
 			cname = rtp_get_sdes(sp->rtp_session[0], rtp_my_ssrc(sp->rtp_session[0]), RTCP_SDES_CNAME);
 
 			sp->logger = fopen(fname, "w");
-			fprintf(sp->logger, "tool_start %lu.%06lu ", t.tv_sec + SECS_BETWEEN_1900_1970, (long)t.tv_usec);
+			fprintf(sp->logger, "tool_start %jd.%06lu ", (intmax_t)t.tv_sec + SECS_BETWEEN_1900_1970, (long)t.tv_usec);
 			fprintf(sp->logger, "0x%08lx\n", (unsigned long) rtp_my_ssrc(sp->rtp_session[0]));
 		} else {
 			fclose(sp->logger);
diff -ruN rat/ui_send_audio.c work.tom/rat-4.2.25/rat/ui_send_audio.c
--- rat/ui_send_audio.c	Mon Jan 12 12:14:46 2004
+++ rat/ui_send_audio.c	Sun Jan  1 02:13:08 2006
@@ -279,7 +279,7 @@
 	if (sp->logger != NULL) {
 		struct timeval	t;
 		gettimeofday(&t, NULL);
-		fprintf(sp->logger, "silence    %lu.%06lu 0x%08lx %s\n", t.tv_sec + SECS_BETWEEN_1900_1970, t.tv_usec,
+		fprintf(sp->logger, "silence    %jd.%06lu 0x%08lx %s\n", (intmax_t)t.tv_sec + SECS_BETWEEN_1900_1970, t.tv_usec,
 		        (unsigned long) rtp_my_ssrc(sp->rtp_session[0]), name);
 	}
 }
@@ -371,8 +371,8 @@
 	if (sp->logger != NULL) {
 		struct timeval	t;
 		gettimeofday(&t, NULL);
-		fprintf(sp->logger, "channel    %lu.%06lu 0x%08lx redundancy %s\n",
-		        t.tv_sec + SECS_BETWEEN_1900_1970, t.tv_usec,
+		fprintf(sp->logger, "channel    %jd.%06lu 0x%08lx redundancy %s\n",
+		        (intmax_t)t.tv_sec + SECS_BETWEEN_1900_1970, t.tv_usec,
 		        (unsigned long) rtp_my_ssrc(sp->rtp_session[0]), scf->long_name);
 	}
 
@@ -453,8 +453,8 @@
 		if (sp->logger != NULL) {
 			struct timeval	t;
 			gettimeofday(&t, NULL);
-			fprintf(sp->logger, "channel    %lu.%06lu 0x%08lx none\n",
-				t.tv_sec+SECS_BETWEEN_1900_1970, t.tv_usec, (unsigned long) rtp_my_ssrc(sp->rtp_session[0]));
+			fprintf(sp->logger, "channel    %jd.%06lu 0x%08lx none\n",
+				(intmax_t)t.tv_sec+SECS_BETWEEN_1900_1970, t.tv_usec, (unsigned long) rtp_my_ssrc(sp->rtp_session[0]));
 		}
                 break;
         case 'r':
@@ -485,8 +485,8 @@
 	if (sp->logger != NULL) {
 		struct timeval	t;
 		gettimeofday(&t, NULL);
-		fprintf(sp->logger, "codec      %lu.%06lu 0x%08lx %s\n",
-			t.tv_sec+SECS_BETWEEN_1900_1970, t.tv_usec, (unsigned long) rtp_my_ssrc(sp->rtp_session[0]),
+		fprintf(sp->logger, "codec      %jd.%06lu 0x%08lx %s\n",
+			(intmax_t)t.tv_sec+SECS_BETWEEN_1900_1970, t.tv_usec, (unsigned long) rtp_my_ssrc(sp->rtp_session[0]),
 			pri_cf->long_name);
 	}
 }
