--- src/WEDImportExport/WED_DSFExport.cpp.orig	2017-04-14 14:21:02 UTC
+++ src/WEDImportExport/WED_DSFExport.cpp
@@ -685,7 +685,7 @@
 
 }
 
-static void	DSF_AccumChainBezier(
+static void	DSF_AccumChainBezier2p(
 						vector<Bezier2p>::const_iterator		start,
 						vector<Bezier2p>::const_iterator		end,
 						const Bbox2&							bounds,
@@ -802,7 +802,7 @@
 	}	
 }
 
-static void	DSF_AccumChain(
+static void	DSF_AccumChain2p(
 						vector<Segment2p>::const_iterator	start,
 						vector<Segment2p>::const_iterator	end,
 						const Bbox2&						bounds,
@@ -1370,7 +1370,7 @@
 						if(fac_is_auto_closed && bad_match(chain.front(),chain.back()))
 							problem_children.insert(what);
 						else
-							DSF_AccumChainBezier(chain.begin(),chain.end(), safe_bounds, cbs,writer, idx, fac->GetHeight(), fac_is_auto_closed);
+							DSF_AccumChainBezier2p(chain.begin(),chain.end(), safe_bounds, cbs,writer, idx, fac->GetHeight(), fac_is_auto_closed);
 					}
 				}
 				else
@@ -1391,7 +1391,7 @@
 						if(fac_is_auto_closed && bad_match(chain.front(),chain.back()))
 							problem_children.insert(what);
 						else						
-							DSF_AccumChain(chain.begin(),chain.end(), safe_bounds, cbs,writer, idx, fac->GetHeight(), fac_is_auto_closed);
+							DSF_AccumChain2p(chain.begin(),chain.end(), safe_bounds, cbs,writer, idx, fac->GetHeight(), fac_is_auto_closed);
 					}
 				}
 			}
