--- cutecommdlg.ui.orig	Thu Jul 15 01:11:45 2004
+++ cutecommdlg.ui	Thu Jul 15 01:12:07 2004
@@ -248,11 +248,6 @@
                     </item>
                     <item>
                         <property name="text">
-                            <string>576000</string>
-                        </property>
-                    </item>
-                    <item>
-                        <property name="text">
                             <string>921600</string>
                         </property>
                     </item>
