--- tools/linguist/shared/profileevaluator.h.orig	2008-04-28 09:11:21.000000000 -0400
+++ tools/linguist/shared/profileevaluator.h	2008-05-09 10:50:40.000000000 -0400
@@ -112,7 +112,12 @@
 private:
     QStringList expandPattern(const QString &pattern);
     void logMessage(const QString &msg, MessageType mt = MT_DebugLevel2);
-    void logMessage(MessageType mt, const char *msg, ...);
+    void logMessage(MessageType mt, const char *msg, ...)
+#if defined(Q_CC_GNU) && !defined(__INSURE__)
+    __attribute__ ((format (printf, 3, 4)))
+#endif
+    ;
+
     QStringList expandVariableReferences(const QString &value);
     QStringList evaluateExpandFunction(const QByteArray &func, const QString &arguments);
 
