diff -ur orig/femon-1.7.7/Makefile femon-1.7.7/Makefile
--- orig/femon-1.7.7/Makefile	2010-03-05 03:20:00.000000000 +0100
+++ femon-1.7.7/Makefile	2010-03-05 17:55:59.000000000 +0100
@@ -115,7 +115,11 @@
 ifndef FEMON_DEBUG
 	@$(STRIP) $@
 endif
+ifdef FREEBSD
+	@cp -f $@ $(LIBDIR)/$@.$(APIVERSION)
+else
 	@cp --remove-destination $@ $(LIBDIR)/$@.$(APIVERSION)
+endif
 
 dist: clean
 	@-rm -rf $(TMPDIR)/$(ARCHIVE)
Only in femon-1.7.7: Makefile.orig
