From 6262552171f7f32d28bb54a6c783bdd6578efa8c Mon Sep 17 00:00:00 2001
From: midipix <writeonce@midipix.org>
Date: Wed, 18 Jul 2018 05:23:28 -0400
Subject: build system: fix shared/static install target dependencies on
 install-pkgconf.

---
 sofort/pkgconf.mk | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/sofort/pkgconf.mk b/sofort/pkgconf.mk
index c28eae6..59a54e0 100644
--- a/sofort/pkgconf.mk
+++ b/sofort/pkgconf.mk
@@ -31,9 +31,9 @@ install-pkgconf: build/$(PACKAGE).pc
 	mkdir -p $(DESTDIR)$(LIBDIR)/pkgconfig
 	cp -p build/$(PACKAGE).pc    $(DESTDIR)$(LIBDIR)/pkgconfig
 
-package-install-shared: install-pkgconf
+install-shared: install-pkgconf
 
-package-install-static: install-pkgconf
+install-static: install-pkgconf
 
 .PHONY: .pkgconf install-pkgconf
 
-- 
cgit v0.11.2

