--- genconfig.sh.orig	2012-06-18 18:14:32.899227837 +0200
+++ genconfig.sh	2012-06-18 18:14:45.089227683 +0200
@@ -302,7 +302,7 @@
 
 echo "/* Uncomment the following line to enable generation of" >> ${CONFIGFILE}
 echo " * filter rules with pf */" >> ${CONFIGFILE}
-echo "/*#define PF_ENABLE_FILTER_RULES*/">> ${CONFIGFILE}
+echo "#define PF_ENABLE_FILTER_RULES">> ${CONFIGFILE}
 echo "" >> ${CONFIGFILE}
 
 echo "/* Uncomment the following line to enable caching of results of" >> ${CONFIGFILE}
