--- work/miniupnpd-1.4.20100921/genconfig.sh	2010-09-21 10:36:12.000000000 -0500
+++ genconfig.sh	2010-10-11 10:23:44.000000000 -0500
@@ -209,7 +209,7 @@
 
 echo "/* Uncomment the following line to enable generation of" >> ${CONFIGFILE}
 echo " * filter rules with pf */" >> ${CONFIGFILE}
-echo "/*#define PF_ENABLE_FILTER_RULES*/">> ${CONFIGFILE}
+echo "#define PF_ENABLE_FILTER_RULES">> ${CONFIGFILE}
 echo "" >> ${CONFIGFILE}
 
 echo "/* Uncomment the following line to enable caching of results of" >> ${CONFIGFILE}
