--- lib/MIME/Head.pm.orig	Sat Nov  4 20:54:46 2000
+++ lib/MIME/Head.pm	Wed Aug 25 11:12:25 2004
@@ -685,11 +685,16 @@
     This is the default value -- that is, "Content-Transfer-Encoding: 7BIT" 
     is assumed if the Content-Transfer-Encoding header field is not present.
 
+I do one other form of fixup: "7_bit", "7-bit", and "7 bit" are
+corrected to "7bit"; likewise for "8bit".
+
 =cut
 
 sub mime_encoding {
     my $self = shift;
-    lc($self->mime_attr('content-transfer-encoding') || '7bit');
+    my $enc = lc($self->mime_attr('content-transfer-encoding') || '7bit');
+    $enc =~ s{^([78])[ _-]bit\Z}{$1bit};
+    $enc;
 }
 
 #------------------------------
